<!-- comp_mut_writup_cut -->

<!-- ```{r results-resistance-mutations, echo = F, warning=F} -->

<!-- # **Resistance mutations, and how many explain the phenotype** -->
<!-- inh_n_samps_KRM <- nrow(subset(inh_bin, KRM_bin == "present")) -->
<!-- rif_n_samps_KRM <- nrow(subset(rif_bin, KRM_bin == "present")) -->

<!-- inh_n_samps_KRM_pc <- fmt_pc(nrow(subset(inh_bin, KRM_bin == "present"))/total_samps) -->
<!-- rif_n_samps_KRM_pc <- fmt_pc(nrow(subset(rif_bin, KRM_bin == "present"))/total_samps) -->

<!-- inh_n_samps_DST_res_KRM_present <- nrow(subset(inh_bin, dst == "1" & KRM_bin == "present")) -->
<!-- inh_n_samps_DST_res_KRM_present_pc <- fmt_pc(inh_n_samps_DST_res_KRM_present/tonum(inh_n_samps_DST_res)) -->

<!-- rif_n_samps_DST_res_KRM_present <- nrow(subset(rif_bin, dst == "1" & KRM_bin == "present")) -->
<!-- rif_n_samps_DST_res_KRM_present_pc <- fmt_pc(rif_n_samps_DST_res_KRM_present/tonum(rif_n_samps_DST_res)) -->

<!-- ``` -->

<!-- **Resistance mutations, and how many explain the phenotype** -->

<!-- *Summary* \ -->
<!-- Number INH samples DST resistant and have a known resistance mutation - `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) \ -->
<!-- Number RIF samples DST resistant and have a known resistance mutation - `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) \ -->

<!-- The total number of samples with a known resistance mutation in an INH target gene (*katG*, *fabG1*, *inhA*) was `r inh_n_samps_KRM` (`r inh_n_samps_KRM_pc`), and `r rif_n_samps_KRM` (`r rif_n_samps_KRM_pc`) for the RIF target gene (*rpoB*). -->

<!-- Explaining the DST resistance status, `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) samples had a known resistance mutation in an INH target gene, while `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) samples had an *rpoB* mutation (RIF). -->

<!-- **Compensatory mutations** -->

<!-- *summary* \ -->
<!-- Number INH samples with a CM - `r inh_3_a_c` \ -->
<!-- Number INH CM haplotypes - `r inh_n_CM_haplotypes` \ -->
<!-- Number INH unique CM - `r length(inh_CM_uniq)` \ -->
<!-- Number INH CM in known CM list - `r length(inh_CM_uniq_in_KCM)` \ -->
<!-- Number INH CM previously non-DR associated - `r length(inh_PCM)` \ -->

<!-- Number RIF samples with a CM - `r rif_3_a_c` \ -->
<!-- Number RIF CM haplotypes - `r rif_n_CM_haplotypes` \ -->
<!-- Number RIF unique CM - `r length(rif_CM_uniq)` \ -->
<!-- Number RIF CM in known CM list - `r length(rif_CM_uniq_in_KCM)` \ -->
<!-- Number RIF CM previously non-DR associated - `r length(rif_PCM)` \ -->

<!-- There were `r fmt(inh_3_a_c)` and `r fmt(rif_3_a_c)` samples with a compensatory mutation, of which `r fmt(inh_3_a)` and `r fmt(rif_3_a)` samples had a known resistance mutation, and `r fmt(inh_3_c)` and `r fmt(rif_3_c)` did not have known resistance mutations in INH and RIF, respectively.  -->

<!-- The number of compensatory mutation haplotypes were `r inh_n_CM_haplotypes` (INH) and `r rif_n_CM_haplotypes` (RIF), while the number of unique compensatory mutations were `r length(inh_CM_uniq)` (INH) and `r length(rif_CM_uniq)` (RIF). -->

<!-- Of these, `r length(inh_CM_uniq_in_KCM)` (INH) and `r length(rif_CM_uniq_in_KCM)` (RIF) were from compensatory mutations previously known, whereas `r length(inh_PCM)` INH and `r length(rif_PCM)` RIF compensatory mutations were putative.  -->

<!-- ```{r results-PRMs, echo = F, warning=F} -->

<!-- # **Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers** -->
<!-- inh_PRM <- select(subset(inh_bin, PRM_bin == "present"), PRM) -->
<!-- inh_n_PRM_haplotypes <- length(unique(inh_PRM$PRM)) -->
<!-- inh_PRM_uniq <- unique(unlist(strsplit(inh_PRM$PRM, "; "))) -->

<!-- rif_PRM <- select(subset(rif_bin, PRM_bin == "present"), PRM) -->
<!-- rif_n_PRM_haplotypes <- length(unique(rif_PRM$PRM)) -->
<!-- rif_PRM_uniq <- unique(unlist(strsplit(rif_PRM$PRM, "; "))) -->

<!-- ``` -->

<!-- **Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers** -->

<!-- *Summary* \ -->
<!-- Number INH samples with a PRM - `r nrow(inh_PRM)` \ -->
<!-- Number INH PRM haplotypes - `r inh_n_PRM_haplotypes` \ -->
<!-- Number INH unique PRM - `r length(inh_PRM_uniq)` \ -->

<!-- Number RIF samples with a PRM - `r nrow(rif_PRM)` \ -->
<!-- Number RIF PRM haplotypes - `r rif_n_PRM_haplotypes` \ -->
<!-- Number RIF unique PRM - `r length(rif_PRM_uniq)` \ -->

<!-- <!-- LAYER 4 --> -->
<!-- Number INH samples with KRM (*katG*) and CM and PRM - `r fmt(length(inh_4_a))`\ -->
<!-- Number INH samples with KRM (*katG*), no CM and PRM - `r fmt(length(inh_4_c))`\ -->
<!-- Number INH samples with no KRM (*katG*) and CM and PRM - `r fmt(length(inh_4_e))`\ -->
<!-- Number INH samples with no KRM (*katG*), no CM and PRM - `r fmt(length(inh_4_g))`\ -->

<!-- <!-- LAYER 5 -->  -->
<!-- Number INH samples with no KRM (*katG*), CM, no PRM, other *katG* - `r n_samps_rare_katg`\ -->
<!-- Number INH unique rare *katG* mutations - `r n_rare_katg`\ -->

<!-- Number RIF samples with no KRM (*ropB*), CM, no PRM, other *rpoB* - `r n_samps_rare_rpob`\ -->
<!-- Number INH unique rare *rpoB* mutations - `r n_rare_rpob`\ -->

<!-- After applying filters to the samples with compensatory mutations to find potential resistance mutations, and then re-searching the full TB-profiler database for samples with said mutations, `r nrow(inh_PRM)` INH isolates were found. In these INH samples, there were `r inh_n_PRM_haplotypes`, with `r length(inh_PRM_uniq)` unique mutations (see **Table \@ref(tab:PRM-summary-table) and Table \@ref(tab:PRM-stats-table)**). No samples were found with potential resistance mutations in *rpoB* after phylogenetic analysis revealed a small number of potential resistance mutations were clearly lineage-specific (`r print_vect(rif_lin_PRM)` and therefore unlikely to be a resistance mutation.  -->

<!-- Within the INH samples with putative resistance mutations, `r fmt(length(inh_4_c))` samples had a known *katG* mutation but no compensatory mutations (see **Table \@ref(tab:PRM-KRM-other-table)**), `r fmt(length(inh_4_e))` samples had no known *katG* mutation, but did have a compensatory mutation, while the majority, `r fmt(length(inh_4_g))` had no known *katG* mutation and no compensatory mutations (see **Figure \@ref(fig:samples-breakdown)**). -->

<!-- Further to the potential resistance mutations, there were also found `r n_samps_rare_katg` (INH) and `r n_samps_rare_rpob` (RIF) samples with no known *katG*/*rpoB* resistance mutations, having a compensatory mutation, but not having a potential resistance mutation by the criteria outlined in *Methods*, nevertheless with a non-drug-resistance-associated *katG*/*rpoB* mutation ('rare' mutation). These rare mutations numbered `r n_rare_katg` in *katG* and `r n_rare_rpob` for *rpoB* (see **Table \@ref(tab:rare-mutations-table)**). -->

<!-- A small number of rare known resistance mutations were selected for sensitivity analysis to verify that these could be recovered by our methods. These mutations occurred in just 7, 6, 5, 4 or 3 samples. When the analysis was re-run with the mutations uncategorised, all subsequently appeared as 'potential resistance mutations'.  -->


<!-- **katG potential resistance mutations - co-occurrence with other resistance mutations and comparison to Ser315Thr** -->

<!-- *Summary* \ -->
<!-- Number of *katG* PRM haplotypes occurring with a KRM - `r inh_n_PRM_hap_KRM` (`r fmt_pc(inh_n_PRM_hap_KRM/inh_n_PRM_haplotypes)`, `r sum(inh_PRM_KRM$n)` samples) \ -->
<!-- Most common KRM occurring with *katG* PRM - `r inh_most_common_KRM_with_PRM` -->

<!-- Number of *katG* PRM haplotypes occurring with an unknown mutation -  `r inh_n_PRM_hap_other` (`r fmt_pc(inh_n_PRM_hap_other/inh_n_PRM_haplotypes)`, `r sum(inh_PRM_other$n)` samples) \ -->
<!-- Most common unknown occurring with *katG* PRM - `r inh_most_common_other_with_PRM` -->


<!-- Mutation *katG*-Ser315Thr incurs little fitness cost and so ought not to co-occur with *ahpC* compensatory mutations. To compare the fitness cost of *katG*-Ser315Thr and our potential resistance mutations, **Table \@ref(tab:fitness-table)** shows comparison of presence/absence of compensatory mutations in samples with (only) *katG*-Ser315Thr and samples with the potential resistance mutations (samples with potential resistance mutations and co-occurring known resistance mutations in *katG* were excluded). A Chi-square test showed a significant difference (`r fitness_x_sq`), with compensatory mutations occurring in a higher proportion in the potential resistance mutations than *katG*-p.Ser315Thr, indicating on average they incur a greater fitness cost.  -->

<!-- *katG*-Ser315Thr is also useful in assessing the resistance level of the potential resistance mutations. Since *katG*-Ser315Thr confers resistance to INH in an all-or-none way, a minority of *katG*-Ser315Thr samples should have co-occurrences with other known resistance mutations, and this is indeed the case - see **Table \@ref(tab:resistance-table)**. Nearly half of the samples with a potential resistance mutation showed co-occurrence with a known resistance mutation, again excluding those with co-occurrence in *katG*, suggesting the potential resistance mutations on average confer a lesser resistance level than *katG*-Ser315Thr. This difference was also significant (`r resistance_x_sq`).  -->

<!-- **Table \@ref(tab:PRM-KRM-other-table)** shows the full list of co-occurrences of the potential resistance mutation haplotypes and co-occurrence of known resistance mutations and unknown mutations in relevant resistance genes (i.e. co-occurrence of *katG* potential resistance mutations with known/unknown *ahpC*, *inhA*, *fabG1* and *kasA*). There were `r inh_n_PRM_hap_KRM` (`r fmt_pc(inh_n_PRM_hap_KRM/inh_n_PRM_haplotypes)`) *katG* potential resistance mutation haplotypes occurring with a known resistance mutation (`r sum(inh_PRM_KRM$n)` samples), the most common of which was `r inh_most_common_KRM_with_PRM`. -->


<!-- ------ -->

<!-- ```{r PRM-stats-table, echo = F, warning=F} -->

<!-- cap <- "Number of samples, DST sensitive proportion, (predicted) DR type proportion and number of lineages for each PRM" -->
<!-- knitr::kable(PRM_stats, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->


<!-- ```{r rare-counts-table, echo = F, warning=F} -->

<!-- cap <- "Counts of occurrence for *katG* (isoniazid) and *rpoB* (rifampicin) among samples with no KRM, at least one CM, but no PRM. These variants are not counted as ‘PRM’ because of their low frequency (only one or two samples)." -->

<!-- # Variants occurring more than twice were rejected because most samples were DR-type sensitive or DST sensitive, or in the case of INH, the samples already have a known resistance mutation in a gene other than *katG* -->

<!-- knitr::kable(rare_cnts, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->


<!-- ```{r fitness-table, echo = F, warning=F} -->

<!-- cap <- "Sample counts comparing presence and absence of compensatory mutations in *katG*-Ser315Thr and potential resistance mutations. Samples with known *katG* resistance mutations co-occurring with potential resistance mutations were excluded from counts." -->
<!-- knitr::kable(fitness_table, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->

<!-- ```{r resistance-table, echo = F, warning=F} -->

<!-- cap <- "Samples counts of presence/absence of known resistance mutations co-occurring with *katG*-Ser315Thr and the potential resistance mutations. Samples with known *katG* resistance mutations co-occurring with potential resistance mutations were excluded from counts." -->
<!-- knitr::kable(resistance_table, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->

<!-- ```{r isoniazid-Ser315Thr-table, echo = F, warning=F} -->

<!-- cap <- "Comparison to katG-Ser315Thr and katG-Ser315Thr-KRM in other isoniazid resistance genes." -->
<!-- knitr::kable(ser315thr_pivot, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->

<!-- ```{r fitness-plot, echo = F, warning=F, fig.cap="Percentage of samples with compensatory mutations among the potential resistance mutations. *katG*-Ser315Thr is shown for comparison. This mutation incurs low fitness cost, hence does not generally co-occur in samples with a compensatory mutation. Red line - median percentage of samples with compensatory mutations in samples with potential resistance mutations. Samples with *katG*-Ser315Thr and a potential resistance mutation have been excluded. Samples with a potential resistance mutation and a known *katG* resistance mutation have been excluded."} -->

<!-- fitness_plot -->

<!-- ``` -->


<!-- ```{r samples-breakdown-rif, echo=FALSE, out.width = "2000px", fig.cap="Breakdown of sample numbers by presence of mutation type for INH" } -->

<!-- knitr::include_graphics(rif_numbers_file) -->

<!-- ``` -->





















