---
title: "Large-scale genomic analysis of *Mycobacterium tuberculosis* reveals extent of target and compensatory mutations linked to isoniazid, rifampicin, and multi-drug resistance"
output: 
  bookdown::word_document2: 
    fig_caption: yes
    table_caption: yes
    number_sections: no
mainfont: Calibri
fontsize: 10pt
bibliography: all_refs.bib
csl: biomed-central.csl
---

```{r setup, include=FALSE, echo = F, warning=F, message=F}

# SETUP ----

rm(list = ls())

knitr::opts_chunk$set(echo = F)
# options(scipen=1, digits=2)
options(scipen=999, digits = 3)
table_font_sz <- 8

```

```{r library, include=FALSE, echo = F, warning=F, message=F}

library(tidyr)
library(plyr)
library(dplyr)
library(knitr)
library(reshape2)
library(janitor)
library(ggsankey)
library(ggplot2)
library(stringr)

```

```{r functions, echo = F, warning=F, message=F}

source("https://raw.githubusercontent.com/GaryNapier/Packages_functions/master/Functions.R")

uniq_col <- function(x, split = "; "){
  unique(unlist(strsplit(x, split)))
}

col_pc <- function(df, col){
  round((df[, col] / sum(df[, col]))*100, 3)
}

len_uniq <- function(x){
  length(unique(x))
}

rbind_force <- function(df1, df2){
  rbind(df1, setNames(df2, names(df1)))
}

# Make quick pivot table summarising total unique values by column
# my_formula arg must be call to the formula() function: formula(<col> ~ 'n')
# Result:
#             PRM            n
#  katG-c.2223A>G  3   (4.05%)
# katG-p.Arg484His  5   (6.76%)
# ...
# katG-p.Tyr413Cys  8  (10.81%)
#           Total 74 (100.00%)
pivot <- function(x, my_formula, value_var = "wgs_id"){
  drop_cols(to_table(reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq),
                     pc_dir = 'col'), 
            'Total')
}

pivot_num <- function(x, my_formula,value_var = "wgs_id"){
  reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq)
}

split_pivot <- function(df, my_formula, leave_col, value_var = "wgs_id", split_on_col = "Drug"){
  
  to_table_split_pivot <- function(x, col){
    x[[col]] <- as.character(x[[col]])
    x <- drop_cols(to_table(x, "col"), "Total")
    x[[col]] <- gsub("\\s*\\([^\\)]+\\)", "", x[[col]])
    x
  }
  
  df_pivot <- reshape2::dcast(df, my_formula, value.var = 'wgs_id', fun.aggregate = len_uniq)
  df_split <- split(df_pivot, df_pivot[, split_on_col])
  
  do.call('rbind', lapply(df_split, to_table_split_pivot, leave_col))
  
}

duped <- function(x, col){
  dups <- x[duplicated(x[col]), col]
  x[x[, col] %in% dups, ]
}

# Swap first and second parts of string, splitting on a charachter (or substring)
# e.g. 
# x <- "katG-p.Ile335Val; katG-p.Ser315Thr"
# swap_str(x, "; ")
# "katG-p.Ser315Thr; katG-p.Ile335Val"
swap_str <- function(x, split_chr = "; "){
  paste0(unlist(strsplit(x, split_chr))[c(2, 1)], collapse = split_chr)
}


# Print each element of vector vertically instead of across the screen
# e.g. 
# > x <- c("a", "b", "c")
# > x
# [1] "a" "b" "c"
# > print_vert(x)
# a
# b
# c
print_vert <- function(x){
  cat(paste0(x, '\n'))
}

# Remove change from second gene-change pair
# e.g.
# x <- "katG-p.Ser315Thr; fabG1-c.-15C>T"
# > drop_change(x, "-")
# [1] "katG-p.Ser315Thr; fabG1"
drop_change <- function(string, split_on = "-"){
  unlist(lapply(strsplit(string, split_on), function(x){paste0(x[c(1, 2)], collapse = split_on)}))
}

clean_binary_table <- function(x){
  x <- data.frame(apply(x, 2, function(x){gsub("\\[\\]", NA, as.character(x))}))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\[||\\]", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\), \\(", "; ", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\', \\'", "-", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\(||\\)", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\'", "", as.character(x)) }))
  x
}

sort_pos <- function(x, col){
  x$pos <- as.numeric(stringr::str_extract(x[, col], "[0-9]+"))
  x <- x[order(x$pos), ]
  drop_cols(x, 'pos')
}

lapply_nms <- function(x, fun, nms){
  res <- lapply(x, fun)
  names(res) <- nms
  res
}

drug_col_front <- function(df_list){
  lapply(df_list, function(x){ select(x, Drug, everything()) })
}

add_drug_col <- function(df_list, drugs){
  drug_col_front(lapply_mod(df_list, function(i){df_list[[i]]$Drug <- rep(drugs[i], nrow(df_list[[i]])); df_list[[i]]}))
}

find_pos <- function(x){
  sub("\\-", "", gsub("[^0-9\\-]+", "", x))
}

fmt_pc <- function(x, ...){
  paste0(fmt(x*100, ...), "%")
}

trimws_df <- function(x){
  data.frame(sapply(x, trimws))
}

tonum <- function(x){
  as.numeric(gsub(",", "", strsplit(x, " ")[[1]][1]))
}

uniq_vars <- function(x){
  unique(unlist(strsplit(x, "; ")))
}

rm_null <- function(x){
  x[!sapply(x,is.null)] 
}

rm_norow <- function(x){
  x[sapply(x, nrow) > 0]
}

rm_na <- function(x){
  x[!is.na(x)]
}

pivot_numeric <- function(x, my_formula, value_var = "wgs_id"){
  reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq)
}

print_chisq <- function(chisq_result, scientific = T, rnd = 3){
  paste0("X-sq = ", round(chisq_result$statistic, rnd), 
        "; df = ", round(chisq_result$parameter, rnd), 
        "; p = ", format(chisq_result$p.value, scientific = scientific))
}

clean_del_ins_dup <- function(x){
  gsub("(del|ins|dup).*", "\\1", x)
}

drop_col_name <- function(x, name){
  # Return col names except named col names
  names(x)[!(names(x) %in% name)]
}

rbind_mod <- function(df1, df2, names){
  rbind(setNames(df1, names), setNames(df2, names))
}


split_df_on_haplotype <- function(df, col){
  # Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
  data.frame(df %>% mutate(col = strsplit(as.character(col), "; ")) %>% unnest(col))
}

strsplit_mod <- function(x,
                     split,
                     type = "remove",
                     perl = FALSE,
                     ...) {
  if (type == "remove") {
    # use base::strsplit
    out <- base::strsplit(x = x, split = split, perl = perl, ...)
  } else if (type == "before") {
    # split before the delimiter and keep it
    out <- base::strsplit(x = x,
                          split = paste0("(?<=.)(?=", split, ")"),
                          perl = TRUE,
                          ...)
  } else if (type == "after") {
    # split after the delimiter and keep it
    out <- base::strsplit(x = x,
                          split = paste0("(?<=", split, ")"),
                          perl = TRUE,
                          ...)
  } else {
    # wrong type input
    stop("type must be remove, after or before!")
  }
  return(out)
}

get_gene <- function(x, split_on = "-"){
  unlist(lapply(strsplit(x, split_on), function(x){x[1]}))
}

drop_gene <- function(x, split_on = "-"){
unlist(lapply(strsplit_mod(x, split_on, "before"), function(x){
  if(length(x) < 3){
    x <- paste0(x[2:length(x)], collapse = "")
    x <- gsub("-p.", "", x)
    gsub("-c.", "", x)
  }else{
    x <- paste0(x[3:length(x)], collapse = "")
    x <- gsub("-p.", "", x)
    gsub("-c.", "", x)
  }
  }))
}

try_read_csv <- function(x){
  tryCatch({read.csv(x)}, error = function(e){})
}

```

```{r variables, echo = F, warning=F, message=F}


# VARIABLES ----

id_col <- "wgs_id"
rnd <- 3
# drugs <- c("rifampicin", "isoniazid", "ethambutol", "pyrazinamide", "streptomycin", 
#              "ofloxacin", "moxifloxacin", "levofloxacin", "amikacin", "kanamycin", "capreomycin", "ciprofloxacin", "prothionamide", 
#              "ethionamide", "para_aminosalicylic_acid", "cycloserine",
#              "clarithromycin", "clofazimine", "bedaquiline", "linezolid", "rifabutin", "delamanid")
drugs <- c("isoniazid", "rifampicin")
fq <- "fluoroquinolones, ofloxacin, levofloxacin, ciprofloxacin, moxifloxacin"
agc <- "capreomycin, streptomycin, amikacin, kanamycin, aminoglycosides"
align <- c("l", "r", "r")
blue_col <- "cornflowerblue"
plot_text_sz <- 0.5
main_cex <- 1.5
# dr_status_vect <- c("Sensitive", "Pre-MDR", "MDR", "Pre-XDR", "XDR", "Other")
dr_status_vect <- c("Sensitive", "Pre-MDR-TB", "HR-TB", "RR-TB", "MDR-TB", "Pre-XDR-TB", "XDR-TB", "Other")
rif_lin_PRM <- c("rpoB-p.Val695Leu", "rpoB-c.-199C>T", "rpoB-p.Ile925Val")

```

```{r paths, echo = F, warning=F, message=F}

# PATHS ----

setwd("~/Documents/comp_mut/r_scripts/")

metadata_path <- "../../metadata/"
local_metadata_path <- "../metadata/"
pipeline_db_path <- "../../pipeline/db/"
db_path <- "../../tbdb/"
results_path <- "../results/"
newick_path <- paste0(results_path, "newick/")

```

```{r files, echo = F, warning=F, message=F}

# FILES ----

# Metadata
metadata_file <- paste0(metadata_path, "tb_data_18_02_2021.csv")
# Database data
lineage_conversion_file <- paste0(pipeline_db_path, "lineage_conversions.txt")
dr_genes_file <- paste0(db_path, "tbdb.csv")
KCM_file <- paste0(pipeline_db_path, "compensatory_mutations.csv")
CM_RM_gene_assoc_file <- paste0(local_metadata_path, "CM_RM_gene_assoc.csv")

```

```{r drug-files, echo = F, warning=F, message=F}

binary_table_files <- paste0(results_path, drugs, "_binary_table.csv")
PCM_KCM_merged_files <- paste0(results_path,  drugs, "_PCM_KCM_merged.csv")
summary_files <- paste0(results_path, drugs, "_summary.csv")
PRM_stats_files <- paste0(results_path, drugs, "_PRM_stats.csv")
tree_files <- paste0(results_path, drugs, "_tree.png")
inh_numbers_file <- paste0(results_path, "isoniazid_numbers.png")
rif_numbers_file <- paste0(results_path, "rifampicin_numbers.png")

```

```{r read-in-data, echo = F, warning=F, message=F}

# READ IN DATA ----

# Metadata
metadata <- read.csv(metadata_file)
# Database data
lineage_conversions <- read.delim(lineage_conversion_file)
# dr_genes <- read.delim(dr_genes_file, header = F)
dr_genes <- read.csv(dr_genes_file)
KCM <- read.csv(KCM_file)
CM_RM_gene_assoc <- read.csv(CM_RM_gene_assoc_file)

```

```{r drugs-read-in-data, echo = F, warning=F, message=F}

binary_tables <- lapply_nms(binary_table_files, read.csv, drugs)
PCM_KCM_merged <- lapply_nms(PCM_KCM_merged_files, read.csv, drugs)
# summaries <- lapply_nms(summary_files, read.csv, drugs)
PRM_stats <- rm_null(lapply_nms(PRM_stats_files, try_read_csv, drugs))

```

```{r clean-data, echo = F, warning=F, message=F}

# Binary tables 

# Remove all the "[]" present in the binary table caused by python's lists
binary_tables <- lapply(binary_tables, clean_binary_table)

# Clean lineage column 
binary_tables <- lapply(binary_tables, function(x) {x$main_lineage <- gsub("lineage", "", x$main_lineage); x})
binary_tables <- lapply(binary_tables, function(x) {x$sublin <- gsub("lineage", "", x$sublin); x})

# Trim white space
binary_tables <- lapply(binary_tables, trimws_df)

# Rifampicin filter - remove lineage-specific PRMS
# NB - TOTAL HACK TO REMOVE THE LINEAGE-SPECIFIC RIF PRMs
binary_tables$rifampicin$PRM <- ifelse(binary_tables$rifampicin$PRM %in% rif_lin_PRM,
       NA,
       binary_tables$rifampicin$PRM)

# Make binary cols for CM KRM and PRM
binary_tables <- lapply(binary_tables, function(x){
  data.frame(cbind(x, 
                   CM_bin = ifelse(is.na(x$CM), "absent", "present"), 
                   KRM_bin = ifelse(is.na(x$KRM), "absent", "present"), 
                   PRM_bin = ifelse(is.na(x$PRM), "absent", "present")))
})

# Make katG and rpoB binary
binary_tables$isoniazid$KRM_katG_bin <- ifelse(is.na(binary_tables$isoniazid$KRM_katG), "absent", "present")
binary_tables$rifampicin$KRM_rpoB_bin <- ifelse(is.na(binary_tables$rifampicin$KRM_rpoB), "absent", "present")

# ---

# Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
binary_tables_dup <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>%
  unnest(other_vars))
})

# Make col for just the other_vars genes, stripping out the change
binary_tables_dup <- lapply(binary_tables_dup, function(x){x$other_vars_gene <- gsub("-.*", "", x$other_vars); x})

# ---

# Split out as sep vars

inh_bin <- binary_tables$isoniazid
rif_bin <- binary_tables$rifampicin

inh_bin_dup <- binary_tables_dup$isoniazid
rif_bin_dup <- binary_tables_dup$rifampicin

# --- 

# KCM - combine gene-change
KCM$gene_change <- paste0(KCM$Gene, "-", KCM$Mutation)

# Subset KCM 
inh_KCM <- KCM[KCM[,"Drug"] == "isoniazid", ]
rif_KCM <- KCM[KCM[,"Drug"] == "rifampicin", ]


# Put PCM_merged together and join in the drug names
PCM_KCM_merged <- do.call("rbind", PCM_KCM_merged)
PCM_KCM_merged$gene_change <- paste0(PCM_KCM_merged$gene, "-", PCM_KCM_merged$change)
PCM_KCM_merged <- merge(PCM_KCM_merged, CM_RM_gene_assoc, 
                    by.x = "gene", by.y = "CM_gene", 
                    sort = F)
# Merge in status
PCM_KCM_merged$status <- ifelse(PCM_KCM_merged$gene_change %in% KCM$gene_change, "known", "unknown")


# Remove the PCMs in the KCM
# PCM_KCM_merged <- PCM_KCM_merged[!(PCM_KCM_merged$gene_change %in% KCM$gene_change), ]

# Transpose summary
# summaries <- do.call('cbind', lapply(summaries, function(x){data.frame(n = t(x))}))
# names(summaries) <- drugs



```

```{r isoniazid-rare-mutations-cnts, echo = F, warning=F}

isoniazid_binary_table <- binary_tables$isoniazid
rare_katg_table <- subset(isoniazid_binary_table, is.na(KRM_katG) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_katG)) )
n_samps_rare_katg <- nrow(rare_katg_table)
n_rare_katg <- length(uniq_col(rare_katg_table$other_katG))
rare_katg_cnts <- tab2df(table(table(unlist(strsplit(rare_katg_table$other_katG, "; ")))))
rare_katgs <- sort_pos(data.frame(Change = uniq_col(rare_katg_table$other_katG)))

```

```{r rifampicin-rare-mutations-cnts, echo = F, warning=F}

rifampicin_binary_table <- binary_tables$rifampicin
rare_rpob_table <- subset(rifampicin_binary_table, is.na(KRM_rpoB) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_rpoB)) )
n_samps_rare_rpob <- nrow(rare_rpob_table)
n_rare_rpob <- length(uniq_col(rare_rpob_table$other_rpoB))
rare_rpob_cnts <- tab2df(table(table(unlist(strsplit(rare_rpob_table$other_rpoB, "; ")))))
rare_rpobs <- sort_pos(data.frame(Change = uniq_col(rare_rpob_table$other_rpoB)))

```

```{r rare-mutations-cnts, echo = F, warning=F}

rare_cnts <- plyr::rbind.fill(rare_katg_cnts, rare_rpob_cnts)

for(i in seq(nrow(rare_cnts))){
  row.names(rare_cnts)[row.names(rare_cnts) == i] <- drugs[i]
}

```

```{r table-1, echo = F, warning=F}

# Make pivot tables of basic stats

# Add drug col, select and bind
meta_tables <- add_drug_col(binary_tables, drugs)
meta_table <- do.call('rbind', lapply(meta_tables, 
                                      function(x){ 
                                        select(x, Drug, wgs_id, main_lineage, sublin, country_code, drtype, dst)
                                        }))

# Trim whitespace
meta_table$dst <- trimws(meta_table$dst)

# Lin
lin_pivot <- pivot(meta_table, formula(main_lineage ~ 'n') )

# DR
drtype_pivot <- pivot(meta_table, formula(drtype ~ 'n') )
drtype_pivot <- trimws_df(drtype_pivot[order(match(drtype_pivot$drtype, dr_status_vect)), ])

# # Predicted DR to individual drugs
KRM_table <- do.call('rbind', lapply(meta_tables, function(x){ select(x, Drug, wgs_id, KRM)}))
KRM_table$KRM_binary <- ifelse(is.na(KRM_table$KRM), 0, 1)
KRM_pivot <- trimws_df(split_pivot(KRM_table, formula(Drug + KRM_binary ~ 'n'), "KRM_binary"))

# DST
# dst_pivot <- pivot(meta_table, formula(Drug + dst ~ 'n') )
dst_pivot <- trimws_df(split_pivot(meta_table, formula(Drug + dst ~ 'n'), "dst"))


table_1 <- data.frame()
lin_dr <- rbind(table_1,
                setNames(lin_pivot, names(table_1)),
                setNames(drtype_pivot, names(table_1)))

lin_dr <- cbind(rep("-", nrow(lin_dr)), lin_dr)

                 
KRM_dst <- rbind(table_1,
                 setNames(KRM_pivot, names(table_1)),
                 setNames(dst_pivot, names(table_1)))

table_1 <- rbind_force(lin_dr, KRM_dst)

names(table_1) <- c("Drug", "Chararcteristic", "n (%)")

table_1$Category <- c(rep("Lineage", nrow(lin_pivot)),
                      rep("DR status", nrow(drtype_pivot)),
                      rep("(predicted) resistance", nrow(KRM_pivot)),
                      rep("DST", nrow(dst_pivot)))

table_1 <- table_1[, c("Category", "Drug", "Chararcteristic", "n (%)")]

```

```{r n-samples-tree-inh, echo = F, warning=F}

# Tree

# INH
# # 1
# total_samps
# # 2
inh_2_a <- nrow(subset(inh_bin, KRM_katG_bin == "present"))
inh_2_b <- nrow(subset(inh_bin, KRM_katG_bin == "absent"))
inh_2_a_b <- inh_2_a + inh_2_b
# # 3
inh_3_a <- nrow(subset(inh_bin, KRM_katG_bin == "present" & CM_bin == "present"))
inh_3_b <- nrow(subset(inh_bin, KRM_katG_bin == "present" & CM_bin == "absent"))
inh_3_c <- nrow(subset(inh_bin, KRM_katG_bin == "absent" & CM_bin == "present"))
inh_3_d <- nrow(subset(inh_bin, KRM_katG_bin == "absent" & CM_bin == "absent"))
inh_3_a_c <- inh_3_a + inh_3_c
# 4
# a
inh_4_a <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present"))
# b
inh_4_b <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# c
inh_4_c <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# d
inh_4_d <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# e
inh_4_e <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present"))
# f
inh_4_f <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# g
inh_4_g <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# h
inh_4_h <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent"))

inh_4_c_g <- inh_4_c + inh_4_g
inh_4_a_c_e_g <- inh_4_a + inh_4_c + inh_4_e + inh_4_g
# 5
# a
inh_5_a <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           !(is.na(other_katG))))
# b
inh_5_b <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           is.na(other_katG)))

layer_2 <- data.frame(n = c(inh_2_a, inh_2_b))
layer_3 <- data.frame(n = c(inh_3_a, 
                            inh_3_b, 
                            inh_3_c, 
                            inh_3_d))
layer_4 <- data.frame(n = c(inh_4_a, 
                            inh_4_b, 
                            inh_4_c, 
                            inh_4_d, 
                            inh_4_e, 
                            inh_4_f, 
                            inh_4_g, 
                            inh_4_h))
layer_5 <- data.frame(n = c(inh_5_a, inh_5_b))

n_df <- rbind(layer_2, layer_3, layer_4, layer_5)

layer <- data.frame(c(
  paste0(rep("2", nrow(layer_2)), "_", letters[1:nrow(layer_2)]), 
  paste0(rep("3", nrow(layer_3)), "_", letters[1:nrow(layer_3)]), 
  paste0(rep("4", nrow(layer_4)), "_", letters[1:nrow(layer_4)]), 
  paste0(rep("5", nrow(layer_5)), "_", letters[1:nrow(layer_5)])
))

inh_samp_tree <- setNames(cbind(layer, n_df), c("layer", "n"))

```

```{r n-samples-tree-rif, echo = F, warning=F}



# # RIF
# # 1
# total_samps
# # 2
rif_2_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present"))
rif_2_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent"))
# 3
rif_3_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present"))
rif_3_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent"))
rif_3_c <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present"))
rif_3_d <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent"))
rif_3_a_c <- rif_3_a + rif_3_c
# 4
# a
rif_4_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present"))
# b
rif_4_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# c
rif_4_c <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# d
rif_4_d <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# e
rif_4_e <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present"))
# f
rif_4_f <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# g
rif_4_g <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# h
rif_4_h <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# 5
# a
rif_5_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           !(is.na(other_rpoB))))
# b
rif_5_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           is.na(other_rpoB)))

layer_2 <- data.frame(n = c(rif_2_a, rif_2_b))
layer_3 <- data.frame(n = c(rif_3_a, 
                            rif_3_b, 
                            rif_3_c, 
                            rif_3_d))
layer_4 <- data.frame(n = c(rif_4_a, 
                            rif_4_b, 
                            rif_4_c, 
                            rif_4_d, 
                            rif_4_e, 
                            rif_4_f, 
                            rif_4_g, 
                            rif_4_h))
layer_5 <- data.frame(n = c(rif_5_a, rif_5_b))

n_df <- rbind(layer_2, layer_3, layer_4, layer_5)

layer <- data.frame(c(
  paste0(rep("2", nrow(layer_2)), "_", letters[1:nrow(layer_2)]), 
  paste0(rep("3", nrow(layer_3)), "_", letters[1:nrow(layer_3)]), 
  paste0(rep("4", nrow(layer_4)), "_", letters[1:nrow(layer_4)]), 
  paste0(rep("5", nrow(layer_5)), "_", letters[1:nrow(layer_5)])
))

rif_samp_tree <- setNames(cbind(layer, n_df), c("layer", "n"))


```

```{r total-samps-and-dst, echo = F, warning=F}

total_samps <- nrow(binary_tables[[1]])

inh_n_samps_dst <- nrow(subset(meta_table, Drug == "isoniazid" & !is.na(dst)))
rif_n_samps_dst <- nrow(subset(meta_table, Drug == "rifampicin" & !is.na(dst)))

inh_n_samps_DST_res <- nrow(subset(meta_table, Drug == "isoniazid" & dst == "1"))
rif_n_samps_DST_res <- nrow(subset(meta_table, Drug == "rifampicin" & dst == "1"))

```

```{r working, echo = F, warning=F}


```

| | |
| ----------- | ----------- |
|Gary Napier^1^ | gary.napier@lshtm.ac.uk |
|Julian Libiseller-Egger^1^ | julian.libiseller-egger@lshtm.ac.uk | 
|Susana Campino^1^ | susana.campino@lshtm.ac.uk |
|Jody E. Phelan^1^,* | jody.phelan@lshtm.ac.uk |
|Taane G. Clark^1^,^2^,* | taane.clark@lshtm.ac.uk |

^1^Faculty of Infectious and Tropical Diseases, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK
^2^Faculty of Epidemiology and Population Health, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK

\* joint authors


## ABSTRACT

**Background:** Resistance to the first-line drugs isoniazid (INH) and rifampicin (RIF) in *Mycobacterium tuberculosis* (multidrug-resistant TB; MDR-TB) threatens tuberculosis disease (TB) control. Mutations in *katG* and *rpoB*, which confer resistance to INH and RIF, often come with a fitness cost. To overcome these fitness costs, M. tuberculosis compensatory mutations have arisen in *rpoC*, *rpoA* (RIF) and *ahpC* (INH) genes. Here we leverage the presence of known and putative compensatory mutations along with phenotypic and genotypic data to detect the presence of novel resistance mutations occurring in the target genes of INH and RIF.

**Results:** Across ~32k isolates, there were `r fmt(inh_n_samps_DST_res)` (`r fmt_pc(inh_n_samps_DST_res/inh_n_samps_dst, digits = 3)`) with INH and `r fmt(rif_n_samps_DST_res)` (`r fmt_pc(rif_n_samps_DST_res/rif_n_samps_dst, digits = 3)` with RIF phenotypic resistance. Known mutations in *katG* and *rpoB* explained most of the resistance. However, `r fmt(inh_3_c)` (`r fmt_pc(inh_3_c/total_samps, digits = 1)`) and `r fmt(rif_3_c)` (`r fmt_pc(rif_3_c/total_samps, digits = 2)`) isolates had *ahpC* and *rpoC* compensatory mutations, respectively, but did not have any known resistance mutations in INH or RIF target genes. By examining isolates with compensatory mutations but no known resistance mutation, we identified a number of inferred novel resistance mutations. We also characterise the putative *katG* mutations in terms of their co-occurrence with other INH variants - *katG*-Ser315Thr and other INH genes, e.g. *fabG1*. 

**Conclusions:** Detecting resistance mutations from whole genome sequencing can inform clinical management, but also revealed new variants linked to resistance. We have identified new resistance markers and, after validation, have added to a list of mutations for genotypic drug-resistance predictions.

**URL:**https://github.com/GaryNapier/comp_mut

**Keywords:** *Mycobacterium tuberculosis*, isoniazid, rifampicin, *katG*, *rpoB*, compensatory/resistance mutations

## BACKGROUND


Tuberculosis, caused by *Mycobacterium tuberculosis* bacteria, is a major global public health problem, with control made difficult by drug resistance. In gaining resistance against anti-tuberculosis drugs, drug target or activating proteins are often compromised in their usual function, and the bacterium therefore also incurs a fitness cost. Costs can be manifest as a phenotypic difference, e.g. reduced virulence or transmissibility. For example, the *katG* gene codes for the KatG enzyme, a catalase-peroxidase which protects the bacterium from reactive oxygen species (ROS) damage and to detoxify hydrogen peroxide [@Trivedi2012], two products of the immune system which attack the pathogen. The enzyme also activates the pro-drug isoniazid (INH), converting it to multiple reactive species [@Zhang2009]. 

When mutations occur in the *katG* gene that disrupt INH binding to KatG, the bacteria is often left simultaneously with drug resistance and a protein with impaired enzymatic function. This is not always the case, since some mutations can confer drug resistance without a punitive fitness cost. For example the *katG*-Ser315Thr mutation confers resistance but minimally affects fitness, hence is highly prevalent among Pre-MDR-TB and MDR-TB strains [@Hazbon2006][@Phelan2019]. In the case of rifampicin (RIF), the target is the $\beta$' subunit of RNA polymerase, coded by *rpoB*. Mutations here prevent RIF from binding, but incur high fitness cost since the intricate machinery of RNA polymerase is intolerant to large structural changes [@Telenti1997], with the exception of Ser450Leu, which is highly prevalent in RIF-resistant strains [@Heep2001]. Indeed, so restrictive are changes to the $\beta$ subunit that over 95% of drug resistance mutations occur in the 'rifampicin resistance determining region' (RRDR), an 81 base-pair section of *rpoB* [@Cao2019].

To overcome this fitness cost, secondary mutations can occur which improve or promote either the target protein itself or in another protein with a similar function. In the case of INH/*katG* the protein AhpC, which has a similar enzymatic function, is often promoted via mutations in the promoter of the *ahpC* gene [@Sherman1999][@Sherman1996]. RIF compensatory mutations occur in the other RNA polymerase subunits $\alpha$ (*rpoA*), $\beta$' (*rpoC*) or even within the $\beta$ subunit itself. These mutations are thought to occur at the interfaces of the subunits, helping to restore overall RNA polymerase function while maintaining an altered binding site in the $\beta$ subunit [@Comas2011]. The TB-profiler platform [@Phelan2019] contains a whole genome sequence (WGS) database of ~32k samples. Each sample is profiled for known drug resistance-associated mutations but also includes any mutation that has not previously been associated with resistance ('non-drug resistance-associated mutation'). By investigating those samples with known (and unknown) compensatory mutations but no known resistance mutations, we reasoned that the presence of these mutations could be leveraged to reveal hitherto non-drug resistance-associated mutations in drug target genes for isoniazid and rifampicin.

## RESULTS

```{r results-isolate-collection, echo = F, warning=F}

# **Isolate collection**
inh_n_samps_TBP_res <- nrow(subset(KRM_table, Drug == "isoniazid" & KRM_binary == "1"))
rif_n_samps_TBP_res <- nrow(subset(KRM_table, Drug == "rifampicin" & KRM_binary == "1"))

n_samps_MDR <- nrow(subset(meta_table, Drug == "isoniazid" & drtype == "MDR-TB"))

```

**Input data**

A total of `r total_samps` *M. tuberculosis* isolates with whole genome sequencing data were suitable for analysis and encompassed all major lineages (**Table \@ref(tab:table-1-display)**). Almost all had undergone drug susceptibility testing, with `r fmt(inh_n_samps_DST_res)` (`r fmt_pc(inh_n_samps_DST_res/inh_n_samps_dst, digits = 2)` and `r fmt(rif_n_samps_DST_res)` (`r fmt_pc(rif_n_samps_DST_res/rif_n_samps_dst, digits = 2)` resistant to isoniazid and rifampicin, respectively. Genotypic resistance prediction using TB-profiler inferred that `r fmt(inh_n_samps_TBP_res)` (`r fmt_pc(inh_n_samps_TBP_res/total_samps, digits = 3)`) and `r fmt(rif_n_samps_TBP_res)` (`r fmt_pc(rif_n_samps_TBP_res/total_samps, digits = 3)`) isolates were resistant to isoniazid and rifampicin, respectively. Further, `r fmt(n_samps_MDR)` (`r fmt_pc(n_samps_MDR/total_samps, digits = 3)`) samples were classified as MDR-TB (see **Table \@ref(tab:table-1-display)**). 

```{r results-compensatory-mutations, echo = F, warning=F}

# **Compensatory mutations**
inh_CM <- select(subset(inh_bin, CM_bin == "present"), CM)

inh_n_CM_haplotypes <- length(unique(inh_CM$CM))
inh_CM_uniq <- uniq_vars(inh_CM$CM)
inh_CM_uniq_in_KCM <- inh_CM_uniq[inh_CM_uniq %in% inh_KCM$gene_change]
inh_PCM <- inh_CM_uniq[!(inh_CM_uniq %in% inh_KCM$gene_change)]

rif_CM <- select(subset(rif_bin, CM_bin == "present"), CM)
rif_n_samps_KRM_CM <- nrow(subset(rif_bin, KRM_bin == "present" & 
                                    CM_bin == "present"))
rif_n_samps_KRM_no_CM <- nrow(subset(rif_bin, KRM_bin == "absent" & 
                                    CM_bin == "present"))
rif_n_CM_haplotypes <- length(unique(rif_CM$CM))
rif_CM_uniq <- uniq_vars(rif_CM$CM)
rif_rpoC <- subset(rif_CM_uniq, grepl("rpoC", rif_CM_uniq))
rif_rpoA <- subset(rif_CM_uniq, grepl("rpoA", rif_CM_uniq))
rif_CM_uniq_in_KCM <- rif_CM_uniq[rif_CM_uniq %in% rif_KCM$gene_change]
rif_PCM <-  rif_CM_uniq[!(rif_CM_uniq %in% rif_KCM$gene_change)]

```

```{r CM-table, echo = F, warning=F}

# Table of compensatory mutations - gene, mutation [n], status 
inh_CM_data <- data.frame(inh_bin %>% mutate(CM = strsplit(as.character(CM), "; ")) %>% unnest(CM))
inh_CM_tab <- sort_pos(pivot_num(subset(inh_CM_data, !is.na(CM)), formula(CM ~ 'n')), "CM")
inh_CM_tab$gene <- get_gene(inh_CM_tab$CM)
inh_CM_tab$mutation <- paste0(clean_del_ins_dup(drop_gene(inh_CM_tab$CM)), " [", inh_CM_tab$n, "]")
inh_CM_tab$status <- ifelse(inh_CM_tab$CM %in% KCM$gene_change, "known", "unknown")
inh_CM_tab$drug <- rep("isoniazid", nrow(inh_CM_tab))
inh_CM_tab <- select(inh_CM_tab, drug, gene, mutation, status)

rif_CM_data <- data.frame(rif_bin %>% mutate(CM = strsplit(as.character(CM), "; ")) %>% unnest(CM))
rif_CM_tab <- pivot_num(subset(rif_CM_data, !is.na(CM)), formula(CM ~ 'n'))
rif_CM_tab$gene <- get_gene(rif_CM_tab$CM)
rif_CM_tab$mutation <- paste0(clean_del_ins_dup(drop_gene(rif_CM_tab$CM)), " [", rif_CM_tab$n, "]")
rif_CM_tab$status <- ifelse(rif_CM_tab$CM %in% KCM$gene_change, "known", "unknown")
rif_CM_tab$drug <- rep("rifampicin", nrow(rif_CM_tab))
rif_CM_tab <- select(rif_CM_tab, drug, gene, mutation, status)

CM_tab <- rbind(inh_CM_tab, rif_CM_tab)
names(CM_tab) <- c("Drug", "Compensatory locus", "Mutation [frequency]", "Status")

```

To characterise potential novel resistance mutations, we looked for samples which had a compensatory mutation, but no known resistance mutation. We used a used a manually curated list of compensatory mutations identified from the literature. For *ahpC*, this list was supplemented with extra mutations which were not previously reported in the literature but showed strong evidence indicating a compensatory mechanism including acquisition through convergent evolution, co-occurrence with frameshift and gene deletion loss of function as well as association with INH resistant isolates. This led to a final list with `r length(inh_CM_uniq)` mutations in *ahpC*, `r length(rif_rpoC)` mutations in *rpoC* and `r length(rif_rpoA)` mutations in *rpoA* (**Table \@ref(tab:CM-table_display)**) 

```{r inh-stats, echo = F, warning = F}

# n katG mutations
inh_n_uniq_KRM_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "KRM_katG"])), 
                                                                    "KRM_katG"]), "; "))))
inh_n_uniq_other_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "other_katG"])), 
                                                                      "other_katG"]), "; "))))
inh_n_uniq_PRM_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "PRM"])), "PRM"]), "; "))))
total_unique_katg <- inh_n_uniq_KRM_katg + inh_n_uniq_other_katg + inh_n_uniq_PRM_katg

```

*Isoniazid*

For INH there were `r inh_3_a_c` samples with a compensatory mutation, of which `r inh_3_c` had no known resistance mutation (**Figure \@ref(fig:samples-trees))**. Within the `r inh_3_c` samples we looked for mutations in *katG* that could potentially explain the emergence of the compensatory mutation. In total, `r total_unique_katg` unique non-synonymous mutations were found in *katG*. To reduce the possibility of identifying non-resistance associated mutations, they were filtered to retain those which occurred in at least three samples, where >50% of the samples had a resistant DST and where >50% of samples had genotypic resistance to at least one other drug. 

```{r PRM-summary-stats, echo = F, warning=F}

# PRM subset and summary
PRM_tabs <- lapply(binary_tables, function(x){subset(x, PRM_bin == "present")})
PRM_pivot <- rm_null(lapply(PRM_tabs, function(x){
  tryCatch({
    sort_pos(pivot_num(x, formula(PRM ~ 'n')), 'PRM')
  }, error = function(e){})
            }))
PRM_pivot <- add_drug_col(PRM_pivot, drugs)
PRM_summary <- do.call("rbind", PRM_pivot)
PRM_summary <- select(PRM_summary, Drug, PRM, n)

# PRM stats
# PRM_stats <- rm_null(lapply_nms(PRM_stats_files, try_read_csv, drugs))
PRM_stats <- lapply(PRM_stats, sort_pos, 'mutation')
# Clean RIF of the lineage-specific PRMs
PRM_stats$rifampicin <- PRM_stats$rifampicin[!(PRM_stats$rifampicin$gene_mutation %in% rif_lin_PRM), ]
PRM_stats <- do.call("rbind", PRM_stats)
# PRM_stats <- select(PRM_stats, drug, gene, mutation, n_samps, dst_prop, dr_prop, n_lins)
PRM_stats$dst_prop <- 1 - PRM_stats$dst_prop
# names(PRM_stats) <- c("Drug", "Gene", "Change", "n samples", "DST sensitive", "DR-type sensitive", "n lineages")

# Add the PRMs which only appear in one lineage to PRM summary
PRMs_one_lin <- select(subset(PRM_stats, n_lins == 1), gene_mutation)[[1]]
PRM_summary$PRM_one_lin <- stringr::str_extract(PRM_summary$PRM, paste0(PRMs_one_lin, collapse = "|"))

# Clean up and format
PRM_summary$PRM <- gsub("; ", "/", gsub("katG-p.|katG-c.", "", PRM_summary$PRM))
PRM_summary$PRM_one_lin <- gsub("katG-p.|katG-c.", "", PRM_summary$PRM_one_lin)
PRM_summary$PRM <- paste0(PRM_summary$PRM, " [", PRM_summary$n, "]")
PRM_summary <- select(PRM_summary, Drug, PRM, PRM_one_lin)
names(PRM_summary) <- c("Drug", "Mutation [frequency]", "PRM_one_lin")

```

Using this approach reduced the number to `r inh_n_uniq_PRM_katg` unique mutations present in `r inh_4_e` samples which we term putative resistance mutations (**Table \@ref(tab:PRM-summary-table)**). 

```{r make-rare-mutations-table, echo = F, warning=F}

rare_mutations_table <- tryCatch({
   cbind(Drug = c(rep("isoniazid", nrow(rare_katgs)), 
                  rep("rifampicin", nrow(rare_rpobs))), 
         rbind(rare_katgs, rare_rpobs))
}, error = function(e){
  cbind(Drug = c(rep("isoniazid", nrow(rare_katgs))), 
        rbind(rare_katgs, rare_rpobs))
})

rare_mutations_table$Change <- gsub("katG-c.|katG-p.", "", rare_mutations_table$Change)

rare_mutations_table <- data.frame(cbind(Drug = unique(rare_mutations_table$Drug), 
             Change = paste(rare_mutations_table$Change, collapse = ", ")))

```

Of the `r inh_3_c` samples which had a CM, `r inh_4_f` did not have a PRM, however upon further inspection, `r inh_5_a` of these samples were found to have a rare *katG* mutation that did not pass the minimum sample cut-off (>=3) in the definition of a putative resistance mutation (**Table \@ref(tab: rare-mutations-table)** ). These rare mutations could also potentially explain the acquisition of a compensatory mutation; however, they were not analysed further due to the low number of samples in which they occur. Although PRMs were found using samples with a CM (n=`r inh_4_e`), additional occurrences were found in `r inh_4_c_g` samples without a CM, leading to a total of  `r inh_4_a_c_e_g` samples which had a putative resistance mutation (**Figure \@ref(fig:samples-trees)**). 

*Rifampicin*

The same approach was used to look for similar mutations that could confer RIF resistance. Using *rpoA* and *rpoC* as compensatory loci and *rpoB* as the resistance associated locus. There were `r ifelse(rif_3_c == 0, "no", rif_3_c)` samples found to have a CM but no known resistance mutation in *rpoB* (**Figure \@ref(fig:samples-trees)**). Within these samples `r ifelse(rif_5_a == 0, "none", rif_5_a)` had an *rpoB* mutation. As such, the rest of the results section details further investigation into INH putative resistance mutations only.

<!-- and `r ifelse(rif_4_e == 0, "none", rif_5_a)` of these mutations met the minimum sample size criteria (>=3) to be characterised as a putative resistance mutation. -->

*Sensitivity analysis*

A small number of rare known resistance mutations (n=5), occurring in just 3-7 samples, were selected for sensitivity analysis to verify that these could be recovered by our methods. Upon re-analysis with the selected mutations taken out of the known resistance list, all mutations were subsequently classified as putative resistance mutations, highlighting the power of this approach. 

*Resistance and co-occurrence with other resistance mutations*

The mutations occurred in multiple lineages (L1-L5) with many showing evidence of convergent evolution (**Figure \@ref(fig:isoniazid-tree)**). Due to the multi-drug regimens used for TB treatment, resistance often develops to multiple drugs in a stepwise manner. Co-occurrence with other resistance mutations was analysed to characterise the sample profile in which PRMs occur. 


```{r PRM-dr-types, echo = F, warning=F}

PRM_DR_pivot <- pivot_num(PRM_tabs$isoniazid, formula(drtype ~ 'n'))
PRM_DR_max <- PRM_DR_pivot[which.max(PRM_DR_pivot$n), "drtype"]
PRM_DR_max_pc <- fmt_pc(PRM_DR_pivot[which.max(PRM_DR_pivot$n), "n"]/inh_4_a_c_e_g)
PRM_DR_pre_MDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Pre-MDR-TB", "n"]/inh_4_a_c_e_g)
PRM_DR_pre_XDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Pre-XDR-TB", "n"]/inh_4_a_c_e_g)
PRM_DR_XDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "XDR-TB", "n"]/inh_4_a_c_e_g, digits = 2)
PRM_DR_sens_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Sensitive", "n"]/inh_4_a_c_e_g, digits = 2)
PRM_DR_other_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Other", "n"]/inh_4_a_c_e_g, digits = 2)

```

The mutations were most frequently found in samples characterised as `r PRM_DR_max` (`r PRM_DR_max_pc`) with representation in Pre-MDR-TB (`r PRM_DR_pre_MDR_pc`), Pre-XDR-TB (`r `PRM_DR_pre_XDR_pc`), XDR-TB (`r `PRM_DR_XDR_pc`), Sensitive (`r PRM_DR_sens_pc`) and Other (`r PRM_DR_other_pc`). 

```{r PRM-KRM-other-co-occurrence, echo = F, warning=F}

PRM_table <- rm_norow(lapply(binary_tables, function(x){subset(x, !(is.na(PRM)))}))

# Pivot table of PRM-KRM only
PRM_KRM <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + KRM ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_KRM <- do.call("rbind", add_drug_col(PRM_KRM, drugs))
inh_PRM_KRM <- subset(PRM_KRM, Drug == "isoniazid" & !(is.na(KRM)))
inh_n_PRM_hap_KRM <- length(unique(inh_PRM_KRM$PRM))
inh_most_common_KRM_with_PRM <- names(which.max(table(inh_PRM_KRM$KRM)))

# PRM-other only
PRM_other <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + other_vars ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_other <- do.call("rbind", add_drug_col(PRM_other, drugs))
inh_PRM_other <- subset(PRM_other, Drug == "isoniazid" & !(is.na(other_vars)))
inh_n_PRM_hap_other <- length(unique(inh_PRM_other$PRM))
inh_most_common_other_with_PRM <- names(which.max(table(inh_PRM_other$other_vars)))

# PIvot table with 'other' - displayed later as PRM-KRM-other-table
PRM_KRM_other <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + KRM + other_vars ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_KRM_other <- do.call("rbind", add_drug_col(PRM_KRM_other, drugs))
PRM_KRM_other <- dplyr::rename(PRM_KRM_other, unknown = other_vars)


PRM_pc_samps_inh_co_occur <- fmt_pc(sum(subset(PRM_KRM_other, !(is.na(KRM)))$n)/inh_4_a_c_e_g)
PRM_pc_fabg_15 <- fmt_pc(sum(subset(PRM_KRM_other, KRM == "fabG1-c.-15C>T")$n)/inh_4_a_c_e_g)


```

Interestingly, around half (`r PRM_pc_samps_inh_co_occur`) of samples co-occurred with other isoniazid mutations, with the *fabG1* -15C>T promoter mutation being the most frequent (`r PRM_pc_fabg_15`) (**Table \@ref(tab:PRM-KRM-other-table)**). 

```{r isoniazid-Ser315Thr, echo = F, warning=F}

ser315thr_tab <- subset(isoniazid_binary_table, grepl("Ser315Thr", isoniazid_binary_table$KRM))
# Move ser315thr string to front e.g. fabG1-c.-15C>T; katG-p.Ser315Thr -> katG-p.Ser315Thr; fabG1-c.-15C>T
ser315thr_tab$KRM <- ifelse(grepl("; katG-p.Ser315Thr", ser315thr_tab$KRM), swap_str(ser315thr_tab$KRM), ser315thr_tab$KRM)
# Drop the specific change for the co-occurring gene - e.g. katG-p.Ser315Thr; fabG1-c.-15C>T ->  katG-p.Ser315Thr; fabG1
ser315thr_tab$KRM_2 <- drop_change(ser315thr_tab$KRM)
ser315thr_pivot <- pivot(ser315thr_tab, formula(KRM_2 ~ 'n'))
ser315thr_pivot <- dplyr::rename(ser315thr_pivot, mutation = KRM_2)

# ser315thr_pivot[grepl("; katG", ser315thr_pivot$mutation)]
 
# PRM_KRM_other[grepl("Ser315Thr", PRM_KRM_other$KRM), ]

```

```{r make-resistance-table, echo = F, warning=F}

# Resistance level

# Multiple KRMs	kat315	PRMs
# Yes	          5	      20
# No	          100	    3

# Get all samples with a Ser315Thr regardless of other KRMs being present (use grepl), with no PRMs
inh_kat315_multi_KRM_no_PRM <- subset(inh_bin, grepl("katG-p.Ser315Thr", KRM) & PRM_bin == "absent")
# Make a binary column - are there other KRMs or not? ("; " in KRM col)
inh_kat315_multi_KRM_no_PRM$Ser315Thr_bin <- ifelse(grepl("; ", inh_kat315_multi_KRM_no_PRM$KRM), "present", "absent")
# Pivot - other KRMs in Ser315Thr vs presence of Ser315Thr
inh_kat315_multi_KRM_no_PRM_pivot <- pivot(inh_kat315_multi_KRM_no_PRM, formula(Ser315Thr_bin ~ KRM_bin))

# Pivot - KRMs occurring with PRM (but not katG)
inh_4_e_g_table <- subset(inh_bin, PRM_bin == "present" & KRM_katG_bin == "absent")
inh_4_e_g_pivot <- pivot(inh_4_e_g_table, formula(KRM_bin ~ PRM_bin))

resistance_table <- merge(inh_kat315_multi_KRM_no_PRM_pivot, inh_4_e_g_pivot, 
                          by.x = "Ser315Thr_bin", by.y = "KRM_bin", sort = F)

names(resistance_table) <- c("KRM", "Ser315Thr", "PRM")

# Pull out % of Ser315Thr co-occurrences with KRMs
pc_315_KRM_co_occur <- str_extract(select(subset(resistance_table, KRM == "present"), Ser315Thr), "\\(.*\\)")

# Do Chi sq
tab <- cbind(table(inh_kat315_multi_KRM_no_PRM$Ser315Thr_bin, inh_kat315_multi_KRM_no_PRM$KRM_bin), table(inh_4_e_g_table$KRM_bin, inh_4_e_g_table$PRM_bin))
resistance_x_sq <- print_chisq(chisq.test(tab), rnd = 2)

```

This is in stark contrast to the Ser315Thr mutation, the most prevalent resistance mutation in INH resistant isolates and which is known to confer a high level of resistance, which only co-occurs with other INH resistance mutations in (`r pc_315_KRM_co_occur`) of samples. 

```{r PRM-DST, echo = F, warning=F}

PRM_DST <- subset(PRM_tabs$isoniazid, !(is.na(dst)))
inh_n_samps_PRM_DST <- nrow(PRM_DST)
inh_n_samps_PRM_DST_1 <- nrow(subset(PRM_DST, dst == 1))
PRM_DST_p <- scientific(chisq.test(table(PRM_DST$dst))$p.value)
PRM_DST_1_no_KRM <- nrow(subset(PRM_DST, dst == 1 & is.na(KRM)))

```

Of the `r inh_4_a_c_e_g` samples with a PRM, `r inh_n_samps_PRM_DST` had an available DST result for isoniazid with `r inh_n_samps_PRM_DST_1` reporting a resistant phenotype leading to a highly significant association of PRMs and DST (p<`r PRM_DST_p`, Table S2). Of those with a resistant DST, `r PRM_DST_1_no_KRM` had no other known mutations that could explain resistance. 

*Mutation fitness*

```{r make-fitness-table, echo = F, warning=F}

# Fitness
# Ser315Thr is low fitness cost so should not have many CMs

# Compensatory	kat315	PRMs
# Yes	          5	      20
# No	          100	    3
# (exclude PRM-katG - katG-p.Ser315Thr pairs)

inh_kat315_no_PRM <- subset(inh_bin, KRM == "katG-p.Ser315Thr" & PRM_bin == "absent")
inh_kat315_no_PRM_pivot <- pivot(inh_kat315_no_PRM, formula(CM_bin ~ KRM))

inh_PRM_no_katG_KRM_pivot <- pivot(inh_4_e_g_table, formula(CM_bin ~ PRM_bin))

fitness_table <- merge(inh_kat315_no_PRM_pivot, inh_PRM_no_katG_KRM_pivot, 
                             by = "CM_bin", sort = F)

names(fitness_table) <- c("CM", "katG-p.Ser315Thr", "PRM")

# Do chi-sq test
x <- table(inh_kat315_no_PRM$CM_bin, inh_kat315_no_PRM$KRM)
y <- table(inh_4_e_g_table$CM_bin, inh_4_e_g_table$PRM_bin)
tab <- cbind(x, y)
fitness_x_sq <- print_chisq(chisq.test(tab), rnd = 2)

inh_n_samps_ser315_no_CM <- nrow(subset(inh_kat315_no_PRM, CM_bin == "absent"))
inh_n_samps_ser315_and_CM <- nrow(subset(inh_kat315_no_PRM, CM_bin == "present"))

```

```{r make-fitness-plot, echo = F, warning=F}

# KRM
# KRM <- subset(inh_bin, KRM_bin == "present")
# KRM <- data.frame(KRM %>% mutate(KRM = strsplit(as.character(KRM), "; ")) %>% unnest(KRM))
# KRM_tab <- pivot_num(KRM, formula(KRM ~ CM_bin))
# KRM_tab$KRM <- clean_del_ins_dup(KRM_tab$KRM)

# Just take Ser315Thr for now - all samples with Ser315Thr (but no PRM)
# KRM_tab <- subset(KRM_tab, grepl("Ser315Thr", KRM_tab$KRM))
# KRM_tab <- pivot_num(inh_kat315_no_PRM, formula(KRM ~ CM_bin))

KRM <- subset(inh_bin, KRM_bin == "present" & PRM_bin == "absent")
KRM <- data.frame(KRM %>% mutate(KRM = strsplit(as.character(KRM), "; ")) %>% unnest(KRM))
KRM_tab <- pivot_num(KRM, formula(KRM ~ CM_bin))
KRM_tab$KRM <- clean_del_ins_dup(KRM_tab$KRM)
KRM_tab <- subset(KRM_tab, grepl("katG", KRM_tab$KRM))
KRM_tab$status <- rep('known', nrow(KRM_tab))

# PRM
# PRM <- subset(inh_bin, PRM_bin == "present")
# Split out PRM haplotypes (one row per PRM, duplicating the ID and everything else in the row)
# PRM <- data.frame(PRM %>% mutate(PRM = strsplit(as.character(PRM), "; ")) %>% unnest(PRM))
# PRM_tab <- pivot_num(PRM, formula(PRM ~ CM_bin))

# Split out PRM haplotypes (one row per PRM, duplicating the ID and everything else in the row)
# Use PRM with no known katG
PRM <- data.frame(inh_PRM_no_katG_KRM %>% 
                    mutate(PRM = strsplit(as.character(PRM), "; ")) 
                  %>% unnest(PRM))
PRM_tab <- pivot_num(PRM, formula(PRM ~ CM_bin))
# Clean up
PRM_tab$PRM <- clean_del_ins_dup(PRM_tab$PRM)
PRM_tab <- sort_pos(PRM_tab, "PRM")
PRM_tab$status <- rep('unknown', nrow(PRM_tab))
# Put together
KRM_PRM_tab <- rbind_mod(KRM_tab, PRM_tab, c("change", drop_col_name(KRM_tab, "KRM")))
KRM_PRM_tab$pc <- (KRM_PRM_tab$present / (KRM_PRM_tab$absent + KRM_PRM_tab$present) )*100

PRM_med <- median(subset(KRM_PRM_tab, status == 'unknown')$pc)

# Plot each PRM % of samples with a CM with only 
fitness_plot <- ggplot() + geom_point(data = subset(KRM_PRM_tab, status == "unknown" | change == "katG-p.Ser315Thr"), 
                                      aes(x = factor(change, level = change), y = pc, 
                                          colour = status), size = 3)
fitness_plot <- fitness_plot + scale_x_discrete(name="") + 
  scale_y_continuous(name="% of samples with CM")
fitness_plot <- fitness_plot + theme_bw()
fitness_plot <- fitness_plot + theme(axis.text.x = element_text(angle=45, hjust = 1))
fitness_plot <- fitness_plot + theme(aspect.ratio=50/100)
fitness_plot <- fitness_plot + geom_hline(yintercept=PRM_med, linetype="dashed", color = "red")

# Plot number of samples vs % of samples with a CM
fitness_pc_total_plot <- ggplot() + geom_point(data = KRM_PRM_tab, 
                                      aes(x = log(absent + present), y = pc, 
                                          colour = status), size = 3)
fitness_pc_total_plot <- fitness_pc_total_plot + scale_x_continuous(name="log(sum) of samples") + scale_y_continuous(name="% samples with CM")
fitness_pc_total_plot <- fitness_pc_total_plot + theme_bw()
fitness_pc_total_plot <- fitness_pc_total_plot + theme(aspect.ratio=50/100)
fitness_pc_total_plot <- fitness_pc_total_plot + geom_text(data=subset(KRM_PRM_tab, change == "katG-p.Ser315Thr"),
                                                           aes(x = log(absent + present), y = pc, label = change), 
                                                           hjust = 1.1)

```

Compensatory mutations are known to occur when mutations with high fitness-costs occur (e.g. *katG* loss of function). To estimate the fitness impact of the PRMs we used the co-occurrence with a CM as a proxy, comparing to Ser315Thr which has reported to confer a low fitness cost. 

<!-- CHECK! --> 
A Chi-square test showed a significant difference (`r fitness_x_sq`), with compensatory mutations occurring in a higher proportion in samples with PRMs (`r inh_4_e`/`r nrow(inh_4_e_g_table)`) 
<!-- CHECK! --> 

<!-- CHECK! --> 
than those with Ser315Thr (`r inh_n_samps_ser315_and_CM`/`r nrow(inh_kat315_no_PRM)`), suggesting that on average they incur a greater fitness cost. 
<!-- CHECK! --> 









```{r results-resistance-mutations, echo = F, warning=F}

# **Resistance mutations, and how many explain the phenotype**
inh_n_samps_KRM <- nrow(subset(inh_bin, KRM_bin == "present"))
rif_n_samps_KRM <- nrow(subset(rif_bin, KRM_bin == "present"))

inh_n_samps_KRM_pc <- fmt_pc(nrow(subset(inh_bin, KRM_bin == "present"))/total_samps)
rif_n_samps_KRM_pc <- fmt_pc(nrow(subset(rif_bin, KRM_bin == "present"))/total_samps)

inh_n_samps_DST_res_KRM_present <- nrow(subset(inh_bin, dst == "1" & KRM_bin == "present"))
inh_n_samps_DST_res_KRM_present_pc <- fmt_pc(inh_n_samps_DST_res_KRM_present/tonum(inh_n_samps_DST_res))

rif_n_samps_DST_res_KRM_present <- nrow(subset(rif_bin, dst == "1" & KRM_bin == "present"))
rif_n_samps_DST_res_KRM_present_pc <- fmt_pc(rif_n_samps_DST_res_KRM_present/tonum(rif_n_samps_DST_res))

```

**Resistance mutations, and how many explain the phenotype**

*Summary* \
Number INH samples DST resistant and have a known resistance mutation - `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) \
Number RIF samples DST resistant and have a known resistance mutation - `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) \

The total number of samples with a known resistance mutation in an INH target gene (*katG*, *fabG1*, *inhA*) was `r inh_n_samps_KRM` (`r inh_n_samps_KRM_pc`), and `r rif_n_samps_KRM` (`r rif_n_samps_KRM_pc`) for the RIF target gene (*rpoB*).

Explaining the DST resistance status, `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) samples had a known resistance mutation in an INH target gene, while `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) samples had an *rpoB* mutation (RIF).

**Compensatory mutations**

*summary* \
Number INH samples with a CM - `r inh_3_a_c` \
Number INH CM haplotypes - `r inh_n_CM_haplotypes` \
Number INH unique CM - `r length(inh_CM_uniq)` \
Number INH CM in known CM list - `r length(inh_CM_uniq_in_KCM)` \
Number INH CM previously non-DR associated - `r length(inh_PCM)` \

Number RIF samples with a CM - `r rif_3_a_c` \
Number RIF CM haplotypes - `r rif_n_CM_haplotypes` \
Number RIF unique CM - `r length(rif_CM_uniq)` \
Number RIF CM in known CM list - `r length(rif_CM_uniq_in_KCM)` \
Number RIF CM previously non-DR associated - `r length(rif_PCM)` \

There were `r fmt(inh_3_a_c)` and `r fmt(rif_3_a_c)` samples with a compensatory mutation, of which `r fmt(inh_3_a)` and `r fmt(rif_3_a)` samples had a known resistance mutation, and `r fmt(inh_3_c)` and `r fmt(rif_3_c)` did not have known resistance mutations in INH and RIF, respectively. 

The number of compensatory mutation haplotypes were `r inh_n_CM_haplotypes` (INH) and `r rif_n_CM_haplotypes` (RIF), while the number of unique compensatory mutations were `r length(inh_CM_uniq)` (INH) and `r length(rif_CM_uniq)` (RIF).

Of these, `r length(inh_CM_uniq_in_KCM)` (INH) and `r length(rif_CM_uniq_in_KCM)` (RIF) were from compensatory mutations previously known, whereas `r length(inh_PCM)` INH and `r length(rif_PCM)` RIF compensatory mutations were putative. 

```{r results-PRMs, echo = F, warning=F}

# **Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers**
inh_PRM <- select(subset(inh_bin, PRM_bin == "present"), PRM)
inh_n_PRM_haplotypes <- length(unique(inh_PRM$PRM))
inh_PRM_uniq <- unique(unlist(strsplit(inh_PRM$PRM, "; ")))

rif_PRM <- select(subset(rif_bin, PRM_bin == "present"), PRM)
rif_n_PRM_haplotypes <- length(unique(rif_PRM$PRM))
rif_PRM_uniq <- unique(unlist(strsplit(rif_PRM$PRM, "; ")))

```

**Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers**

*Summary* \
Number INH samples with a PRM - `r nrow(inh_PRM)` \
Number INH PRM haplotypes - `r inh_n_PRM_haplotypes` \
Number INH unique PRM - `r length(inh_PRM_uniq)` \

Number RIF samples with a PRM - `r nrow(rif_PRM)` \
Number RIF PRM haplotypes - `r rif_n_PRM_haplotypes` \
Number RIF unique PRM - `r length(rif_PRM_uniq)` \

<!-- LAYER 4 -->
Number INH samples with KRM (*katG*) and CM and PRM - `r fmt(length(inh_4_a))`\
Number INH samples with KRM (*katG*), no CM and PRM - `r fmt(length(inh_4_c))`\
Number INH samples with no KRM (*katG*) and CM and PRM - `r fmt(length(inh_4_e))`\
Number INH samples with no KRM (*katG*), no CM and PRM - `r fmt(length(inh_4_g))`\

<!-- LAYER 5 --> 
Number INH samples with no KRM (*katG*), CM, no PRM, other *katG* - `r n_samps_rare_katg`\
Number INH unique rare *katG* mutations - `r n_rare_katg`\

Number RIF samples with no KRM (*ropB*), CM, no PRM, other *rpoB* - `r n_samps_rare_rpob`\
Number INH unique rare *rpoB* mutations - `r n_rare_rpob`\

After applying filters to the samples with compensatory mutations to find potential resistance mutations, and then re-searching the full TB-profiler database for samples with said mutations, `r nrow(inh_PRM)` INH isolates were found. In these INH samples, there were `r inh_n_PRM_haplotypes`, with `r length(inh_PRM_uniq)` unique mutations (see **Table \@ref(tab:PRM-summary-table) and Table \@ref(tab:PRM-stats-table)**). No samples were found with potential resistance mutations in *rpoB* after phylogenetic analysis revealed a small number of potential resistance mutations were clearly lineage-specific (`r print_vect(rif_lin_PRM)` and therefore unlikely to be a resistance mutation. 

Within the INH samples with putative resistance mutations, `r fmt(length(inh_4_c))` samples had a known *katG* mutation but no compensatory mutations (see **Table \@ref(tab:PRM-KRM-other-table)**), `r fmt(length(inh_4_e))` samples had no known *katG* mutation, but did have a compensatory mutation, while the majority, `r fmt(length(inh_4_g))` had no known *katG* mutation and no compensatory mutations (see **Figure \@ref(fig:samples-breakdown)**).

Further to the potential resistance mutations, there were also found `r n_samps_rare_katg` (INH) and `r n_samps_rare_rpob` (RIF) samples with no known *katG*/*rpoB* resistance mutations, having a compensatory mutation, but not having a potential resistance mutation by the criteria outlined in *Methods*, nevertheless with a non-drug-resistance-associated *katG*/*rpoB* mutation ('rare' mutation). These rare mutations numbered `r n_rare_katg` in *katG* and `r n_rare_rpob` for *rpoB* (see **Table \@ref(tab:rare-mutations-table)**).

A small number of rare known resistance mutations were selected for sensitivity analysis to verify that these could be recovered by our methods. These mutations occurred in just 7, 6, 5, 4 or 3 samples. When the analysis was re-run with the mutations uncategorised, all subsequently appeared as 'potential resistance mutations'. 


**katG potential resistance mutations - co-occurrence with other resistance mutations and comparison to Ser315Thr**

*Summary* \
Number of *katG* PRM haplotypes occurring with a KRM - `r inh_n_PRM_hap_KRM` (`r fmt_pc(inh_n_PRM_hap_KRM/inh_n_PRM_haplotypes)`, `r sum(inh_PRM_KRM$n)` samples) \
Most common KRM occurring with *katG* PRM - `r inh_most_common_KRM_with_PRM`

Number of *katG* PRM haplotypes occurring with an unknown mutation -  `r inh_n_PRM_hap_other` (`r fmt_pc(inh_n_PRM_hap_other/inh_n_PRM_haplotypes)`, `r sum(inh_PRM_other$n)` samples) \
Most common unknown occurring with *katG* PRM - `r inh_most_common_other_with_PRM`


Mutation *katG*-Ser315Thr incurs little fitness cost and so ought not to co-occur with *ahpC* compensatory mutations. To compare the fitness cost of *katG*-Ser315Thr and our potential resistance mutations, **Table \@ref(tab:fitness-table)** shows comparison of presence/absence of compensatory mutations in samples with (only) *katG*-Ser315Thr and samples with the potential resistance mutations (samples with potential resistance mutations and co-occurring known resistance mutations in *katG* were excluded). A Chi-square test showed a significant difference (`r fitness_x_sq`), with compensatory mutations occurring in a higher proportion in the potential resistance mutations than *katG*-p.Ser315Thr, indicating on average they incur a greater fitness cost. 

*katG*-Ser315Thr is also useful in assessing the resistance level of the potential resistance mutations. Since *katG*-Ser315Thr confers resistance to INH in an all-or-none way, a minority of *katG*-Ser315Thr samples should have co-occurrences with other known resistance mutations, and this is indeed the case - see **Table \@ref(tab:resistance-table)**. Nearly half of the samples with a potential resistance mutation showed co-occurrence with a known resistance mutation, again excluding those with co-occurrence in *katG*, suggesting the potential resistance mutations on average confer a lesser resistance level than *katG*-Ser315Thr. This difference was also significant (`r resistance_x_sq`). 

**Table \@ref(tab:PRM-KRM-other-table)** shows the full list of co-occurrences of the potential resistance mutation haplotypes and co-occurrence of known resistance mutations and unknown mutations in relevant resistance genes (i.e. co-occurrence of *katG* potential resistance mutations with known/unknown *ahpC*, *inhA*, *fabG1* and *kasA*). There were `r inh_n_PRM_hap_KRM` (`r fmt_pc(inh_n_PRM_hap_KRM/inh_n_PRM_haplotypes)`) *katG* potential resistance mutation haplotypes occurring with a known resistance mutation (`r sum(inh_PRM_KRM$n)` samples), the most common of which was `r inh_most_common_KRM_with_PRM`.

## Discussion

That no RIF *rpoB* potential resistance mutations were found is unsurprising, since there are few ways in which the precise machinery of RNA polymerase can change without loss of function. In contrast, *katG* can change in many ways while the bacteria is largely unaffected.

The ‘rare’ mutations could be considered potential resistance mutations with a larger sample size. These mutations were discarded because they only appeared in one or two samples.

Further research could explore *in silico* prediction of resistance mechanisms in the relevant target proteins, i.e. modelling of how the mutations manifest in the final protein and how drug binding is thereby affected. 

## Conclusions

We have presented a method of quickly finding potential resistance mutations in order to monitor development of resistance mechanisms to two of the most important anti-TB drugs. 

Our list of potential resistance mutations ought to be verified as causing resistance and added to the database of known resistance mutations. 


## Methods

*Sequence data and processing*

Samples were obtained from multiple sources, isolates belonging to individual patients. Quality inspection of sequence reads was condcucted in fastQC (v0.11.9) (www.bioinformatics.babraham.ac.uk/projects/fastqc/). Reads were trimmed to remove low-quality sequences in Trimmomatic (v0.39)	- parameters: *PE, -phred33, LEADING:3, TRAILING:3, SLIDINGWINDOW:4:20, MINLEN:36* [@Bolger2014] and sequences were aligned to the H37Rv reference genome (AL123456) with BWA mem (v0.7.17) [@Li2013], [@Cole1998]. Joint SNP calling was carried out in gatk GenotypeGVCFs (v4.1.3.0) - --java-options "-Xmx40g" - [@Depristo2011].

SNP were filtered to exclude indels and heterozygous SNPs. Monomorphic SNPs and those in non-unique regions of the genome (e.g. ppe genes) were excluded. These procedures were conducted using bcftools view; bcftools filter (v1.9) (using htslib 1.9); parameters *-V indels -e 'GT="het"' -S . -a -c 1 -Oz* [@Li2009]. Multiple variant calling was done with gatk GenotypeGVCFs (v4.1.3.0) (gatk.broadinstitute.org). 

Multi-FASTA formated filed were created from the filtered SNP file and H37Rv reference fasta with bedtools makewindows (v2.28.0); parameters -w 50000 -s 50000 [@Quinlan2010]. 

Phylogenetic trees were made with IQ-TREE (v1.6.12), involving a general time reversible model with rate heterogeneity set to a discrete Gamma model and an ascertainment bias correction (parameters −m GTR+G+ASC), with 1000 bootstrap samples [@Nguyen2015].

Drug resistance types and lineages were predicted *in silico* with TB-Profiler (v2.4) [@Phelan2019] [@Coll2015]. TB-Profiler also used identified all known drug-resistance mutations and all other mutations in relevant genes for finding compensatory mutations and putative novel resistance mutations. 

*Finding putative resistance markers using compensatory mutations*

From the database of TB-profiler - all non-synonymous mutations in the relevant compensatory genes were found (*ahpC* in the case of INH and *rpoC* and *rpoA* for RIF).

All non-drug resistance associated ('unknown') mutations in compensatory genes were filtered such that they were present in three or more samples, <50% samples were predicted *Sensitive* by TB-profiler and <50% of samples were DST sensitive, for each relevant drug. 

As there were many potential *ahpC* mutations, further filtering criteria were applied to these - mutations were retained if they were associated with loss of function in *katG* mutations, occurred in the same position as known *ahpC* mutations and if they appeared in multiple lineages. Only one of these criteria need be met for retention in the final list of potential compensatory *ahpC* mutations. 

These successful mutations were then combined with the list of known compensatory mutations [@Phelan2019] to form a full list of compensatory mutations.

To find putative resistance mutations, all non-synonymous mutations in the relevant resistance genes were pulled from the TB-profiler database (*katG* for INH and *rpoB* for RIF), with the exclusion of isolates with mixed lineages. Some variants were excluded from consideration for PRM. *katG*-Arg463Leu is known not to be associated with INH resistance [@Doorn2001]. 

For each drug, mutations were found in samples where a compensatory mutation was present and no known resistance mutations were present in the relevant genes, but a non-resistance-associated mutation was present in the relevant target genes. 

These mutations were then filtered to exclude known drug-resistance-associated variants, and subjected to the same criteria as the putative compensatory mutations - i.e. present in three or more samples, <50% samples were predicted *Sensitive* by TB-profiler and <50% of samples were DST sensitive. Mutations occurring in promoter regions of *rpoB* were excluded as candidate potential resistance mutations as there is no known mechanism of resistance that could result from increased expression of the RNA polymerase beta subunit.

Using this list of potential resistance mutations, all TB-profiler samples were then searched for their presence, regardless of compensatory mutation status. It should be noted that therefore not all samples with potential resistance mutations necessarily have compensatory mutations, and vice versa. 

## List of abbreviations

* *Mtb* = *Mycobacterium tuberculosis*
* TB = tuberculosis
<!-- * RM = any resistance mutation; -->
<!-- * KRM = known RM; -->
<!-- * PRM = potential RM; -->
<!-- * CM = any compensatory mutation; -->
<!-- * KCM = known CM; -->
<!-- * PCM = potential CM -->
* INH = isoniazid
* RIF = rifampicin
* WGS = whole genome sequencing

## Declarations

## Ethics approval and consent to participate
Not applicable

## Consent for publication
Not applicable 

## Competing interests
The authors declare that they have no competing interests

## Funding
GN is funded by an BBSRC-LIDo PhD studentship. JEP is funded by a Newton Institutional Links Grant (British Council, no. 261868591). TGC is funded by the Medical Research Council UK (Grant no. MR/M01360X/1, MR/N010469/1, MR/R025576/1, and MR/R020973/1). SC is funded by Medical Research Council UK grants (ref. MR/M01360X/1, MR/R025576/1, and MR/R020973/1). The authors declare no conflicts of interest. The funders had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript should be declared.

## Authors' contributions
JEP and TGC conceived and directed the project. GN performed bioinformatic and statistical analyses under the supervision of SC, JEP and TGC. GN, SC, JEP and TGC interpreted results. GN wrote the first draft of the manuscript with inputs from JEP and TGC. All authors commented and edited on various versions of the draft manuscript and approved the final version. GN, SC, JEP, and TGC compiled the final manuscript.

## Acknowledgements

## Tables

```{r table-1-display, echo = F, warning=F}

cap <- sprintf("*Mycobacterium tuberculosis* samples (n=%s)", total_samps)
knitr::kable(table_1, caption = cap, format = "pipe", row.names = F)

```

```{r PRM-summary-table, echo = F, warning=F}

cap <- sprintf("List of potential resistance mutations for isoniazid in *katG* (n=%s)", inh_4_a_c_e_g)
knitr::kable(PRM_summary, row.names = F, caption = cap, format = "pipe")

```

<!-- ```{r PRM-stats-table, echo = F, warning=F} -->

<!-- cap <- "Number of samples, DST sensitive proportion, (predicted) DR type proportion and number of lineages for each PRM" -->
<!-- knitr::kable(PRM_stats, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->


<!-- ```{r rare-counts-table, echo = F, warning=F} -->

<!-- cap <- "Counts of occurrence for *katG* (isoniazid) and *rpoB* (rifampicin) among samples with no KRM, at least one CM, but no PRM. These variants are not counted as ‘PRM’ because of their low frequency (only one or two samples)." -->

<!-- # Variants occurring more than twice were rejected because most samples were DR-type sensitive or DST sensitive, or in the case of INH, the samples already have a known resistance mutation in a gene other than *katG* -->

<!-- knitr::kable(rare_cnts, row.names = F, caption = cap, format = "pipe") -->

<!-- ``` -->

```{r rare-mutations-table, echo = F, warning=F}

cap <- sprintf("Mutations in *katG* (INH) and from samples with no known resistance mutations, with compensatory mutations but no potential resistance mutation by the criteria outlined in Methods. These mutations only occur in one or two samples among the %s samples from Supplementary figure 1", inh_5_a)
knitr::kable(rare_mutations_table, row.names = F, caption = cap, format = "pipe")

```

```{r PRM-KRM-other-table, echo = F, warning=F}

cap <- "Known resistance mutations and non-DR-associated ('unknown') mutations co-occurring in samples with putative drug resistance"
knitr::kable(PRM_KRM_other, row.names = F, caption = cap, format = "pipe")

```

**Isoniazid**

```{r fitness-table, echo = F, warning=F}

cap <- "Sample counts comparing presence and absence of compensatory mutations in *katG*-Ser315Thr and potential resistance mutations. Samples with known *katG* resistance mutations co-occurring with potential resistance mutations were excluded from counts."
knitr::kable(fitness_table, row.names = F, caption = cap, format = "pipe")

```

```{r fitness-plot, echo = F, warning=F, fig.cap="Percentage of samples with compensatory mutations among the potential resistance mutations. *katG*-Ser315Thr is shown for comparison. This mutation incurs low fitness cost, hence does not generally co-occur in samples with a compensatory mutation. Red line - median percentage of samples with compensatory mutations in samples with potential resistance mutations. Samples with *katG*-Ser315Thr and a potential resistance mutation have been excluded. Samples with a potential resistance mutation and a known *katG* resistance mutation have been excluded."}

fitness_plot

```

```{r fitness-pc-plot, echo = F, warning=F, fig.cap="Number of samples vs percentage of samples with a compensatory mutation for known *katG* mutations and potential *katG* resistance mutations. *katG*-Ser315Thr is labeled for comparison. This mutation incurs low fitness cost, hence does not generally co-occur in samples with a compensatory mutation. Samples with *katG*-Ser315Thr and a potential resistance mutation have been excluded. Samples with a potential resistance mutation and a known *katG* resistance mutation have been excluded."}

fitness_pc_total_plot

```

```{r resistance-table, echo = F, warning=F}

cap <- "Samples counts of presence/absence of known resistance mutations co-occurring with *katG*-Ser315Thr and the potential resistance mutations. Samples with known *katG* resistance mutations co-occurring with potential resistance mutations were excluded from counts."
knitr::kable(resistance_table, row.names = F, caption = cap, format = "pipe")

```

```{r isoniazid-Ser315Thr-table, echo = F, warning=F}

cap <- "Comparison to katG-Ser315Thr and katG-Ser315Thr-KRM in other isoniazid resistance genes."
knitr::kable(ser315thr_pivot, row.names = F, caption = cap, format = "pipe")

```

```{r CM-table-display, echo = F, warning=F}

cap <- "Mutations in known compensatory genes and frequencies in 32k samples"
knitr::kable(CM_tab)

```

## Figures

```{r isoniazid-tree, echo=FALSE, out.width = "2000px", fig.cap="Phylogenetic tree of isolates with resistance; Isoniazid resistance (n=XX), with potentially new katG mutations"}

knitr::include_graphics(tree_files[grep("isoniazid", tree_files)])

```

```{r samples-trees, echo=FALSE, out.width = "2000px", fig.show='hold', fig.cap="Analysis strategy and numbers of mutations" }

knitr::include_graphics("../results/samples_trees.png")
# knitr::include_graphics(rif_numbers_file)

```


<!-- ```{r samples-breakdown-rif, echo=FALSE, out.width = "2000px", fig.cap="Breakdown of sample numbers by presence of mutation type for INH" } -->

<!-- knitr::include_graphics(rif_numbers_file) -->

<!-- ``` -->

```{r inh-ahpc-katg-links, echo=FALSE, warning = F, out.width = "2000px", fig.cap="Co-occurrences between *ahpC* compensatory mutation positions and *katG* positions. Gaps in *katG* positions indicate no compensatory mutation was found for that sample." }

binary_expand_PRM <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(PRM = strsplit(as.character(PRM), "; ")) %>%
  unnest(PRM))
})

inh <- binary_expand_PRM$isoniazid
inh_PRM <- subset(inh, !(is.na(PRM)))
inh_PRM <- select(inh_PRM, CM, PRM)
# Split and process del/dup/ins separately 
del_dup_ins <- inh_PRM[grepl("del|dup|ins", inh_PRM$PRM), ]
inh_PRM <- inh_PRM[!(grepl("del|dup|ins", inh_PRM$PRM)), ]
inh_PRM_pos <- data.frame(apply(inh_PRM, 2, find_pos))
del_dup_ins$CM <- find_pos(del_dup_ins$CM)
del_dup_ins$PRM <- gsub("[^0-9_del|dup|ins]", "", del_dup_ins$PRM)
inh_PRM_pos <- rbind(inh_PRM_pos, del_dup_ins)

colnames(inh_PRM_pos) <- c("ahpC", "katG")

df <- inh_PRM_pos %>% make_long(ahpC, katG)
# Remove NAs from first node col - removes ahpCs that do not linek to a katG
df <- subset(df, !(is.na(node)))

# Move the node col to two separate cols according to the gene (i.e. group) so that the text can be plotted separately 
# - each row of text can be adjusted separately 
df$katG <- ifelse(df$x == 'katG', df$node, NA)
df$ahpC <- ifelse(df$x == 'ahpC', df$node, NA)

sank <- ggplot(df, aes(x = x, next_x = next_x, node = node, next_node = next_node))
sank <- sank + coord_flip()
sank <- sank + geom_sankey(flow.alpha = .6, fill = 'lightblue', na.rm = T)
sank <- sank + geom_sankey_text(aes(label = katG), size = 3, angle = 45, width = 2, na.rm = T, vjust = -2, hjust = 0)
sank <- sank + geom_sankey_text(aes(label = ahpC), size = 3, angle = 48, width = 2, na.rm = T, vjust = 2, hjust = 1.5)
sank <- sank + theme_minimal()
sank <- sank + labs(x = NULL)
sank <- sank + theme(legend.position = "none",
        plot.title = element_text(hjust = .5),
        axis.text.x=element_blank(), 
        panel.grid = element_blank())
sank <- sank + ggtitle("INH")
sank

sank_file <- "../results/ahpc_katg_links.png"

ggsave(sank, file = sank_file)

knitr::include_graphics(sank_file)

```

### References
































