---
title: "Large-scale genomic analysis of *Mycobacterium tuberculosis* reveals extent of target and compensatory mutations linked to isoniazid, rifampicin, and multi-drug resistance"
output: 
  bookdown::word_document2: 
    fig_caption: yes
    table_caption: yes
    number_sections: no
mainfont: Calibri
fontsize: 10pt
bibliography: all_refs.bib
csl: biomed-central.csl
---

```{r setup, include=FALSE, echo = F, warning=F, message=F}

# SETUP ----

rm(list = ls())

knitr::opts_chunk$set(echo = F)
# options(scipen=1, digits=2)
options(scipen=999, digits = 3)
table_font_sz <- 8

```

```{r library, include=FALSE, echo = F, warning=F, message=F}

library(tidyr)
library(plyr)
library(dplyr)
library(knitr)
library(reshape2)
library(janitor)
library(ggsankey)
library(ggplot2)
library(stringr)
library(scales)
library(ggrepel)

```

```{r functions, echo = F, warning=F, message=F}

source("https://raw.githubusercontent.com/GaryNapier/Packages_functions/master/Functions.R")

uniq_col <- function(x, split = "; "){
  unique(unlist(strsplit(x, split)))
}

col_pc <- function(df, col){
  round((df[, col] / sum(df[, col]))*100, 3)
}

len_uniq <- function(x){
  length(unique(x))
}

rbind_force <- function(df1, df2){
  rbind(df1, setNames(df2, names(df1)))
}

# Make quick pivot table summarising total unique values by column
# my_formula arg must be call to the formula() function: formula(<col> ~ 'n')
# Result:
#             PRM            n
#  katG-c.2223A>G  3   (4.05%)
# katG-p.Arg484His  5   (6.76%)
# ...
# katG-p.Tyr413Cys  8  (10.81%)
#           Total 74 (100.00%)
pivot <- function(x, my_formula, value_var = "wgs_id"){
  drop_cols(to_table(reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq),
                     pc_dir = 'col'), 
            'Total')
}

pivot_num <- function(x, my_formula,value_var = "wgs_id", ...){
  reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq, ...)
}

split_pivot <- function(df, my_formula, leave_col, value_var = "wgs_id", split_on_col = "Drug"){
  
  to_table_split_pivot <- function(x, col){
    x[[col]] <- as.character(x[[col]])
    x <- drop_cols(to_table(x, "col"), "Total")
    x[[col]] <- gsub("\\s*\\([^\\)]+\\)", "", x[[col]])
    x
  }
  
  df_pivot <- reshape2::dcast(df, my_formula, value.var = 'wgs_id', fun.aggregate = len_uniq)
  df_split <- split(df_pivot, df_pivot[, split_on_col])
  
  do.call('rbind', lapply(df_split, to_table_split_pivot, leave_col))
  
}

duped <- function(x, col){
  dups <- x[duplicated(x[col]), col]
  x[x[, col] %in% dups, ]
}

# Swap first and second parts of string, splitting on a charachter (or substring)
# e.g. 
# x <- "katG-p.Ile335Val; katG-p.Ser315Thr"
# swap_str(x, "; ")
# "katG-p.Ser315Thr; katG-p.Ile335Val"
swap_str <- function(x, split_chr = "; "){
  paste0(unlist(strsplit(x, split_chr))[c(2, 1)], collapse = split_chr)
}


# Print each element of vector vertically instead of across the screen
# e.g. 
# > x <- c("a", "b", "c")
# > x
# [1] "a" "b" "c"
# > print_vert(x)
# a
# b
# c
print_vert <- function(x){
  cat(paste0(x, '\n'))
}

# Remove change from second gene-change pair
# e.g.
# x <- "katG-p.Ser315Thr; fabG1-c.-15C>T"
# > drop_change(x, "-")
# [1] "katG-p.Ser315Thr; fabG1"
drop_change <- function(string, split_on = "-"){
  unlist(lapply(strsplit(string, split_on), function(x){paste0(x[c(1, 2)], collapse = split_on)}))
}

clean_binary_table <- function(x){
  x <- data.frame(apply(x, 2, function(x){gsub("\\[\\]", NA, as.character(x))}))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\[||\\]", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\), \\(", "; ", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\', \\'", "-", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\(||\\)", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\'", "", as.character(x)) }))
  x
}

sort_pos <- function(x, col){
  x$pos <- as.numeric(stringr::str_extract(x[, col], "[0-9]+"))
  x <- x[order(x$pos), ]
  drop_cols(x, 'pos')
}

lapply_nms <- function(x, fun, nms){
  res <- lapply(x, fun)
  names(res) <- nms
  res
}

drug_col_front <- function(df_list){
  lapply(df_list, function(x){ select(x, Drug, everything()) })
}

add_drug_col <- function(df_list, drugs){
  drug_col_front(lapply_mod(df_list, function(i){df_list[[i]]$Drug <- rep(drugs[i], nrow(df_list[[i]])); df_list[[i]]}))
}

find_pos <- function(x){
  sub("\\-", "", gsub("[^0-9\\-]+", "", x))
}

fmt_pc <- function(x, rnd = 2, ...){
  paste0(fmt(round(x*100, rnd), ...), "%")
}

trimws_df <- function(x){
  data.frame(sapply(x, trimws))
}

tonum <- function(x){
  as.numeric(gsub(",", "", strsplit(x, " ")[[1]][1]))
}

uniq_vars <- function(x){
  unique(unlist(strsplit(x, "; ")))
}

rm_null <- function(x){
  x[!sapply(x,is.null)] 
}

rm_norow <- function(x){
  x[sapply(x, nrow) > 0]
}

rm_na <- function(x){
  x[!is.na(x)]
}

pivot_numeric <- function(x, my_formula, value_var = "wgs_id"){
  reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq)
}

print_chisq <- function(chisq_result, scientific = T, rnd = 3){
  paste0("X-sq = ", round(chisq_result$statistic, rnd), 
        "; df = ", round(chisq_result$parameter, rnd), 
        "; p = ", format(chisq_result$p.value, scientific = scientific))
}

clean_del_ins_dup <- function(x){
  gsub("(del|ins|dup).*", "\\1", x)
}

drop_col_name <- function(x, name){
  # Return col names except named col names
  names(x)[!(names(x) %in% name)]
}

rbind_mod <- function(df1, df2, names){
  rbind(setNames(df1, names), setNames(df2, names))
}


split_df_on_haplotype <- function(df, col){
  # Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
  data.frame(df %>% mutate(col = strsplit(as.character(col), "; ")) %>% unnest(col))
}

strsplit_mod <- function(x,
                     split,
                     type = "remove",
                     perl = FALSE,
                     ...) {
  if (type == "remove") {
    # use base::strsplit
    out <- base::strsplit(x = x, split = split, perl = perl, ...)
  } else if (type == "before") {
    # split before the delimiter and keep it
    out <- base::strsplit(x = x,
                          split = paste0("(?<=.)(?=", split, ")"),
                          perl = TRUE,
                          ...)
  } else if (type == "after") {
    # split after the delimiter and keep it
    out <- base::strsplit(x = x,
                          split = paste0("(?<=", split, ")"),
                          perl = TRUE,
                          ...)
  } else {
    # wrong type input
    stop("type must be remove, after or before!")
  }
  return(out)
}

get_gene <- function(x, split_on = "-"){
  unlist(lapply(strsplit(x, split_on), function(x){x[1]}))
}

drop_gene <- function(x, split_on = "-"){
unlist(lapply(strsplit_mod(x, split_on, "before"), function(x){
  if(length(x) < 3){
    x <- paste0(x[2:length(x)], collapse = "")
    x <- gsub("-p.", "", x)
    gsub("-c.", "", x)
  }else{
    x <- paste0(x[3:length(x)], collapse = "")
    x <- gsub("-p.", "", x)
    gsub("-c.", "", x)
  }
  }))
}

try_read_csv <- function(x){
  tryCatch({read.csv(x)}, error = function(e){})
}

add_pc_col <- function(x, col, divide_by, rnd = 2){
  x$pc <- fmt_pc(x[, col]/divide_by, rnd)
  x[col] <- fmt(x[, col])
  x
}

changeSciNot <- function(n) {
  output <- format(n, scientific = TRUE) #Transforms the number into scientific notation even if small
  # output <- sub("e", "*10^", output) #Replace e with 10^
  output <- sub("e", "x10", output) #Replace e with 10^
  output <- sub("\\+0?", "", output) #Remove + symbol and leading zeros on expoent, if > 1
  output <- sub("-0?", "-", output) #Leaves - symbol but removes leading zeros on expoent, if < 1
  output
}

lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

rm_dup_group <- function(df, col){
  # If rows of groups are repeated in a col, replace repeats with blank ("")
  # For final display of tables
  # e.g. 
  # in
  # lineage value
  # L1      10
  # L1      20
  # L1      25
  # L2      100
  # L2      3
  # L2      21
  # out
  # lineage value
  # L1      10
  #         20
  #         25
  # L2      100
  #         3
  #         21
  df[which(duplicated(df[col])), col] <- ""
  df
}

```

```{r variables, echo = F, warning=F, message=F}


# VARIABLES ----

id_col <- "wgs_id"
rnd <- 3
# drugs <- c("rifampicin", "isoniazid", "ethambutol", "pyrazinamide", "streptomycin", 
#              "ofloxacin", "moxifloxacin", "levofloxacin", "amikacin", "kanamycin", "capreomycin", "ciprofloxacin", "prothionamide", 
#              "ethionamide", "para_aminosalicylic_acid", "cycloserine",
#              "clarithromycin", "clofazimine", "bedaquiline", "linezolid", "rifabutin", "delamanid")
drugs <- c("isoniazid", "rifampicin")
fq <- "fluoroquinolones, ofloxacin, levofloxacin, ciprofloxacin, moxifloxacin"
agc <- "capreomycin, streptomycin, amikacin, kanamycin, aminoglycosides"
align <- c("l", "r", "r")
blue_col <- "cornflowerblue"
plot_text_sz <- 0.5
main_cex <- 1.5
# dr_status_vect <- c("Sensitive", "Pre-MDR", "MDR", "Pre-XDR", "XDR", "Other")
dr_status_vect <- c("Sensitive", "Pre-MDR-TB", "HR-TB", "RR-TB", "MDR-TB", "Pre-XDR-TB", "XDR-TB", "Other")
rif_lin_PRM <- c("rpoB-p.Val695Leu", "rpoB-c.-199C>T", "rpoB-p.Ile925Val")

```

```{r paths, echo = F, warning=F, message=F}

# PATHS ----

setwd("~/Documents/comp_mut/r_scripts/")

metadata_path <- "../../metadata/"
local_metadata_path <- "../metadata/"
pipeline_db_path <- "../../pipeline/db/"
db_path <- "../../tbdb/"
results_path <- "../results/"
newick_path <- paste0(results_path, "newick/")

```

```{r files, echo = F, warning=F, message=F}

# FILES ----

# Metadata
metadata_file <- paste0(metadata_path, "tb_data_18_02_2021.csv")
# Database data
lineage_conversion_file <- paste0(pipeline_db_path, "lineage_conversions.txt")
dr_genes_file <- paste0(db_path, "tbdb.csv")
KCM_file <- paste0(pipeline_db_path, "compensatory_mutations.csv")
CM_RM_gene_assoc_file <- paste0(local_metadata_path, "CM_RM_gene_assoc.csv")
protein_model_file <- paste0(results_path, "protein_model.png")


```

```{r drug-files, echo = F, warning=F, message=F}

binary_table_files <- paste0(results_path, drugs, "_binary_table.csv")
PCM_KCM_merged_files <- paste0(results_path,  drugs, "_PCM_KCM_merged.csv")
summary_files <- paste0(results_path, drugs, "_summary.csv")
PRM_stats_files <- paste0(results_path, drugs, "_PRM_stats.csv")
tree_files <- paste0(results_path, drugs, "_tree.png")
inh_numbers_file <- paste0(results_path, "isoniazid_numbers.png")
rif_numbers_file <- paste0(results_path, "rifampicin_numbers.png")

```

```{r read-in-data, echo = F, warning=F, message=F}

# READ IN DATA ----

# Metadata
metadata <- read.csv(metadata_file)
# Database data
lineage_conversions <- read.delim(lineage_conversion_file)
# dr_genes <- read.delim(dr_genes_file, header = F)
dr_genes <- read.csv(dr_genes_file)
KCM <- read.csv(KCM_file)
CM_RM_gene_assoc <- read.csv(CM_RM_gene_assoc_file)

```

```{r drugs-read-in-data, echo = F, warning=F, message=F}

binary_tables <- lapply_nms(binary_table_files, read.csv, drugs)
PCM_KCM_merged <- lapply_nms(PCM_KCM_merged_files, read.csv, drugs)
# summaries <- lapply_nms(summary_files, read.csv, drugs)
# PRM_stats <- rm_null(lapply_nms(PRM_stats_files, try_read_csv, drugs))

```

```{r clean-data, echo = F, warning=F, message=F}

# Binary tables 

# Remove all the "[]" present in the binary table caused by python's lists
binary_tables <- lapply(binary_tables, clean_binary_table)

# Clean lineage column 
binary_tables <- lapply(binary_tables, function(x) {x$main_lineage <- gsub("lineage", "", x$main_lineage); x})
binary_tables <- lapply(binary_tables, function(x) {x$sublin <- gsub("lineage", "", x$sublin); x})

# Trim white space
binary_tables <- lapply(binary_tables, trimws_df)

# Rifampicin filter - remove lineage-specific PRMS
# NB - TOTAL HACK TO REMOVE THE LINEAGE-SPECIFIC RIF PRMs
binary_tables$rifampicin$PRM <- ifelse(binary_tables$rifampicin$PRM %in% rif_lin_PRM, NA, binary_tables$rifampicin$PRM)

# REMOVE NON-p. FROM KATG PRMs
binary_tables$isoniazid$PRM <- ifelse(!(grepl("p.", binary_tables$isoniazid$PRM)), NA, binary_tables$isoniazid$PRM)

# Make binary cols for CM KRM and PRM
binary_tables <- lapply(binary_tables, function(x){
  data.frame(cbind(x, 
                   CM_bin = ifelse(is.na(x$CM), "absent", "present"), 
                   KRM_bin = ifelse(is.na(x$KRM), "absent", "present"), 
                   PRM_bin = ifelse(is.na(x$PRM), "absent", "present")))
})

# Make katG and rpoB binary
binary_tables$isoniazid$KRM_katG_bin <- ifelse(is.na(binary_tables$isoniazid$KRM_katG), "absent", "present")
binary_tables$rifampicin$KRM_rpoB_bin <- ifelse(is.na(binary_tables$rifampicin$KRM_rpoB), "absent", "present")


# Write out 
write.csv(binary_tables$isoniazid, file = paste0(results_path, "isoniazid_binary_table.csv"), quote = F, row.names = F)


# ---

# Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
binary_tables_dup <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>%
  unnest(other_vars))
})

# Make col for just the other_vars genes, stripping out the change
binary_tables_dup <- lapply(binary_tables_dup, function(x){x$other_vars_gene <- gsub("-.*", "", x$other_vars); x})

# ---

# Split out as sep vars

inh_bin <- binary_tables$isoniazid
rif_bin <- binary_tables$rifampicin

inh_bin_dup <- binary_tables_dup$isoniazid
rif_bin_dup <- binary_tables_dup$rifampicin

# --- 

# KCM - combine gene-change
KCM$gene_change <- paste0(KCM$Gene, "-", KCM$Mutation)

# Subset KCM 
inh_KCM <- KCM[KCM[,"Drug"] == "isoniazid", ]
rif_KCM <- KCM[KCM[,"Drug"] == "rifampicin", ]


# Put PCM_merged together and join in the drug names
PCM_KCM_merged <- do.call("rbind", PCM_KCM_merged)
PCM_KCM_merged$gene_change <- paste0(PCM_KCM_merged$gene, "-", PCM_KCM_merged$change)
PCM_KCM_merged <- merge(PCM_KCM_merged, CM_RM_gene_assoc, 
                    by.x = "gene", by.y = "CM_gene", 
                    sort = F)
# Merge in status
PCM_KCM_merged$status <- ifelse(PCM_KCM_merged$gene_change %in% KCM$gene_change, "known", "unknown")

# Remove the PCMs in the KCM
# PCM_KCM_merged <- PCM_KCM_merged[!(PCM_KCM_merged$gene_change %in% KCM$gene_change), ]

# Transpose summary
# summaries <- do.call('cbind', lapply(summaries, function(x){data.frame(n = t(x))}))
# names(summaries) <- drugs



```

```{r total-samps, echo = F, warning=F}

total_samps <- nrow(binary_tables[[1]])

```

```{r isoniazid-PRM-cnts, echo = F, warning=F}

isoniazid_binary_table <- binary_tables$isoniazid
inh_PRM_samps <- subset(isoniazid_binary_table, PRM_bin == "present")
inh_unique_PRMs <- unique(unlist(strsplit(inh_PRM_samps$PRM, "; ")))
n_inh_unique_PRMs <- length(inh_unique_PRMs)


```

```{r isoniazid-rare-mutations-cnts, echo = F, warning=F}


rare_katg_table <- subset(isoniazid_binary_table, is.na(KRM_katG) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_katG)) )
n_samps_rare_katg <- nrow(rare_katg_table)
n_rare_katg <- length(uniq_col(rare_katg_table$other_katG))
rare_katg_cnts <- tab2df(table(table(unlist(strsplit(rare_katg_table$other_katG, "; ")))))
rare_katgs <- sort_pos(data.frame(Change = uniq_col(rare_katg_table$other_katG)))
inh_n_unique_rare_katg <- length(unique(rare_katgs$Change))

```

```{r rifampicin-rare-mutations-cnts, echo = F, warning=F}

rifampicin_binary_table <- binary_tables$rifampicin
rare_rpob_table <- subset(rifampicin_binary_table, is.na(KRM_rpoB) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_rpoB)) )
n_samps_rare_rpob <- nrow(rare_rpob_table)
n_rare_rpob <- length(uniq_col(rare_rpob_table$other_rpoB))
rare_rpob_cnts <- tab2df(table(table(unlist(strsplit(rare_rpob_table$other_rpoB, "; ")))))
rare_rpobs <- sort_pos(data.frame(Change = uniq_col(rare_rpob_table$other_rpoB)))

```

```{r rare-mutations-cnts, echo = F, warning=F}

rare_cnts <- plyr::rbind.fill(rare_katg_cnts, rare_rpob_cnts)

for(i in seq(nrow(rare_cnts))){
  row.names(rare_cnts)[row.names(rare_cnts) == i] <- drugs[i]
}

```

```{r table-1, echo = F, warning=F}

# Make pivot tables of basic stats

# Wrangling

# Add drug col, select and bind
meta_tables <- add_drug_col(binary_tables, drugs)
meta_table <- do.call('rbind', lapply(meta_tables, 
                                      function(x){ 
                                        select(x, Drug, wgs_id, main_lineage, sublin, country_code, drtype, dst, KRM)
                                        }))

# Trim whitespace
meta_table$dst <- trimws(meta_table$dst)

# Aggregate lineages
meta_table$lin_agg <- ifelse(!(meta_table$main_lineage == "1" | 
                                 meta_table$main_lineage == "2" | 
                                 meta_table$main_lineage == "3" |
                                 meta_table$main_lineage == "4"), "Other", meta_table$main_lineage)

# Use the KRM to define resistance to INH or RIF and create new DR type col
meta_table$drug_res <- ifelse(meta_table$Drug == "isoniazid" & !(is.na(meta_table$KRM)), "HR-TB",
                              ifelse(meta_table$Drug == "rifampicin" & !(is.na(meta_table$KRM)), "RR-TB",
                                                                         meta_table$KRM))


# Pivot tables

# Lin pivots
lin_pivot <- add_pc_col(pivot_num(meta_table, formula(lin_agg ~ 'n')), "n", total_samps)

# DR pivots
# INH and RIF resistance
drugs_pivot <- add_pc_col(pivot_num(meta_table, formula(drug_res ~ 'n')), 'n', total_samps, rnd = 3)
drugs_pivot <- subset(drugs_pivot, !(is.na(drug_res)))
# All other types resistance
drtype_pivot <- add_pc_col(pivot_num(meta_table, formula(drtype ~ 'n')), 'n', total_samps, rnd = 2)
drtype_pivot <- rbind_force(drtype_pivot, drugs_pivot)
drtype_pivot <- subset(drtype_pivot, !(drtype == "Pre-MDR-TB"))
order_by <- c("Sensitive", "RR-TB", "HR-TB", "MDR-TB", "Pre-XDR-TB", "XDR-TB", "Other")
drtype_pivot <- trimws_df(drtype_pivot[order(match(drtype_pivot$drtype, order_by)), ])

# # Predicted DR to individual drugs
KRM_table <- do.call('rbind', lapply(meta_tables, function(x){ select(x, Drug, wgs_id, KRM)}))
KRM_table$KRM_binary <- ifelse(is.na(KRM_table$KRM), 0, 1)
# KRM_pivot <- trimws_df(split_pivot(KRM_table, formula(Drug + KRM_binary ~ 'n'), "KRM_binary"))

# DST
# dst_pivot <- pivot(meta_table, formula(Drug + dst ~ 'n') )
# dst_pivot <- trimws_df(split_pivot(meta_table, formula(Drug + dst ~ 'n'), "dst"))

table_1 <- rbind_force(lin_pivot, drtype_pivot)
table_1$Characteristic <- c(rep("Lineage", nrow(lin_pivot)),
                      rep("Genotypic status", nrow(drtype_pivot)))
names(table_1) <- c("-", "N", "%", "Characteristic")
table_1 <- table_1[, c("Characteristic", "-", "N", "%")]

 
# table_1 <- data.frame()
# lin_dr <- rbind(table_1,
#                 setNames(lin_pivot, names(table_1)),
#                 setNames(drtype_pivot, names(table_1)))
# 
# lin_dr <- cbind(rep("-", nrow(lin_dr)), lin_dr)
# 
#                  
# KRM_dst <- rbind(table_1,
#                  setNames(KRM_pivot, names(table_1)),
#                  setNames(dst_pivot, names(table_1)))
# 
# table_1 <- rbind_force(lin_dr, KRM_dst)
# 
# names(table_1) <- c("Drug", "Chararcteristic", "n (%)")
# 
# table_1$Category <- c(rep("Lineage", nrow(lin_pivot)),
#                       rep("DR status", nrow(drtype_pivot)),
#                       rep("(predicted) resistance", nrow(KRM_pivot)),
#                       rep("DST", nrow(dst_pivot)))
# 
# table_1 <- table_1[, c("Category", "Drug", "Chararcteristic", "n (%)")]

```

```{r DST-counts, echo = F, warning = F}

inh_n_samps_dst <- nrow(subset(meta_table, Drug == "isoniazid" & !is.na(dst)))
rif_n_samps_dst <- nrow(subset(meta_table, Drug == "rifampicin" & !is.na(dst)))

inh_n_samps_DST_res <- nrow(subset(meta_table, Drug == "isoniazid" & dst == "1"))
rif_n_samps_DST_res <- nrow(subset(meta_table, Drug == "rifampicin" & dst == "1"))

n_samps_dst_mdr <- nrow(intersect(dplyr::select(subset(meta_table, Drug == "isoniazid" & dst == "1"), wgs_id), 
dplyr::select(subset(meta_table, Drug == "rifampicin" & dst == "1"), wgs_id)))


```

```{r n-samples-tree-inh, echo = F, warning=F}

# Tree

# INH
# # 1
# total_samps
# # 2
inh_2_a <- nrow(subset(inh_bin, KRM_katG_bin == "present"))
inh_2_b <- nrow(subset(inh_bin, KRM_katG_bin == "absent"))
inh_2_a_b <- inh_2_a + inh_2_b
# # 3
inh_3_a <- nrow(subset(inh_bin, KRM_katG_bin == "present" & CM_bin == "present"))
inh_3_b <- nrow(subset(inh_bin, KRM_katG_bin == "present" & CM_bin == "absent"))
inh_3_c <- nrow(subset(inh_bin, KRM_katG_bin == "absent" & CM_bin == "present"))
inh_3_d <- nrow(subset(inh_bin, KRM_katG_bin == "absent" & CM_bin == "absent"))
inh_3_a_c <- inh_3_a + inh_3_c
# 4
# a
inh_4_a <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present"))
# b
inh_4_b <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# c
inh_4_c <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# d
inh_4_d <- nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# e
inh_4_e <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present"))
# f
inh_4_f <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# g
inh_4_g <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# h
inh_4_h <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent"))

inh_4_c_g <- inh_4_c + inh_4_g
inh_4_a_c_e_g <- inh_4_a + inh_4_c + inh_4_e + inh_4_g
# 5
# a
inh_5_a <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           !(is.na(other_katG))))
# b
inh_5_b <- nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           is.na(other_katG)))

layer_2 <- data.frame(n = c(inh_2_a, inh_2_b))
layer_3 <- data.frame(n = c(inh_3_a, 
                            inh_3_b, 
                            inh_3_c, 
                            inh_3_d))
layer_4 <- data.frame(n = c(inh_4_a, 
                            inh_4_b, 
                            inh_4_c, 
                            inh_4_d, 
                            inh_4_e, 
                            inh_4_f, 
                            inh_4_g, 
                            inh_4_h))
layer_5 <- data.frame(n = c(inh_5_a, inh_5_b))

n_df <- rbind(layer_2, layer_3, layer_4, layer_5)

layer <- data.frame(c(
  paste0(rep("2", nrow(layer_2)), "_", letters[1:nrow(layer_2)]), 
  paste0(rep("3", nrow(layer_3)), "_", letters[1:nrow(layer_3)]), 
  paste0(rep("4", nrow(layer_4)), "_", letters[1:nrow(layer_4)]), 
  paste0(rep("5", nrow(layer_5)), "_", letters[1:nrow(layer_5)])
))

inh_samp_tree <- setNames(cbind(layer, n_df), c("layer", "n"))

```

```{r n-samples-tree-rif, echo = F, warning=F}



# # RIF
# # 1
# total_samps
# # 2
rif_2_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present"))
rif_2_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent"))
# 3
rif_3_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present"))
rif_3_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent"))
rif_3_c <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present"))
rif_3_d <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent"))
rif_3_a_c <- rif_3_a + rif_3_c
# 4
# a
rif_4_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present"))
# b
rif_4_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# c
rif_4_c <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# d
rif_4_d <- nrow(subset(rif_bin, KRM_rpoB_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# e
rif_4_e <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present"))
# f
rif_4_f <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# g
rif_4_g <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# h
rif_4_h <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# 5
# a
rif_5_a <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           !(is.na(other_rpoB))))
# b
rif_5_b <- nrow(subset(rif_bin, KRM_rpoB_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           is.na(other_rpoB)))

layer_2 <- data.frame(n = c(rif_2_a, rif_2_b))
layer_3 <- data.frame(n = c(rif_3_a, 
                            rif_3_b, 
                            rif_3_c, 
                            rif_3_d))
layer_4 <- data.frame(n = c(rif_4_a, 
                            rif_4_b, 
                            rif_4_c, 
                            rif_4_d, 
                            rif_4_e, 
                            rif_4_f, 
                            rif_4_g, 
                            rif_4_h))
layer_5 <- data.frame(n = c(rif_5_a, rif_5_b))

n_df <- rbind(layer_2, layer_3, layer_4, layer_5)

layer <- data.frame(c(
  paste0(rep("2", nrow(layer_2)), "_", letters[1:nrow(layer_2)]), 
  paste0(rep("3", nrow(layer_3)), "_", letters[1:nrow(layer_3)]), 
  paste0(rep("4", nrow(layer_4)), "_", letters[1:nrow(layer_4)]), 
  paste0(rep("5", nrow(layer_5)), "_", letters[1:nrow(layer_5)])
))

rif_samp_tree <- setNames(cbind(layer, n_df), c("layer", "n"))


```

```{r working, echo = F, warning=F}


```

| | |
| ----------- | ----------- |
|Gary Napier^1^ | gary.napier@lshtm.ac.uk |
|Julian Libiseller-Egger^1^ | julian.libiseller-egger@lshtm.ac.uk | 
|Susana Campino^1^ | susana.campino@lshtm.ac.uk |
|Jody E. Phelan^1^,* | jody.phelan@lshtm.ac.uk |
|Taane G. Clark^1^,^2^,* | taane.clark@lshtm.ac.uk |

^1^Faculty of Infectious and Tropical Diseases, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK
^2^Faculty of Epidemiology and Population Health, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK

\* joint authors

Corresponding authors:
Dr. Jody Phelan and Prof. Taane Clark 
Department of Infection Biology
Faculty of Infectious and Tropical Diseases
London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK 


## ABSTRACT

**Background:** Resistance to isoniazid (INH) and rifampicin (RIF) first-line drugs in *Mycobacterium tuberculosis* threatens tuberculosis disease control. Mutations in *katG* and *rpoB*, which confer resistance to INH and RIF, respectively, often come with a fitness cost. To overcome these fitness costs, *M. tuberculosis* compensatory mutations have arisen in *rpoC*, *rpoA* (RIF) and *ahpC* (INH) genes. Here we leverage the presence of known and putative compensatory mutations along with phenotypic and genotypic data to detect the presence of novel resistance mutations occurring in the target genes of INH and RIF.

**Results:** Across ~32k *M. tuberculosis* isolates with whole genome sequence data, there were `r fmt(inh_n_samps_DST_res)` (`r fmt_pc(inh_n_samps_DST_res/inh_n_samps_dst, digits = 3)`)  with INH and `r fmt(rif_n_samps_DST_res)` (`r fmt_pc(rif_n_samps_DST_res/rif_n_samps_dst, digits = 3)`)  with RIF phenotypic resistance. Known mutations in *katG* and *rpoB* explained ~99% of the resistance. However, `r fmt(inh_3_c)` (`r fmt_pc(inh_3_c/total_samps, digits = 1)`) isolates had *ahpC* compensatory mutations but did not have any known resistance mutations in the INH target gene *katG*, leading the identification of `r n_inh_unique_PRMs` novel putative resistance mutations in *katG* observed in at least three isolates. These putative *katG* mutations co-occur with other INH variants (e.g., *katG*-Ser315Thr, *fabG1* mutations). There were no isolates which had *rpoC*/*rpoA* compensatory mutations but no known resistance mutations for RIF.

**Conclusions:** Using whole genome sequence to establish and detect the full repertoire of resistance mutations will inform the clinical management of tuberculosis. We have identified new resistance markers which, after validation, can be used for genotypic drug-resistance profiling.

**URL:**https://github.com/GaryNapier/comp_mut

**Keywords:** *Mycobacterium tuberculosis*, isoniazid, rifampicin, *katG*, *rpoB*, compensatory/resistance mutations

## BACKGROUND

Tuberculosis (TB), caused by *Mycobacterium tuberculosis* (Mtb) bacteria, is a major global public health problem. TB control is being made difficult by drug resistance, especially to first-line rifampicin (RIF) and isoniazid (INH), together called multi-drug resistance (MDR-TB). In gaining resistance against anti-TB drugs, Mtb drug targets or activating proteins are often compromised in their usual function, and therefore the bacterium incurs a fitness cost. These costs can manifest as a phenotypic difference, such as reduced virulence or transmissibility. For example, the *katG* gene codes for the KatG enzyme, a catalase-peroxidase that protects the bacterium from reactive oxygen species damage and is used to detoxify hydrogen peroxide [@Trivedi2012], improving survival within macrophages and the host immune response. The enzyme also activates the pro-drug INH, converting it to multiple reactive species [@Zhang2009].



When mutations occur in the *katG* gene that disrupt INH binding to KatG, the bacteria are often left with drug resistance and a protein with impaired enzymatic function. In some cases, mutations can confer drug resistance without a punitive fitness cost. For example, the *katG* Ser315Thr mutation confers resistance but minimally affects fitness, hence is highly prevalent among (pre-)MDR-TB strains [@Hazbon2006][@Phelan2019]. For RIF, the target is the $\beta$' subunit of RNA polymerase, coded by *rpoB*. Mutations in *rpoB* prevent RIF from binding, but incur high fitness cost since the intricate machinery of RNA polymerase is intolerant to large structural changes [@Telenti1997]. One notable exception is *rpoB* Ser450Leu, which is highly prevalent in RIF-resistant strains [@Heep2001]. Indeed, so restrictive are changes to the $\beta$ subunit, more than 95% of drug resistance mutations occur in the RIF resistance determining region (RRDR), an 81 base-pair section of the rpoB gene [@Cao2019].

To overcome this fitness cost, secondary mutations can occur which improve or promote either the target protein itself or an alternative with a similar function. In the case of INH/*katG* the protein AhpC, which has a similar enzymatic function, is often promoted via mutations in the promoter of the *ahpC* gene [@Sherman1999][@Sherman1996]. RIF compensatory mutations occur in the other RNA polymerase subunits $\alpha$ (*rpoA*), $\beta$' (*rpoC*) or even within the $\beta$ subunit itself. These mutations are thought to occur at the interfaces of the subunits, helping to restore overall RNA polymerase function while maintaining an altered binding site in the $\beta$ subunit [@Comas2011]. 

The TB-profiler platform [@Phelan2019] uses 2,300 mutations across 35 loci to profile Mtb resistance across 21 anti-TB drugs, including RIF and INH. However, the full repertoire of resistance mutations, including for MDR-TB is not fully characterised. The accompanying TB-profiler database consists of ~32k isolates has whole genome sequence and drug susceptibility test (DST) phenotypic data, with genotypic profiles. Here, by investigating those samples with compensatory mutations but no known resistance mutations, we attempt to identify the presence of novel mutations linked to genes for INH, RIF, and MDR-TB. Further, we attempt to understand the patterns of co-existence between resistance and compensatory mutations in relation to INH and RIF drug resistance.


## RESULTS

```{r results-isolate-collection, echo = F, warning=F}

# **Isolate collection**
inh_n_samps_TBP_res <- nrow(subset(KRM_table, Drug == "isoniazid" & KRM_binary == "1"))
rif_n_samps_TBP_res <- nrow(subset(KRM_table, Drug == "rifampicin" & KRM_binary == "1"))

n_samps_MDR <- nrow(subset(meta_table, Drug == "isoniazid" & drtype == "MDR-TB"))

n_samps_DST_total <- sum(!is.na(inh_bin$dst))

```

*Isolate data*

A total of `r fmt(total_samps)` Mtb isolates with whole genome sequencing and DST data were analysed, and encompassed all major lineages (L4 51.1%, L2 25.3%, L3 11.5%, L1 9.7%) (**Table \@ref(tab:table-1-display)**). 

Across the `r fmt(n_samps_DST_total)` samples with DST data, `r fmt(inh_n_samps_DST_res)` (`r fmt_pc(inh_n_samps_DST_res/inh_n_samps_dst, digits = 3)`), `r fmt(rif_n_samps_DST_res)` (`r fmt_pc(rif_n_samps_DST_res/rif_n_samps_dst, digits = 3)`) and `r n_samps_dst_mdr` (`r fmt_pc(n_samps_dst_mdr/n_samps_DST_total)`) were resistant to INH, RIF, and MDR-TB, respectively. Genotypic resistance prediction using TB-Profiler software across all isolates inferred that `r fmt(inh_n_samps_TBP_res)` (`r fmt_pc(inh_n_samps_TBP_res/total_samps, digits = 3)`) and `r fmt(rif_n_samps_TBP_res)` (`r fmt_pc(rif_n_samps_TBP_res/total_samps, digits = 3)`), `r fmt(n_samps_MDR)` (`r fmt_pc(n_samps_MDR/total_samps, digits = 3)`) were resistant to INH, RIF, and MDR-TB, respectively (**Table \@ref(tab:table-1-display)**). 

```{r results-common-KRM-mutations, echo = F, warning=F}

# Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
binary_tables_KRM_dup <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(KRM = strsplit(as.character(KRM), "; ")) %>%
  unnest(KRM))
})

inh_KRM_dup <- binary_tables_KRM_dup$isoniazid
inh_KRM_cnt <- data.frame(inh_KRM_dup %>% dplyr::count(KRM))
inh_KRM_cnt <- inh_KRM_cnt[order(inh_KRM_cnt$n, decreasing = T),]
inh_KRM_top <- head(subset(inh_KRM_cnt, !(is.na(KRM))), 3)

inh_KRM_top_1 <- inh_KRM_top$KRM[1]
inh_KRM_top_1_n <- inh_KRM_top$n[1]
inh_KRM_top_1_pc <- fmt_pc(inh_KRM_top$n[1]/total_samps)

inh_KRM_top_2 <- inh_KRM_top$KRM[2]
inh_KRM_top_2_n <- inh_KRM_top$n[2]
inh_KRM_top_2_pc <- fmt_pc(inh_KRM_top$n[2]/total_samps)

inh_KRM_top_3 <- inh_KRM_top$KRM[3]
inh_KRM_top_3_n <- inh_KRM_top$n[3]
inh_KRM_top_3_pc <- fmt_pc(inh_KRM_top$n[3]/total_samps)

rif_KRM_dup <- binary_tables_KRM_dup$rifampicin
rif_KRM_cnt <- data.frame(rif_KRM_dup %>% dplyr::count(KRM))
rif_KRM_cnt <- rif_KRM_cnt[order(rif_KRM_cnt$n, decreasing = T),]
rif_KRM_top <- head(subset(rif_KRM_cnt, !(is.na(KRM))), 3)

rif_KRM_top_1 <- rif_KRM_top$KRM[1]
rif_KRM_top_1_pc <- fmt_pc(rif_KRM_top$n[1]/total_samps)
rif_KRM_top_2 <- rif_KRM_top$KRM[2]
rif_KRM_top_2_pc <- fmt_pc(rif_KRM_top$n[2]/total_samps)
rif_KRM_top_3 <- rif_KRM_top$KRM[3]
rif_KRM_top_3_pc <- fmt_pc(rif_KRM_top$n[3]/total_samps)



common_KRM_tab <- do.call("rbind", add_drug_col(list(inh_KRM_cnt, rif_KRM_cnt), drugs))

# Clean 
common_KRM_tab <- subset(common_KRM_tab, !(is.na(KRM)))
common_KRM_tab$KRM <- clean_del_ins_dup(common_KRM_tab$KRM)
common_KRM_tab <- subset(common_KRM_tab, n > 4)

```

The most common mutations underlying INH resistance were `r inh_KRM_top_1` (n=`r inh_KRM_top_1_n`; `r inh_KRM_top_1_pc`), `r inh_KRM_top_2` (n=`r inh_KRM_top_2_n`; `r inh_KRM_top_2_pc`), and `r inh_KRM_top_3` (n=`r inh_KRM_top_3_n`; `r inh_KRM_top_3_pc`). Similarly, for the RIF resistance, the most frequent mutations were `r rif_KRM_top_1` (`r rif_KRM_top_1_pc`), `r rif_KRM_top_2` (`r rif_KRM_top_2_pc`) and `r rif_KRM_top_3` (`r rif_KRM_top_3_pc`) (**Table S1**). 

```{r results-compensatory-mutations, echo = F, warning=F}

# **Compensatory mutations**
inh_CM <- select(subset(inh_bin, CM_bin == "present"), CM)

inh_n_CM_haplotypes <- length(unique(inh_CM$CM))
inh_CM_uniq <- uniq_vars(inh_CM$CM)
inh_CM_uniq_in_KCM <- inh_CM_uniq[inh_CM_uniq %in% inh_KCM$gene_change]
inh_PCM <- inh_CM_uniq[!(inh_CM_uniq %in% inh_KCM$gene_change)]

rif_CM <- select(subset(rif_bin, CM_bin == "present"), CM)
rif_n_samps_KRM_CM <- nrow(subset(rif_bin, KRM_bin == "present" & 
                                    CM_bin == "present"))
rif_n_samps_KRM_no_CM <- nrow(subset(rif_bin, KRM_bin == "absent" & 
                                    CM_bin == "present"))
rif_n_CM_haplotypes <- length(unique(rif_CM$CM))
rif_CM_uniq <- uniq_vars(rif_CM$CM)
rif_rpoC <- subset(rif_CM_uniq, grepl("rpoC", rif_CM_uniq))
rif_rpoA <- subset(rif_CM_uniq, grepl("rpoA", rif_CM_uniq))
rif_CM_uniq_in_KCM <- rif_CM_uniq[rif_CM_uniq %in% rif_KCM$gene_change]
rif_PCM <-  rif_CM_uniq[!(rif_CM_uniq %in% rif_KCM$gene_change)]

```

```{r CM-table, echo = F, warning=F}

# Table of compensatory mutations - gene, mutation [n], status 
inh_CM_data <- data.frame(inh_bin %>% mutate(CM = strsplit(as.character(CM), "; ")) %>% unnest(CM))
inh_CM_tab <- sort_pos(pivot_num(subset(inh_CM_data, !is.na(CM)), formula(CM ~ 'n')), "CM")
inh_CM_tab$gene <- get_gene(inh_CM_tab$CM)
inh_CM_tab$mutation <- paste0(clean_del_ins_dup(drop_gene(inh_CM_tab$CM)), " [", inh_CM_tab$n, "]")
inh_CM_tab$status <- ifelse(inh_CM_tab$CM %in% KCM$gene_change, "known", "unknown")
inh_CM_tab$drug <- rep("isoniazid", nrow(inh_CM_tab))
inh_CM_tab <- select(inh_CM_tab, drug, gene, mutation, status)

rif_CM_data <- data.frame(rif_bin %>% mutate(CM = strsplit(as.character(CM), "; ")) %>% unnest(CM))
rif_CM_tab <- pivot_num(subset(rif_CM_data, !is.na(CM)), formula(CM ~ 'n'))
rif_CM_tab$gene <- get_gene(rif_CM_tab$CM)
rif_CM_tab$mutation <- paste0(clean_del_ins_dup(drop_gene(rif_CM_tab$CM)), " [", rif_CM_tab$n, "]")
rif_CM_tab$status <- ifelse(rif_CM_tab$CM %in% KCM$gene_change, "known", "unknown")
rif_CM_tab$drug <- rep("rifampicin", nrow(rif_CM_tab))
rif_CM_tab <- select(rif_CM_tab, drug, gene, mutation, status)
CM_tab <- rbind(inh_CM_tab, rif_CM_tab)
CM_tab$status <- ifelse(CM_tab$status == "unknown", "*", "")
CM_tab <- rm_dup_group(CM_tab, "drug")
CM_tab <- rm_dup_group(CM_tab, "gene")
names(CM_tab) <- c("Drug", "Compensatory locus", "Mutation [frequency]", "Status")

```

To characterise putative novel resistance mutations, we considered samples which had a compensatory mutation, but no known resistance mutation. A manually curated list of `r nrow(CM_tab)` compensatory mutations from the literature was generated (**Table S2**), which includes mutations in *ahpC* (n=`r length(inh_CM_uniq)`; e.g., -47_-46ins, -48G>A, -51G>A, -52C>A, -52C>T, -81C>T), *rpoC* (n=`r length(rif_rpoC)`; e.g., Asn698Ser, Asp485Asn, Ile491Thr, Ile491Val, Leu516Pro, Trp484Gly, Val483Ala, Val483Gly), and *rpoA* (n=`r length(rif_rpoA)`; e.g., Thr187Ala). The number of occurrences of individual compensatory mutations within the 32k isolates varied for *rpoC*/*A* (RIF, range: 5 – 427 isolates) and *ahpC* (INH, range: 3 – 97 isolates) genes. No isolate had more than one compensatory mutation for RIF or INH, and across MDR-TB.

*Putative novel resistance mutations*

```{r inh-stats, echo = F, warning = F}

# n katG mutations
inh_n_uniq_KRM_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "KRM_katG"])), 
                                                                    "KRM_katG"]), "; "))))
inh_n_uniq_other_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "other_katG"])), 
                                                                      "other_katG"]), "; "))))


# ***** MOST IMPORTANT NUMBER ******
inh_n_uniq_PRM_katg <- length(unique(unlist(strsplit(unique(inh_bin[!(is.na(inh_bin[, "PRM"])), "PRM"]), "; "))))
MOST_IMPORTANT_NUMBER <- inh_n_uniq_PRM_katg
# ***** MOST IMPORTANT NUMBER ******


total_unique_katg <- inh_n_uniq_KRM_katg + inh_n_uniq_other_katg + inh_n_uniq_PRM_katg

```

Using the *rpoA* and *rpoC* compensatory mutations, there were no RIF resistant isolates without a known *rpoB* resistance mutation (**Figure S1**). For INH, there were `r inh_3_a_c` samples with a compensatory mutation, of which `r inh_3_c` (`r fmt_pc(inh_3_c/inh_3_a_c)`) had no known *katG* resistance mutation (**Figure S1**). Within the `r inh_3_c` samples we looked for mutations in *katG* that could potentially explain the emergence of the compensatory mutation. In total, `r total_unique_katg` unique non-synonymous mutations were found in the *katG* gene. Only `r inh_n_uniq_PRM_katg` (`r fmt_pc(inh_n_uniq_PRM_katg/total_unique_katg, 1)`) of these *katG* mutations occurred in at least three isolates, and had >50% of isolates with a resistant DST and genotypic resistance to at least one other drug. These `r inh_n_uniq_PRM_katg` high-quality *katG* mutations were present in `r inh_4_a_c_e_g` isolates, including `r inh_4_e` and `r inh_4_c_g` with and without compensatory mutations, respectively (**Table \@ref(tab:PRM-summary-table)**; **Figure S1**). Of the `r inh_3_c` isolates that had a compensatory mutation, `r inh_4_f` (`r fmt_pc(inh_4_f/inh_3_c)`) did not have any of the `r inh_n_uniq_PRM_katg` highly quality *katG* mutations, but `r inh_5_a` (/`r inh_4_f`; `r fmt_pc(inh_5_a/inh_4_f)`) were found to have rare *katG* mutations that did not pass the minimum sample cut-off (>=3) used to define putative resistance mutations (**Table S3**). These rare *katG* mutations could also potentially explain the acquisition of a compensatory mutation but were not analysed further.

```{r PRM-summary-stats, echo = F, warning=F}

# PRM subset and summary
PRM_tabs <- lapply(binary_tables, function(x){subset(x, PRM_bin == "present")})
PRM_pivot <- rm_null(lapply(PRM_tabs, function(x){
  tryCatch({
    sort_pos(pivot_num(x, formula(PRM ~ 'n')), 'PRM')
  }, error = function(e){})
            }))
PRM_pivot <- add_drug_col(PRM_pivot, drugs)
PRM_summary <- do.call("rbind", PRM_pivot)
PRM_summary <- select(PRM_summary, Drug, PRM, n)

# PRM stats
PRM_stats <- rm_null(lapply_nms(PRM_stats_files, try_read_csv, drugs))
PRM_stats <- lapply(PRM_stats, sort_pos, 'mutation')
# Clean RIF of the lineage-specific PRMs
PRM_stats$rifampicin <- PRM_stats$rifampicin[!(PRM_stats$rifampicin$gene_mutation %in% rif_lin_PRM), ]
PRM_stats <- do.call("rbind", PRM_stats)
PRM_stats <- select(PRM_stats, drug, gene, mutation, n_samps, dst_prop, dr_prop, n_lins)
PRM_stats$dst_prop <- 1 - PRM_stats$dst_prop
PRM_stats$gene_mutation <- paste0(PRM_stats$gene, "-", PRM_stats$mutation)
# Add the PRMs which only appear in one lineage to PRM summary
PRMs_one_lin <- select(subset(PRM_stats, n_lins == 1), gene_mutation)[[1]]
# names(PRM_stats) <- c("Drug", "Gene", "Change", "n samples", "DST sensitive", "DR-type sensitive", "n lineages")
PRM_summary$PRM_one_lin <- stringr::str_extract(PRM_summary$PRM, paste0(PRMs_one_lin, collapse = "|"))

# Clean up and format
PRM_summary$PRM <- gsub("; ", "/", gsub("katG-p.|katG-c.", "", PRM_summary$PRM))
PRM_summary$PRM_one_lin <- gsub("katG-p.|katG-c.", "", PRM_summary$PRM_one_lin)
PRM_summary$PRM <- paste0(PRM_summary$PRM, " [", PRM_summary$n, "]")
PRM_summary <- select(PRM_summary, Drug, PRM, PRM_one_lin)
names(PRM_summary) <- c("Drug", "Mutation [frequency]", "PRM_one_lin")

```

```{r make-rare-mutations-table, echo = F, warning=F}

rare_mutations_table <- tryCatch({
   cbind(Drug = c(rep("isoniazid", nrow(rare_katgs)), 
                  rep("rifampicin", nrow(rare_rpobs))), 
         rbind(rare_katgs, rare_rpobs))
}, error = function(e){
  cbind(Drug = c(rep("isoniazid", nrow(rare_katgs))), 
        rbind(rare_katgs, rare_rpobs))
})

rare_mutations_table$Change <- gsub("katG-c.|katG-p.", "", rare_mutations_table$Change)

# rare_mutations_table <- data.frame(cbind(Drug = unique(rare_mutations_table$Drug), 
#              Change = paste(rare_mutations_table$Change, collapse = ", ")))

rare_mutations_table <- data.frame(Change = paste(rare_mutations_table$Change, collapse = ", "))
names(rare_mutations_table) <- "Mutations in katG gene"

```

*Resistance and co-occurrence with other resistance mutations*

```{r PRM-dr-types, echo = F, warning=F}

PRM_DR_pivot <- pivot_num(PRM_tabs$isoniazid, formula(drtype ~ 'n'))
PRM_DR_max <- PRM_DR_pivot[which.max(PRM_DR_pivot$n), "drtype"]
PRM_DR_max_pc <- fmt_pc(PRM_DR_pivot[which.max(PRM_DR_pivot$n), "n"]/inh_4_a_c_e_g)
PRM_DR_pre_MDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Pre-MDR-TB", "n"]/inh_4_a_c_e_g)
PRM_DR_pre_XDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Pre-XDR-TB", "n"]/inh_4_a_c_e_g)
PRM_DR_XDR_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "XDR-TB", "n"]/inh_4_a_c_e_g, digits = 2)
PRM_DR_sens_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Sensitive", "n"]/inh_4_a_c_e_g, digits = 2)
PRM_DR_other_pc <- fmt_pc(PRM_DR_pivot[PRM_DR_pivot$drtype == "Other", "n"]/inh_4_a_c_e_g, digits = 2)

```

```{r PRM-KRM-other-co-occurrence, echo = F, warning=F}

PRM_table <- rm_norow(lapply(binary_tables, function(x){subset(x, !(is.na(PRM)))}))

# Pivot table of PRM-KRM only
PRM_KRM <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + KRM ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_KRM <- do.call("rbind", add_drug_col(PRM_KRM, drugs))
inh_PRM_KRM <- subset(PRM_KRM, Drug == "isoniazid" & !(is.na(KRM)))
inh_n_PRM_hap_KRM <- length(unique(inh_PRM_KRM$PRM))
inh_most_common_KRM_with_PRM <- names(which.max(table(inh_PRM_KRM$KRM)))

# PRM-other only
PRM_other <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + other_vars ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_other <- do.call("rbind", add_drug_col(PRM_other, drugs))
inh_PRM_other <- subset(PRM_other, Drug == "isoniazid" & !(is.na(other_vars)))
inh_n_PRM_hap_other <- length(unique(inh_PRM_other$PRM))
inh_most_common_other_with_PRM <- names(which.max(table(inh_PRM_other$other_vars)))

# Pivot table with 'other' - displayed later as PRM-KRM-other-table
PRM_KRM_other <- lapply(PRM_table, function(x) {
  reshape2::dcast(x, PRM + KRM + other_vars ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_KRM_other <- do.call("rbind", add_drug_col(PRM_KRM_other, drugs))
PRM_KRM_other <- dplyr::rename(PRM_KRM_other, unknown = other_vars)

# Get stats before clean up
PRM_n_samps_inh_co_occur <- sum(subset(PRM_KRM_other, !(is.na(KRM)))$n)
PRM_pc_samps_inh_co_occur <- fmt_pc(PRM_n_samps_inh_co_occur/inh_4_a_c_e_g)
PRM_n_fabg_15 <- sum(subset(PRM_KRM_other, KRM == "fabG1-c.-15C>T")$n)
PRM_pc_fabg_15 <- fmt_pc(PRM_n_fabg_15/inh_4_a_c_e_g)

# Clean up
PRM_KRM_other <- data.frame(apply(PRM_KRM_other, 2, function(x){
  ifelse(is.na(x), "-", x)
}))
PRM_KRM_other$PRM <- gsub("katG-p.|katG-c.", "", PRM_KRM_other$PRM)
names(PRM_KRM_other) <- c("Drug", "Possible resistance mutations (*katG*)", "Known resistance mutations", "Unknown", "n")

```

```{r isoniazid-Ser315Thr, echo = F, warning=F}

ser315thr_tab <- subset(isoniazid_binary_table, grepl("Ser315Thr", isoniazid_binary_table$KRM))
# Move ser315thr string to front e.g. fabG1-c.-15C>T; katG-p.Ser315Thr -> katG-p.Ser315Thr; fabG1-c.-15C>T
ser315thr_tab$KRM <- ifelse(grepl("; katG-p.Ser315Thr", ser315thr_tab$KRM), swap_str(ser315thr_tab$KRM), ser315thr_tab$KRM)
# Drop the specific change for the co-occurring gene - e.g. katG-p.Ser315Thr; fabG1-c.-15C>T ->  katG-p.Ser315Thr; fabG1
ser315thr_tab$KRM_2 <- drop_change(ser315thr_tab$KRM)
ser315thr_pivot <- pivot(ser315thr_tab, formula(KRM_2 ~ 'n'))
ser315thr_pivot <- dplyr::rename(ser315thr_pivot, mutation = KRM_2)

# ser315thr_pivot[grepl("; katG", ser315thr_pivot$mutation)]
 
# PRM_KRM_other[grepl("Ser315Thr", PRM_KRM_other$KRM), ]

```

```{r make-resistance-table-and-chisq, echo = F, warning=F}

# Resistance level

# Multiple KRMs	kat315	PRMs
# Yes	          5	      20
# No	          100	    3

# Get all samples with a Ser315Thr regardless of other KRMs being present (use grepl), with no PRMs
inh_kat315_multi_KRM_no_PRM <- subset(inh_bin, grepl("katG-p.Ser315Thr", KRM) & PRM_bin == "absent")
# Make a binary column - are there other KRMs or not? ("; " in KRM col)
inh_kat315_multi_KRM_no_PRM$Ser315Thr_bin <- ifelse(grepl("; ", inh_kat315_multi_KRM_no_PRM$KRM), "present", "absent")
# Pivot - other KRMs in Ser315Thr vs presence of Ser315Thr
inh_kat315_multi_KRM_no_PRM_pivot <- pivot(inh_kat315_multi_KRM_no_PRM, formula(Ser315Thr_bin ~ KRM_bin))

# Pivot - KRMs occurring with PRM (but not katG)
inh_4_e_g_table <- subset(inh_bin, PRM_bin == "present" & KRM_katG_bin == "absent")
inh_4_e_g_pivot <- pivot(inh_4_e_g_table, formula(KRM_bin ~ PRM_bin))

resistance_table <- merge(inh_kat315_multi_KRM_no_PRM_pivot, inh_4_e_g_pivot, 
                          by.x = "Ser315Thr_bin", by.y = "KRM_bin", sort = F)

names(resistance_table) <- c("KRM", "Ser315Thr", "PRM")

# Pull out % of Ser315Thr co-occurrences with KRMs
pc_315_KRM_co_occur <- str_extract(select(subset(resistance_table, KRM == "present"), Ser315Thr), "\\(.*\\)")
pc_315_KRM_co_occur <- gsub("\\(|\\)", "", pc_315_KRM_co_occur)

# Do Chi sq
res_tab <- cbind(table(inh_kat315_multi_KRM_no_PRM$Ser315Thr_bin, inh_kat315_multi_KRM_no_PRM$KRM_bin), table(inh_4_e_g_table$KRM_bin, inh_4_e_g_table$PRM_bin))
resistance_x_sq <- print_chisq(chisq.test(res_tab), rnd = 2)

```

```{r PRM-DST, echo = F, warning=F}

PRM_DST <- subset(PRM_tabs$isoniazid, !(is.na(dst)))
inh_n_samps_PRM_DST <- nrow(PRM_DST)
inh_n_samps_PRM_DST_1 <- nrow(subset(PRM_DST, dst == 1))
PRM_DST_p <- scientific(chisq.test(table(PRM_DST$dst))$p.value)
PRM_DST_1_no_KRM <- nrow(subset(PRM_DST, dst == 1 & is.na(KRM)))



```

The `r inh_n_uniq_PRM_katg` putative INH-katG resistance mutations occurred in multiple lineages (L1-L5) with many showing evidence of convergent evolution (**Figure \@ref(fig:isoniazid-tree)**). The putative resistance mutations occur in similar numbers of sub-lineages and at similar *katG* gene positions compared to known resistance mutations, indicating that they show comparable phylogenetic and gene location characteristics (**Figure \ref@(fig:fitness-pc-plot)A**). Due to the multi-drug regimens used for TB treatment, resistance often develops to multiple drugs in a stepwise manner [@Libiseller-Egger2020]. The co-occurrence of the `r inh_n_uniq_PRM_katg` *katG* mutations with other resistance mutations was analysed to characterise the isolate profiles in which putative resistance mutations occur. The `r inh_n_uniq_PRM_katg` *katG* mutations were most frequently found in isolates characterised as `r PRM_DR_max` (`r PRM_DR_max_pc`), but also common representation in pre-MDR-TB (`r PRM_DR_pre_MDR_pc`) and pre-XDR-TB (`r PRM_DR_pre_XDR_pc`). Interestingly, around half (`r PRM_n_samps_inh_co_occur`/`r inh_4_a_c_e_g`; `r PRM_pc_samps_inh_co_occur`) of isolates with any of the `r MOST_IMPORTANT_NUMBER` *katG* mutations, had co-occurrence with others linked to INH, with the *fabG1* -15C>T promoter mutation being the most frequent (`r PRM_n_fabg_15`/`r inh_4_a_c_e_g`; `r PRM_pc_fabg_15`) (**Table S4**). This observation is in stark contrast to the *katG* Ser315Thr mutation, the most prevalent resistance mutation in INH resistant isolates and known to confer a high level of resistance, which only co-occurs with other INH resistance mutations in `r pc_315_KRM_co_occur` of isolates. Of the `r inh_4_a_c_e_g` samples with a putative resistance mutation (**Figure S1**), `r inh_n_samps_PRM_DST` (`r fmt_pc(inh_n_samps_PRM_DST/inh_4_a_c_e_g)`) had an available DST result for INH with `r inh_n_samps_PRM_DST_1` reporting a resistant phenotype leading to a highly significant association of putative drug resistance mutation and DST (P<`r changeSciNot(PRM_DST_p)`). Of those with a resistant DST (n=`r inh_n_samps_PRM_DST_1`), `r PRM_DST_1_no_KRM` (`r fmt_pc(PRM_DST_1_no_KRM/inh_n_samps_PRM_DST_1)`) had no other known mutations that could explain resistance.

Isolates with mutations with a high level of drug resistance tend to have low numbers of co-occurring resistance mutations linked to that resistance. As a proxy for measuring resistance level, we measured the percentages of known and putative resistance mutation samples with co-occurring (non-*katG* - *fabG1*, *inhA*, *kasA*) mutations. Mutations at the *katG* 315 codon position mutation, which are known to confer high resistance [@Coll2018a], had a relatively low proportion of samples with co-occurring resistance mutations, with four out of the five known codon 315 mutations have ~20% or fewer proportion of samples with co-occurring resistance mutations (**Figure \ref@(fig:fitness-pc-plot)B**). There was no significant difference between the 40 known and `r MOST_IMPORTANT_NUMBER` putative *katG* resistance mutations in terms of proportions of co-occurrence of other resistance mutations (mean resistance co-occurrence proportion: known mutations 0.302 vs. putative 0.387; T-test P = 0.4).

*Mutation fitness*

```{r make-fitness-table, echo = F, warning=F}

# Fitness
# Ser315Thr is low fitness cost so should not have many CMs

# Compensatory	kat315	PRMs
# Yes	          5	      20
# No	          100	    3
# (exclude PRM-katG - katG-p.Ser315Thr pairs)

inh_kat315_no_PRM <- subset(inh_bin, KRM == "katG-p.Ser315Thr" & PRM_bin == "absent")
inh_kat315_no_PRM_pivot <- pivot(inh_kat315_no_PRM, formula(CM_bin ~ KRM))

inh_PRM_no_katG_KRM_pivot <- pivot(inh_4_e_g_table, formula(CM_bin ~ PRM_bin))

fitness_table <- merge(inh_kat315_no_PRM_pivot, inh_PRM_no_katG_KRM_pivot, 
                             by = "CM_bin", sort = F)

names(fitness_table) <- c("CM", "katG-p.Ser315Thr", "PRM")

# Do chi-sq test
x <- table(inh_kat315_no_PRM$CM_bin, inh_kat315_no_PRM$KRM)
y <- table(inh_4_e_g_table$CM_bin, inh_4_e_g_table$PRM_bin)
fitness_tab <- cbind(x, y)
fitness_x_sq <- print_chisq(chisq.test(fitness_tab), rnd = 2)

inh_n_samps_ser315_no_CM <- nrow(subset(inh_kat315_no_PRM, CM_bin == "absent"))
inh_n_samps_ser315_and_CM <- nrow(subset(inh_kat315_no_PRM, CM_bin == "present"))

```

```{r make-fitness-plot, echo = F, warning=F}

# KRM
# KRM <- subset(inh_bin, KRM_bin == "present")
# KRM <- data.frame(KRM %>% mutate(KRM = strsplit(as.character(KRM), "; ")) %>% unnest(KRM))
# KRM_tab <- pivot_num(KRM, formula(KRM ~ CM_bin))
# KRM_tab$KRM <- clean_del_ins_dup(KRM_tab$KRM)

# Just take Ser315Thr for now - all samples with Ser315Thr (but no PRM)
# KRM_tab <- subset(KRM_tab, grepl("Ser315Thr", KRM_tab$KRM))
# KRM_tab <- pivot_num(inh_kat315_no_PRM, formula(KRM ~ CM_bin))

KRM <- subset(inh_bin, KRM_bin == "present" & PRM_bin == "absent")
KRM <- data.frame(KRM %>% mutate(KRM = strsplit(as.character(KRM), "; ")) %>% unnest(KRM))
KRM_tab <- pivot_num(KRM, formula(KRM ~ CM_bin))
KRM_tab$KRM <- clean_del_ins_dup(KRM_tab$KRM)
KRM_tab <- subset(KRM_tab, grepl("katG", KRM_tab$KRM))
KRM_tab$status <- rep('known', nrow(KRM_tab))

# PRM
# PRM <- subset(inh_bin, PRM_bin == "present")
# Split out PRM haplotypes (one row per PRM, duplicating the ID and everything else in the row)
# PRM <- data.frame(PRM %>% mutate(PRM = strsplit(as.character(PRM), "; ")) %>% unnest(PRM))
# PRM_tab <- pivot_num(PRM, formula(PRM ~ CM_bin))

# Split out PRM haplotypes (one row per PRM, duplicating the ID and everything else in the row)
# Use PRM with no known katG
PRM <- data.frame(inh_4_e_g_table %>% 
                    mutate(PRM = strsplit(as.character(PRM), "; ")) 
                  %>% unnest(PRM))
PRM_tab <- pivot_num(PRM, formula(PRM ~ CM_bin))
# Clean up
PRM_tab$PRM <- clean_del_ins_dup(PRM_tab$PRM)
PRM_tab <- sort_pos(PRM_tab, "PRM")
PRM_tab$status <- rep('unknown', nrow(PRM_tab))
# Put together
KRM_PRM_tab <- rbind_mod(KRM_tab, PRM_tab, c("change", drop_col_name(KRM_tab, "KRM")))
KRM_PRM_tab$pc <- round((KRM_PRM_tab$present / (KRM_PRM_tab$absent + KRM_PRM_tab$present) )*100, 2)

PRM_med <- median(subset(KRM_PRM_tab, status == 'unknown')$pc)

# Plot each PRM % of samples with a CM with only 
fitness_plot <- ggplot() + geom_point(data = subset(KRM_PRM_tab, status == "unknown" | change == "katG-p.Ser315Thr"), 
                                      aes(x = factor(change, level = change), y = pc, 
                                          colour = status), size = 3)
fitness_plot <- fitness_plot + scale_x_discrete(name="") + 
  scale_y_continuous(name="% of samples with CM")
fitness_plot <- fitness_plot + theme_bw()
fitness_plot <- fitness_plot + theme(axis.text.x = element_text(angle=45, hjust = 1))
fitness_plot <- fitness_plot + theme(aspect.ratio=50/100)
fitness_plot <- fitness_plot + geom_hline(yintercept=PRM_med, linetype="dashed", color = "red")

# Plot number of samples vs % of samples with a CM

max_n <- max(KRM_PRM_tab$absent + KRM_PRM_tab$present)
# x_ax_A <- seq(0, max_n/10, 10)
x_ax_A <- seq(0, 10, 2)
x_ax_B <- seq(20, 100, 20)
x_ax_C <- seq(200, 1000, 200)
x_ax_D <- seq(2000, max_n, 2000)
x_ax <- c(x_ax_A, x_ax_B, x_ax_C, x_ax_D)

fitness_pc_total_plot <- ggplot() + geom_point(data = KRM_PRM_tab, 
                                      aes(x = absent + present, y = pc, 
                                          fill = status), 
                                      colour="black", pch=21,
                                      size = 3)
# fitness_pc_total_plot <- fitness_pc_total_plot + scale_x_continuous(name="n samples") + scale_y_continuous(name="% samples with compensatory mutations")
fitness_pc_total_plot <- fitness_pc_total_plot + scale_y_continuous(name="% samples with compensatory mutations")
fitness_pc_total_plot <- fitness_pc_total_plot + theme(aspect.ratio=50/100)
fitness_pc_total_plot <- fitness_pc_total_plot + geom_text(data=subset(KRM_PRM_tab, change == "katG-p.Ser315Thr"),
                                                           aes(x = absent + present, y = pc, label = "Ser315Thr"), 
                                                           hjust = 1.1)
fitness_pc_total_plot <- fitness_pc_total_plot + scale_x_log10(breaks = x_ax, name="n samples")
fitness_pc_total_plot <- fitness_pc_total_plot + theme_bw()
fitness_pc_total_plot <- fitness_pc_total_plot + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r make-KRM-PRM-CM-pc-tab, echo = F, warning=F}

# % co-occurence with another resistance mutation and % co-occurence with a CM

# All samples with a katG KRM and no PRM
KRM_katg <- subset(inh_bin, KRM_katG_bin == "present" & PRM_bin == "absent")

# Split out the katG KRMs
KRM_katg <- data.frame(KRM_katg %>% mutate(KRM_katG = strsplit(as.character(KRM_katG), "; ")) %>% unnest(KRM_katG))

# Remove katG KRM with less than 5 samples
KRM_katg <- data.frame(KRM_katg %>% group_by(KRM_katG) %>% filter(n() > 4))

# Make a binary column for presence of non-katG KRMs
KRM_katg$KRM_non_katG_bin <- ifelse(!(is.na(KRM_katg$KRM_fabG1)) | !(is.na(KRM_katg$KRM_inhA)) | !(is.na(KRM_katg$KRM_kasA)), "KRM_present", "KRM_absent")

# Modify CM column so that it's distinctive from just 'present' or 'absent', otherwise causes problems
KRM_katg$CM_bin <- ifelse(KRM_katg$CM_bin == "absent", "CM_absent", "CM_present")

# Pivot the katG KRMs against the non-katG presence/absence
katg_KRM_pivot <- pivot_num(KRM_katg, formula(KRM_katG ~ KRM_non_katG_bin))

# Pivot the katG KRMs against CM presence/absence
katg_CM_pivot <- pivot_num(KRM_katg, formula(KRM_katG ~ CM_bin))

# Merge together pivots
katg_KRM_CM_pivot <- merge(katg_KRM_pivot, katg_CM_pivot)

# Get percentages
katg_KRM_CM_pivot$KRM_pc <- round((katg_KRM_CM_pivot$KRM_present / (katg_KRM_CM_pivot$KRM_absent + katg_KRM_CM_pivot$KRM_present))*100, 2)
katg_KRM_CM_pivot$CM_pc <- round((katg_KRM_CM_pivot$CM_present / (katg_KRM_CM_pivot$CM_absent + katg_KRM_CM_pivot$CM_present))*100, 2)

# Annotate with 'known'
katg_KRM_CM_pivot$status <- rep("known", nrow(katg_KRM_CM_pivot))

# Clean up
katg_KRM_CM_pivot <- sort_pos(katg_KRM_CM_pivot, "KRM_katG") 


# Same for PRM

PRM <- subset(inh_bin, PRM_bin == "present" & KRM_katG_bin == "absent")
PRM <- data.frame(PRM %>% mutate(PRM = strsplit(as.character(PRM), "; ")) %>% unnest(PRM))


# Filter?
PRM <- data.frame(PRM %>% group_by(PRM) %>% filter(n() > 2))


PRM$KRM_non_katG_bin <- ifelse(!(is.na(PRM$KRM_fabG1)) | !(is.na(PRM$KRM_inhA)) | !(is.na(PRM$KRM_kasA)), "KRM_present", "KRM_absent")
PRM$CM_bin <- ifelse(PRM$CM_bin == "absent", "CM_absent", "CM_present")

PRM_KRM_pivot <- pivot_num(PRM, formula(PRM ~ KRM_non_katG_bin))
PRM_CM_pivot <- pivot_num(PRM, formula(PRM ~ CM_bin))

PRM_KRM_CM_pivot <- merge(PRM_KRM_pivot, PRM_CM_pivot)
PRM_KRM_CM_pivot$KRM_pc <- round((PRM_KRM_CM_pivot$KRM_present / (PRM_KRM_CM_pivot$KRM_absent + PRM_KRM_CM_pivot$KRM_present))*100, 2)
PRM_KRM_CM_pivot$CM_pc <- round((PRM_KRM_CM_pivot$CM_present / (PRM_KRM_CM_pivot$CM_absent + PRM_KRM_CM_pivot$CM_present))*100, 2)
PRM_KRM_CM_pivot$status <- rep("unknown", nrow(PRM_KRM_CM_pivot))
PRM_KRM_CM_pivot <- sort_pos(PRM_KRM_CM_pivot, "PRM")

# Put together
KRM_PRM_CM_pc_tab <- rbind_mod(katg_KRM_CM_pivot, PRM_KRM_CM_pivot, c("change", drop_col_name(katg_KRM_CM_pivot, "KRM_katG")))

# Clean up
KRM_PRM_CM_pc_tab$change <- clean_del_ins_dup(gsub("katG-p.|katG-c.", "", KRM_PRM_CM_pc_tab$change))
# KRM_PRM_CM_pc_tab <- dplyr::select(KRM_PRM_CM_pc_tab, change, KRM_pc, CM_pc, status)

KRM_ter <- quantile(KRM_PRM_CM_pc_tab$KRM_pc, probs = seq(0, 1, 0.33333))
CM_ter <- quantile(KRM_PRM_CM_pc_tab$CM_pc, probs = seq(0, 1, 0.33333))

# Classify resistance level based on Ser315Thr % of KRM (15%)
# KRM_PRM_CM_pc_tab$resistance_level <- ifelse(KRM_PRM_CM_pc_tab$KRM_pc < 15, "high", 
#                                                ifelse(KRM_PRM_CM_pc_tab$KRM_pc >= 15 & KRM_PRM_CM_pc_tab$KRM_pc <= 90, 
#                                                       "medium", "low"))

KRM_PRM_CM_pc_tab$resistance_level <- ifelse(KRM_PRM_CM_pc_tab$KRM_pc < 15, "high", 
                                               ifelse(KRM_PRM_CM_pc_tab$KRM_pc >= 15 & KRM_PRM_CM_pc_tab$KRM_pc <= KRM_ter[3], 
                                                      "medium", "low"))

KRM_PRM_CM_pc_tab$resistance_level <- factor(KRM_PRM_CM_pc_tab$resistance_level, levels = c("high", "medium", "low"))

# Classify fitness level based on Ser315Thr % of CM (2.71%)
# KRM_PRM_CM_pc_tab$fitness_cost_class <- ifelse(KRM_PRM_CM_pc_tab$CM_pc < 3, "low", 
#                                                ifelse(KRM_PRM_CM_pc_tab$CM_pc >= 3 & KRM_PRM_CM_pc_tab$CM_pc <= 90, 
#                                                       "medium", "high"))

KRM_PRM_CM_pc_tab$fitness_cost_class <- ifelse(KRM_PRM_CM_pc_tab$CM_pc < CM_ter[2], "low", 
                                               ifelse(KRM_PRM_CM_pc_tab$CM_pc >= CM_ter[2] & KRM_PRM_CM_pc_tab$CM_pc <= CM_ter[3], 
                                                      "medium", "high"))

KRM_PRM_CM_pc_tab$fitness_cost_class <- factor(KRM_PRM_CM_pc_tab$fitness_cost_class, levels = c("low", "medium", "high"))

# Arrange cols
KRM_PRM_CM_pc_tab <- select(KRM_PRM_CM_pc_tab, 
                            status, change, 
                            KRM_absent, KRM_present, KRM_pc,
                            CM_absent, CM_present, CM_pc, fitness_cost_class)

# Make duplicate before renaming and cleaning for display
KRM_PRM_CM_pc_tab_plot <- KRM_PRM_CM_pc_tab
KRM_PRM_CM_pc_tab_plot$n <- KRM_PRM_CM_pc_tab_plot$KRM_absent + KRM_PRM_CM_pc_tab_plot$KRM_present

# Build new table for display
KRM_PRM_CM_pc_tab <- data.frame(cbind(status = KRM_PRM_CM_pc_tab$status, 
      change = KRM_PRM_CM_pc_tab$change,
      KRM = paste0("[", KRM_PRM_CM_pc_tab$KRM_absent, "/",
                   KRM_PRM_CM_pc_tab$KRM_present, "/",
                   KRM_PRM_CM_pc_tab$KRM_pc, "]"), 
      CM = paste0("[", KRM_PRM_CM_pc_tab$CM_absent, "/",
                  KRM_PRM_CM_pc_tab$CM_present, "/",
                  KRM_PRM_CM_pc_tab$CM_pc, "]"), 
      fitness_cost_class = as.character(KRM_PRM_CM_pc_tab$fitness_cost_class)))

names(KRM_PRM_CM_pc_tab) <- c("Status", "Change", 
                              "n samples resistance mutations [absent/present/%]",
                              "n samples compensatory mutations [absent/present/%]",
                              "Fitness cost classification")

```

```{r make-KRM-PRM-CM-pc-plot, echo = F, warning=F}

# Plot data above using numeric table
KRM_PRM_CM_pc_tab_plot$KRM_present <- ifelse(KRM_PRM_CM_pc_tab_plot$KRM_present == 0, 0, log(KRM_PRM_CM_pc_tab_plot$KRM_present))
KRM_PRM_CM_pc_tab_plot$CM_present <- ifelse(KRM_PRM_CM_pc_tab_plot$CM_present == 0, 0, log(KRM_PRM_CM_pc_tab_plot$CM_present))

KRM_PRM_CM_pc_plot <- ggplot()+
  geom_point(data = KRM_PRM_CM_pc_tab_plot, aes(x = CM_pc, y = KRM_pc, fill = status, size = log(n)), colour="black", pch=21)+
  # geom_point(data = KRM_PRM_CM_pc_tab_plot, aes(x = CM_pc, y = KRM_pc, colour = status))+
  # geom_jitter(data = KRM_PRM_CM_pc_tab_plot, aes(x = KRM_pc, y = CM_pc, fill = status, size = log(n)), colour="black", pch=21, size=5)+
  geom_text(data=subset(KRM_PRM_CM_pc_tab_plot, change == "Ser315Thr"),
            aes(x = CM_pc, y = KRM_pc, label = "Ser315Thr"), 
            hjust = -0.2, size = 3)+
  ylab("% samples with known resistance mutation present (in fabG1, inhA or kasA)")+
  xlab("% samples with compensatory mutation present")+
  theme_bw() + geom_smooth(method=lm) 

ggsave("../results/KRM_PRM_CM_pc_plot.png", KRM_PRM_CM_pc_plot, width = 1100/5, height = 700/5, units = "mm")

max_log_n <- max(log(KRM_PRM_CM_pc_tab_plot$n))
sz_leg <- round_any(exp(c(2, 4, 6, max_log_n)), 10)
  
KRM_CM_PC_boxplot <- ggplot()+
  geom_boxplot(data = KRM_PRM_CM_pc_tab_plot, aes(x = fitness_cost_class, y = KRM_pc, fill = status),
               position=position_dodge(0.8), 
               alpha = 0.7)+
  geom_point(data = KRM_PRM_CM_pc_tab_plot,
           aes(x = fitness_cost_class, y = KRM_pc, fill = status, size = log(n)),
           colour="black", pch=21,
           position = position_dodge(0.8))+
  scale_size(name   = "n",
             breaks = log(sz_leg),
             labels = sz_leg)+
  geom_label_repel(
    data = subset(KRM_PRM_CM_pc_tab_plot, grepl("315", change)),
    aes(x = fitness_cost_class, y = KRM_pc, label = change),
    position = position_nudge(x = -0.2),
    size = 4,
    box.padding = unit(1, "lines"),
    point.padding = unit(0.1, "lines")
    )+
  xlab("Fitness cost")+
  ylab("% samples with co-occurring resistance mutations (fabG1, inhA, kasA)")+
  theme_bw()

KRM_CM_PC_boxplot_file <- "../results/KRM_CM_PC_boxplot.png"
ggsave(KRM_CM_PC_boxplot_file, KRM_CM_PC_boxplot, width = 1100/5, height = 700/5, units = "mm")


```

```{r KRM-CM-stats, echo = F, warning=F}



# t.test(KRM_pc ~ status, data = KRM_PRM_CM_pc_tab_plot)
# t.test(CM_pc ~ status, data = KRM_PRM_CM_pc_tab_plot)
# t.test(CM_pc ~ status, data = subset(KRM_PRM_CM_pc_tab_plot, !(change == "Ser315Thr")))
# t.test(CM_pc ~ status, data = subset(KRM_PRM_CM_pc_tab_plot, !(grepl("315", change))))


# Linear model values for relationship between fitness and resistance
# known_CM_KRM_lm <- lm(data = subset(KRM_PRM_CM_pc_tab_plot, status == "known"), KRM_pc ~ CM_pc)
# unknown_CM_KRM_lm <- lm(data = subset(KRM_PRM_CM_pc_tab_plot, status == "unknown"), KRM_pc ~ CM_pc)

KRM_CM_lm <- lm(data = KRM_PRM_CM_pc_tab_plot, KRM_pc ~ CM_pc + status)

# summary(lm(data = KRM_PRM_CM_pc_tab_plot, status ~ CM_pc + KRM_pc))

# summary(lm(data = iris, Petal.Length + Sepal.Length ~ Species - 1))

# vers <- subset(iris, Species == "versicolor")
# plot(vers$Sepal.Length, vers$Petal.Length)
# summary(lm(data = vers, Petal.Length ~ Sepal.Length))
# vers_vir <- subset(iris, Species %in% c("versicolor", "virginica"))
# 
# summary(lm(data = vers_vir, Petal.Length ~ Sepal.Length + Species))
# 
# ggplot(data = vers_vir, 
#        aes(x = Sepal.Length,
#            y = Petal.Length,
#            colour = Species)) + 
#   geom_smooth(method = "lm") + 
#   geom_point()+
#   geom_abline(aes(slope = 0.7248, 
#                   intercept = -0.0425))+
#   geom_abline(aes(slope = 0.7248, 
#                   intercept = -0.0425 + 0.8194))

# ggplot(data = KRM_PRM_CM_pc_tab_plot, 
#        aes(x = CM_pc,
#            y = KRM_pc,
#            colour = status)) + 
#   geom_smooth(method = lm) + 
#   geom_point() 

# summary(known_CM_KRM_lm)
# summary(unknown_CM_KRM_lm)
# 
# ggplot(data = KRM_PRM_CM_pc_tab_plot, 
#        aes(x = CM_pc,
#            y = KRM_pc,
#            colour = status)) + 
#   geom_smooth(method = lm) + 
#   geom_point()+
#   geom_abline(aes(slope = -0.131, 
#                   intercept = 32.786))+
#   geom_abline(aes(slope = -0.131, 
#                   intercept = 32.786 + 12.664))



# t.test(data = subset(KRM_PRM_CM_pc_tab_plot, fitness_cost_class == "high"), KRM_pc ~ status)
# t.test(data = subset(KRM_PRM_CM_pc_tab_plot, fitness_cost_class == "medium"), KRM_pc ~ status)
# t.test(data = subset(KRM_PRM_CM_pc_tab_plot, fitness_cost_class == "low"), KRM_pc ~ status)


```

Compensatory mutations are known to occur when mutations with high fitness costs occur (e.g. *katG* loss of function). To estimate the fitness impact of the `r inh_n_uniq_PRM_katg` putative resistance mutations we used the co-occurrence with a compensatory mutation as a proxy, and also included the Ser315Thr and other codon 315 mutations for comparison. The proxy fitness cost was discretised into 'low', 'medium' and 'high' categories (see Methods). The katG Ser315Thr is known to confer a low fitness cost [@Pym2002], was classified into our 'low' category, and has only ~3% of samples containing the mutation co-occurring with a compensatory mutation. 

In fact, mutations at the codon 315 position appear to have low fitness cost (**Figure \ref@(fig:fitness-pc-plot)B**), where four out of the five known codon 315 resistance mutations were classified into our 'low' fitness cost category.

Overall, compensatory mutations seem to occur in a higher proportion in isolates with our putative `r MOST_IMPORTANT_NUMBER` *katG* resistance mutations (`r inh_4_e`/`r nrow(inh_4_e_g_table)`) compared to Ser315Thr (`r inh_n_samps_ser315_and_CM`/`r fmt(nrow(inh_kat315_no_PRM))`) (Chisq P < 10^-16^), suggesting that on average they incur a greater fitness cost compared to this high frequency global mutation. Similarly, there was a higher proportion of isolates with a compensatory mutation in those with the `r MOST_IMPORTANT_NUMBER` putative mutations compared to those with the 40 known mutations (proportion with compensation mutation: known 0.196 vs. putative 0.515; T-test P = 3×10^-5^). This difference remained statistically significant even when excluding the 315 codon positions (P = 2×10^-4^).




There appears to be little overall relationship between known resistance mutation co-occurrence (resistance level) and fitness cost in both known and unknown resistance mutations (p = `r lmp(KRM_CM_lm)`. Nor were there significant differences between known and putative resistance mutations in each of our three fitness cost categories of low, medium and high (minimum p = 0.4; **Figure S4**). 

That there were no differences in the co-variation between resistance level and fitness cost between known and putative variants lends evidence as to the veracity of our putative resistance mutations. 

```{r protein-structure, echo = F, warning=F}

system("../shell_scripts/dist_and_ddg_parse.sh")

distances_file <- paste0(results_path, "distances.txt")
distances <- read.delim(distances_file)

dist_t <- t.test(dist ~ status, data = distances)

fitness_est_file <- paste0(results_path, "ddg.txt")
fitness_est <- read.delim(fitness_est_file)

fitness_t <- t.test(ddg ~ status, data = fitness_est)


```

*Protein structure modelling*

To explore the functional effects of the `r inh_n_uniq_PRM_katg` putative resistance mutations, *in silico* predictions of effects on the *katG* target protein were assessed. 

The distances from the *katG* heme-binding site, thought to be close to the active INH binding site and crucial to enzymatic activity [@Munir2021], did not differ significantly between the known and putative resistance mutations (mean known = `r dist_t$estimate[1]` A^o^; mean putative = `r dist_t$estimate[2]` A^o^; p = `r round(dist_t$p.value, 2)`).

To infer fitness cost in the putative resistance mutations, the protein stability change induced by mutations were estimated as the change in Gibbs free energy ($\Delta\Delta$G) using mCSM. There was however, no significant difference between the known resistance mutations and the putative resistance mutations (mean known = `r fitness_t$estimate[1]`; mean putative = `r fitness_t$estimate[2]` p = `r round(fitness_t$p.value, 2)`). 

## DISCUSSION

Our goal was to identify putative novel mutations underlying resistance to RIF and INH, by identifying isolates with established compensatory mutations. No novel *rpoB* gene mutations potentially linked to resistance to RIF were found, but not surprisingly since there are few ways in which the precise machinery of RNA polymerase can change without loss of function. In contrast, *katG* can change in many ways while the bacteria is largely unaffected. The 'rare' mutations identified in one or two isolates were not analysed further, but could be monitored. 

In lieu of DST testing, we have provided evidence that our *katG* putative resistance mutations are causally resistance-conferring. Our mutations occur in multiple sublineages, indicating convergence independent of phylogeny, a pattern which is well established in the known *katG* resistance mutations.
	
No significant differences were found in the percentage of samples with (non-*katG*) known resistance mutation co-occurrence (our proxy for resistance level) between the known resistance mutations and our putative resistance mutations. In showing a similar pattern of resistance mutation co-occurrence we infer that our putative resistance mutations confer a similar level of resistance as known mutations on average and that our putative resistance mutations are likely to be genuine. 
			
There was however a difference in the fitness cost between the known and putative resistance mutations, with the putative mutations appearing to have on average a higher cost. That the known and putative mutations diverge in respect of fitness cost is not necessarily indicative that our mutations are less likely to be genuine. Known resistance mutations are likely to converge on the most stable protein configurations and hence proliferate. Novel resistance mutations will therefore be more likely to occur in positions which are less stable. 
		
There was no relationship between resistance level and fitness cost in either the known resistance mutations or putative resistance mutations. Again, this similar pattern of variation indicates the veracity of the putative resistance mutations. 

Further evidence for our mutations comes from *in silico* protein modelling. The putative resistance mutations showed no significant difference to known mutations in terms of their distances from the *katG* heme active binding site, indicating that overall they likely confer a similar pattern of resistance. However, in contrast to the differences between known and putative mutations in their percentages of samples with compensatory mutations, the was no difference in the *in silico* prediction of known and putative protein stability. This result is perhaps surprising given the significant difference between known and putative mutations in terms of the percentages of samples with compensatory mutations. However, the $\Delta\Delta$G measure is only a prediction of stability, and therefore an indirect indication of fitness cost.

## CONCLUSIONS

We have presented a method of quickly finding potential resistance mutations in order to monitor development of resistance mechanisms to two of the most important anti-TB drugs.
Our list of potential resistance mutations ought to be verified as causing resistance and added to the database of known resistance mutations.

## METHODS

*Sequence data and processing*

The database of 32k isolates with DST and sequence data has been described previously [@Napier2020]. Sequences were aligned to the H37Rv reference genome (AL123456) with BWA mem (v0.7.17) [@Li2013], [@Cole1998]. Joint SNP and indel calling was carried out in gatk GenotypeGVCFs (v4.1.3.0) - –java-options “-Xmx40g” [@Depristo2011]. Monomorphic SNPs/indels and those in non-unique regions of the genome (e.g., ppe genes) were excluded. Multi-FASTA formatted filed were created from the filtered SNP/indel file and H37Rv reference fasta with bedtools makewindows (v2.28.0); parameters -w 50000 -s 50000 [@Quinlan2010]. Phylogenetic trees were constructed using IQ-TREE (v1.6.12), involving a general time reversible model with rate heterogeneity set to a discrete Gamma model and an ascertainment bias correction (parameters −m GTR+G+ASC), with 1000 bootstrap samples [@Nguyen2015]. Drug resistance types and lineages were predicted *in silico* with TB-profiler (v4.2.0) [@Phelan2019] [@Coll2015]. TB-profiler software [@Phelan2019] was used to identify all known drug-resistance, compensatory and putative novel resistance mutations. 

*Finding putative resistance markers using compensatory mutations*

From the database of TB-Profiler (n=32k), all non-synonymous mutations in the relevant compensatory genes were found (*ahpC* for INH, *rpoC*/*rpoA* for RIF). All mutations in compensatory genes were filtered such that they were present in three or more samples, <50% samples were predicted sensitive by TB-Profiler and <50% of samples were DST sensitive, for each relevant drug. As there were many potential *ahpC* mutations, further filtering criteria were applied to these - mutations were retained if they were associated with loss of function in *katG* mutations, occurred in the same position as known *ahpC* mutations and if they appeared in multiple lineages. Only one of these criteria needed to be met to be considered a potential compensatory *ahpC* mutations. The full list of compensatory mutations consisted of `r inh_n_uniq_PRM_katg` mutations (**Table S1**). To find putative resistance mutations, all non-synonymous mutations in the relevant resistance genes were extracted from the TB-Profiler database (*katG* for INH, *rpoB* for RIF). Some variants known not to be associated with INH resistance [@Doorn2001] (e.g., *katG*-Arg463Leu) were excluded.

For each drug, mutations were found in samples where a compensatory mutation was present and no known resistance mutations were present in the relevant genes, but a non-resistance-associated mutation was present in the relevant target genes. These mutations were then filtered to exclude known drug-resistance-associated variants, and subjected to the same criteria as the putative compensatory mutations - i.e., present in three or more samples, <50% samples were predicted *Sensitive* by TB-profiler and <50% of samples were DST sensitive. Mutations occurring in promoter regions of *rpoB* were excluded as candidate potential resistance mutations as there is no known mechanism of resistance that could result from increased expression of the RNA polymerase beta subunit.

Using this list of potential resistance mutations, all TB-profiler samples were then searched for their presence, regardless of compensatory mutation status. It should be noted that therefore not all samples with potential resistance mutations necessarily have compensatory mutations, and vice versa. 

*Protein structural modelling*

The open source software ChimeraX [@Pettersen2021] was used to model distances from the INH heme binding site and infer change in Gibbs free energy ($\Delta\Delta$G) *in silico* change in Gibbs free energy (G) of mutations was inferred using mCSM [@Pires2014].

## List of abbreviations

* *Mtb* = *Mycobacterium tuberculosis*
* TB = tuberculosis
* INH = isoniazid
* RIF = rifampicin
* WGS = whole genome sequencing
* CM = compensatory mutation
* PRM = putative resistance mutation

## Declarations

## Ethics approval and consent to participate
Not applicable

## Consent for publication
Not applicable 

## Competing interests
The authors declare that they have no competing interests

## Funding
GN is funded by an BBSRC-LIDo PhD studentship. JEP is funded by a Newton Institutional Links Grant (British Council, no. 261868591). TGC is funded by the Medical Research Council UK (Grant no. MR/M01360X/1, MR/N010469/1, MR/R025576/1, and MR/R020973/1). SC is funded by Medical Research Council UK grants (ref. MR/M01360X/1, MR/R025576/1, and MR/R020973/1). The authors declare no conflicts of interest. The funders had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript should be declared.

## Authors' contributions
JEP and TGC conceived and directed the project. GN performed bioinformatic and statistical analyses under the supervision of SC, JEP and TGC. GN, SC, JEP and TGC interpreted results. GN wrote the first draft of the manuscript with inputs from JEP and TGC. All authors commented and edited on various versions of the draft manuscript and approved the final version. GN, SC, JEP, and TGC compiled the final manuscript.

## Acknowledgements

## TABLES

<!-- TABLE 1 --> 
```{r table-1-display, echo = F, warning=F}

cap <- sprintf("*Mycobacterium tuberculosis* isolates (n=%s)", fmt(total_samps))
knitr::kable(table_1, caption = cap, format = "pipe", row.names = F)

```
MDR-TB = multi-drug resistant; XDR-TB = extensively drug resistant


<!-- TABLE 2 -->
```{r PRM-summary-table, echo = F, warning=F}

cap <- sprintf("List of potential resistance mutations for isoniazid in the *katG* gene (n=%s). Underlined present in only a single lineage.", inh_4_a_c_e_g)
knitr::kable(PRM_summary, row.names = F, caption = cap, format = "pipe")

```

## FIGURES

<!-- Figure 1 -->
```{r isoniazid-tree, echo=FALSE, out.width = "2000px", fig.cap=sprintf("Phylogenetic tree of (n=%s) isolates with %s putative novel *katG* gene mutations for Isoniazid resistance", inh_4_a_c_e_g, inh_n_uniq_PRM_katg)}

# source("comp_mut_tree.R")
knitr::include_graphics(tree_files[grep("isoniazid", tree_files)])

```

<!-- Figure 2 -->
```{r fitness-pc-plot, echo = F, warning=F, fig.cap="Number of samples vs percentage of samples with a compensatory mutation for known *katG* mutations and potential *katG* resistance mutations. *katG*-Ser315Thr is labeled for comparison. This mutation incurs low fitness cost, hence does not generally co-occur in samples with a compensatory mutation. Samples with *katG*-Ser315Thr and a potential resistance mutation have been excluded. Samples with a potential resistance mutation and a known *katG* resistance mutation have been excluded."}

fitness_pc_total_plot_file <- "../results/fitness_pc_total_plot.png"
ggsave(fitness_pc_total_plot, file = fitness_pc_total_plot_file, width = 1100/4, height = 700/4, units = "mm")
knitr::include_graphics(fitness_pc_total_plot_file)


```

## SUPPLEMENTARY TABLES

**Table S1**

```{r common-KRM-tab, echo = F, warning=F}

cap <- "Most common known resistance mutations for INH and RIF in TBprofiler. Mutations occurring fewer than five times are omitted."

knitr::kable(common_KRM_tab, row.names = F, caption = cap, format = "pipe")

```

**Table S2**

```{r CM-table-display, echo = F, warning=F}

cap <- "Mutations in compensatory genes and frequencies in 32k samples. Underlined if known compensatory mutations. Bolded if occurring across >1 lineage"
knitr::kable(CM_tab, caption = cap)

```
Ins = insertion, del = deletion * novel markers with strong evidence for compensatory effects through convergent evolution, co-occurrence with loss of function mutations in katG as well as association with INH resistant isolates.

**Table S3**

```{r rare-mutations-table, echo = F, warning=F}

cap <- sprintf("Mutations (n=%s) in *katG* (INH) and from samples with no known resistance mutations, with compensatory mutations but no potential resistance mutation by the criteria outlined in Methods. These mutations only occur in one or two samples among the %s samples from Supplementary figure 1", inh_n_unique_rare_katg, inh_5_a)
knitr::kable(rare_mutations_table, row.names = F, caption = cap, format = "pipe")

```



```{r PRM-KRM-other-table, echo = F, warning=F}

cap <- "Known resistance mutations and non-drug resistance-associated (‘unknown’) mutations co-occurring in samples with putative drug resistance"
knitr::kable(PRM_KRM_other, row.names = F, caption = cap, format = "pipe")

```

**Table S4**

```{r KRM-PRM-CM-pc-tab, echo = F, warning=F}

# % co-occurence with another resistance mutation and % co-occurence with a CM

cap <- "Percentages of co-occurring known resistance mutations and compensatory mutations among known and putative ('unknown') resistance mutations. Mutations in fewer than three samples are omitted. Co-occurring known resistance mutations are in genes *fabG1*, *inhA*, *kasA*. The three categories of high medium and low resistance/fitness are based on tertiles of the respective percentages, except for the 'high' category of resistance level, which is based on the ~15% Ser315Thr resistance mutation co-occurrence. See **Figure S4**"

knitr::kable(KRM_PRM_CM_pc_tab, row.names = F, caption = cap, format = "pipe")

```

**Table S5**



## SUPPLEMENTARY FIGURES

**Figure S1**

```{r samples-trees, echo=FALSE, out.width = "2000px", fig.show='hold', fig.cap="Analysis strategy and numbers of mutations" }

knitr::include_graphics("../results/samples_trees.png")
# knitr::include_graphics(rif_numbers_file)

```

**Figure S2**

```{r inh-ahpc-katg-links, echo=FALSE, warning = F, out.width = "2000px", fig.cap="Co-occurrences between *ahpC* compensatory mutation positions and *katG* putative resistance mutation positions for isoniazid"}

binary_expand_PRM <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(PRM = strsplit(as.character(PRM), "; ")) %>%
  unnest(PRM))
})

inh <- binary_expand_PRM$isoniazid
inh_PRM <- subset(inh, !(is.na(PRM)))
inh_PRM <- select(inh_PRM, CM, PRM)
# Split and process del/dup/ins separately 
del_dup_ins <- inh_PRM[grepl("del|dup|ins", inh_PRM$PRM), ]
inh_PRM <- inh_PRM[!(grepl("del|dup|ins", inh_PRM$PRM)), ]
inh_PRM_pos <- data.frame(apply(inh_PRM, 2, find_pos))
del_dup_ins$CM <- find_pos(del_dup_ins$CM)
del_dup_ins$PRM <- gsub("[^0-9_del|dup|ins]", "", del_dup_ins$PRM)
inh_PRM_pos <- rbind(inh_PRM_pos, del_dup_ins)

colnames(inh_PRM_pos) <- c("ahpC", "katG")

df <- inh_PRM_pos %>% make_long(ahpC, katG)
# Remove NAs from first node col - removes ahpCs that do not linek to a katG
df <- subset(df, !(is.na(node)))

# Move the node col to two separate cols according to the gene (i.e. group) so that the text can be plotted separately 
# - each row of text can be adjusted separately 
df$katG <- ifelse(df$x == 'katG', df$node, NA)
df$ahpC <- ifelse(df$x == 'ahpC', df$node, NA)

detach("package:plyr", unload = TRUE)

sank <- ggplot(df, aes(x = x, next_x = next_x, node = node, next_node = next_node))
sank <- sank + coord_flip()
sank <- sank + geom_sankey(flow.alpha = .6, fill = 'lightblue', na.rm = T)
sank <- sank + geom_sankey_text(aes(label = katG), size = 3, angle = 45, width = 2, na.rm = T, vjust = -2, hjust = 0)
sank <- sank + geom_sankey_text(aes(label = ahpC), size = 3, angle = 48, width = 2, na.rm = T, vjust = 2, hjust = 1.5)
sank <- sank + theme_minimal()
sank <- sank + labs(x = NULL)
sank <- sank + theme(legend.position = "none",
        plot.title = element_text(hjust = .5),
        axis.text.x=element_blank(), 
        panel.grid = element_blank())
sank <- sank + ggtitle("INH")

sank_file <- "../results/ahpc_katg_links.png"

ggsave(sank, file = sank_file, width = 1100/4, height = 700/4, units = "mm")

knitr::include_graphics(sank_file)

```

**Figure S3**

```{r n-known-katg-homoplasy-plot, echo=FALSE, warning = F}

# Get stupid number of KRMs in > 2 samps - see homoplasy_plot.R
known_data <- select(inh_bin, wgs_id, sublin, KRM_katG)
# Split out
known_data <- data.frame(known_data %>% mutate(KRM_katG = strsplit(as.character(KRM_katG), "; ")) %>% unnest(KRM_katG))
# Remove NA
known_data <- subset(known_data, !(is.na(KRM_katG)))
known_n <- reshape2::dcast(known_data, KRM_katG ~ 'n', value.var = "KRM_katG", fun.aggregate = length)
known_sublin_n <- reshape2::dcast(reshape2::dcast(known_data, KRM_katG + sublin ~ 'n', value.var = "KRM_katG", fun.aggregate = length), KRM_katG ~ 'n_sublin', value.var = "KRM_katG", fun.aggregate = length)
known_pivot <- merge(known_n, known_sublin_n, by = "KRM_katG", all.x = T, sort = F)
known_pivot$pos <- find_pos(known_pivot$KRM_katG)
known_pivot$pos <- as.numeric(known_pivot$pos)
known_pivot <- subset(known_pivot, n>2)
known_pivot <- subset(known_pivot, pos > 0)
known_pivot <- subset(known_pivot, !(grepl("c.", KRM_katG)))
inh_n_know_katg_filt <- length(unique(known_pivot$KRM_katG))

```

```{r homoplasy-plot, echo=FALSE, warning = F, out.width = "2000px", fig.cap=sprintf("Homoplasy among known (n=%s) and putative (n=%s) resistance mutations in the *katG* gene, as indicated by number of sub-lineages in which mutations occur. The common *katG* Ser315Thr mutation is highlighted. Mutations in fewer than three samples and non-protein coding regions are omitted.", inh_n_know_katg_filt, MOST_IMPORTANT_NUMBER)}

# source("homoplasy_plot.R")
knitr::include_graphics("../results/homoplasy_plot.png")

```

**Figure S4**

```{r KRM-CM-pc-plot, echo=FALSE, warning = F, out.width = "2000px", fig.cap="Fitness cost versus percentage of samples with co-occurring known (non-*katG*) resistance mutations in known and putative *katG* resistance mutations. The three categories of high medium and low fitness cost are based on tertiles of the percentage of samples with compensatory mutations (low:<=17%, medium:>17%<=40%, high:>40%<=100%). Known *katG* resistance mutations at the 315 position are highlighted. Ser315Thr in particular is known to confer a high level of resistance to INH, hence its samples co-occur with relatively few other (non-*katG*) resistance mutations (~15%). Ser315Thr also incurs little fitness cost, therefore a low percentage of its samples also have compensatory mutations (~3%). Mutations occurring in fewer than three samples are omitted. See **Table S4**"}

# knitr::include_graphics("../results/KRM_PRM_CM_pc_plot.png")

knitr::include_graphics(KRM_CM_PC_boxplot_file)

```

**Figure S5**

```{r protein-model, echo=FALSE, warning = F, out.width = "2000px", fig.cap="Protein structural model of KatG (7ag8) with known (cyan) and putative (red) mutations highlighted on chain A (blue). The vast majority of putative mutations cluster around the bound heme (green)."}

knitr::include_graphics(protein_model_file)

```

### References
































