---
title: "Large-scale genomic analysis of *Mycobacterium tuberculosis* reveals extent of target and compensatory mutations linked to isoniazid, rifampicin and multi-drug resistance"
output: 
  bookdown::word_document2: 
    fig_caption: yes
    table_caption: yes
    number_sections: no
mainfont: Calibri
fontsize: 10pt
bibliography: all_refs.bib
csl: biomed-central.csl
---

```{r setup, include=FALSE, echo = F, warning=F, message=F}

# SETUP ----

rm(list = ls())

knitr::opts_chunk$set(echo = F)
# options(scipen=1, digits=2)
options(scipen=999, digits = 3)
table_font_sz <- 8

```

```{r library, include=FALSE, echo = F, warning=F, message=F}

library(tidyr)
library(plyr)
library(dplyr)
library(knitr)
library(reshape2)
library(janitor)
library(ggsankey)
library(ggplot2)

```

```{r functions, echo = F, warning=F, message=F}

source("https://raw.githubusercontent.com/GaryNapier/Packages_functions/master/Functions.R")

uniq_col <- function(x, split = "; "){
  unique(unlist(strsplit(x, split)))
}

col_pc <- function(df, col){
  round((df[, col] / sum(df[, col]))*100, 3)
}

len_uniq <- function(x){
  length(unique(x))
}

rbind_force <- function(df1, df2){
  rbind(df1, setNames(df2, names(df1)))
}

# Make quick pivot table summarising total unique values by column
# my_formula arg must be call to the formula() function: formula(<col> ~ 'n')
# Result:
#             PRM            n
#  katG-c.2223A>G  3   (4.05%)
# katG-p.Arg484His  5   (6.76%)
# ...
# katG-p.Tyr413Cys  8  (10.81%)
#           Total 74 (100.00%)
pivot <- function(x, my_formula, value_var = "wgs_id"){
  drop_cols(to_table(reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq),
                     pc_dir = 'col'), 
            'Total')
}

split_pivot <- function(df, my_formula, leave_col, value_var = "wgs_id", split_on_col = "Drug"){
  
  to_table_split_pivot <- function(x, col){
    x[[col]] <- as.character(x[[col]])
    x <- drop_cols(to_table(x, "col"), "Total")
    x[[col]] <- gsub("\\s*\\([^\\)]+\\)", "", x[[col]])
    x
  }
  
  df_pivot <- reshape2::dcast(df, my_formula, value.var = 'wgs_id', fun.aggregate = len_uniq)
  df_split <- split(df_pivot, df_pivot[, split_on_col])
  
  do.call('rbind', lapply(df_split, to_table_split_pivot, leave_col))
  
}

duped <- function(x, col){
  dups <- x[duplicated(x[col]), col]
  x[x[, col] %in% dups, ]
}

# Swap first and second parts of string, splitting on a charachter (or substring)
# e.g. 
# x <- "katG-p.Ile335Val; katG-p.Ser315Thr"
# swap_str(x, "; ")
# "katG-p.Ser315Thr; katG-p.Ile335Val"
swap_str <- function(x, split_chr = "; "){
  paste0(unlist(strsplit(x, split_chr))[c(2, 1)], collapse = split_chr)
}


# Print each element of vector vertically instead of across the screen
# e.g. 
# > x <- c("a", "b", "c")
# > x
# [1] "a" "b" "c"
# > print_vert(x)
# a
# b
# c
print_vert <- function(x){
  cat(paste0(x, '\n'))
}

# Remove change from second gene-change pair
# e.g.
# x <- "katG-p.Ser315Thr; fabG1-c.-15C>T"
# > drop_change(x, "-")
# [1] "katG-p.Ser315Thr; fabG1"
drop_change <- function(string, split_on = "-"){
  unlist(lapply(strsplit(string, split_on), function(x){paste0(x[c(1, 2)], collapse = split_on)}))
}

clean_binary_table <- function(x){
  x <- data.frame(apply(x, 2, function(x){gsub("\\[\\]", NA, as.character(x))}))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\[||\\]", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\), \\(", "; ", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\', \\'", "-", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\(||\\)", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\'", "", as.character(x)) }))
  x
}

sort_pos <- function(x, col){
  x$pos <- as.numeric(stringr::str_extract(x[, col], "[0-9]+"))
  x <- x[order(x$pos), ]
  drop_cols(x, 'pos')
}

lapply_nms <- function(x, fun, nms){
  res <- lapply(x, fun)
  names(res) <- nms
  res
}

drug_col_front <- function(df_list){
  lapply(df_list, function(x){ select(x, Drug, everything()) })
}

add_drug_col <- function(df_list, drugs){
  drug_col_front(lapply_mod(df_list, function(i){df_list[[i]]$Drug <- rep(drugs[i], nrow(df_list[[i]])); df_list[[i]]}))
}

find_pos <- function(x){
  sub("\\-", "", gsub("[^0-9\\-]+", "", x))
}

fmt_pc <- function(x){
  paste0(fmt(x*100), "%")
}

trimws_df <- function(x){
  data.frame(sapply(x, trimws))
}

tonum <- function(x){
  as.numeric(gsub(",", "", strsplit(x, " ")[[1]][1]))
}

uniq_vars <- function(x){
  unique(unlist(strsplit(x, "; ")))
}

rm_null <- function(x){
  x[!sapply(x,is.null)] 
}

rm_norow <- function(x){
  x[sapply(x, nrow) > 0]
}

rm_na <- function(x){
  x[!is.na(x)]
}

```

```{r variables, echo = F, warning=F, message=F}


# VARIABLES ----

id_col <- "wgs_id"
rnd <- 3
# drugs <- c("rifampicin", "isoniazid", "ethambutol", "pyrazinamide", "streptomycin", 
#              "ofloxacin", "moxifloxacin", "levofloxacin", "amikacin", "kanamycin", "capreomycin", "ciprofloxacin", "prothionamide", 
#              "ethionamide", "para_aminosalicylic_acid", "cycloserine",
#              "clarithromycin", "clofazimine", "bedaquiline", "linezolid", "rifabutin", "delamanid")
drugs <- c("isoniazid", "rifampicin")
fq <- "fluoroquinolones, ofloxacin, levofloxacin, ciprofloxacin, moxifloxacin"
agc <- "capreomycin, streptomycin, amikacin, kanamycin, aminoglycosides"
align <- c("l", "r", "r")
blue_col <- "cornflowerblue"
plot_text_sz <- 0.5
main_cex <- 1.5
# dr_status_vect <- c("Sensitive", "Pre-MDR", "MDR", "Pre-XDR", "XDR", "Other")
dr_status_vect <- c("Sensitive", "Pre-MDR-TB", "HR-TB", "RR-TB", "MDR-TB", "Pre-XDR-TB", "XDR-TB", "Other")

```

```{r paths, echo = F, warning=F, message=F}

# PATHS ----

setwd("~/Documents/comp_mut/r_scripts/")

metadata_path <- "../../metadata/"
local_metadata_path <- "../metadata/"
pipeline_db_path <- "../../pipeline/db/"
db_path <- "../../tbdb/"
results_path <- "../results/"
newick_path <- paste0(results_path, "newick/")

```

```{r files, echo = F, warning=F, message=F}

# FILES ----

# Metadata
metadata_file <- paste0(metadata_path, "tb_data_18_02_2021.csv")
# Database data
lineage_conversion_file <- paste0(pipeline_db_path, "lineage_conversions.txt")
dr_genes_file <- paste0(db_path, "tbdb.csv")
KCM_file <- paste0(pipeline_db_path, "compensatory_mutations.csv")
CM_RM_gene_assoc_file <- paste0(local_metadata_path, "CM_RM_gene_assoc.csv")

```

```{r drug-files, echo = F, warning=F, message=F}

binary_table_files <- paste0(results_path, drugs, "_binary_table.csv")
PCM_merged_files <- paste0(results_path,  drugs, "_PCM_merged.csv")
summary_files <- paste0(results_path, drugs, "_summary.csv")
PRM_stats_files <- paste0(results_path, drugs, "_PRM_stats.csv")
tree_files <- paste0(results_path, drugs, "_tree.png")
inh_numbers_file <- paste0(results_path, "isoniazid_numbers.png")
rif_numbers_file <- paste0(results_path, "rifampicin_numbers.png")

```

```{r read-in-data, echo = F, warning=F, message=F}

# READ IN DATA ----

# Metadata
metadata <- read.csv(metadata_file)
# Database data
lineage_conversions <- read.delim(lineage_conversion_file)
# dr_genes <- read.delim(dr_genes_file, header = F)
dr_genes <- read.csv(dr_genes_file)
KCM <- read.csv(KCM_file)
CM_RM_gene_assoc <- read.csv(CM_RM_gene_assoc_file)

```

```{r drugs-read-in-data, echo = F, warning=F, message=F}

binary_tables <- lapply_nms(binary_table_files, read.csv, drugs)
PCM_merged <- lapply_nms(PCM_merged_files, read.csv, drugs)
summaries <- lapply_nms(summary_files, read.csv, drugs)
PRM_stats <- lapply_nms(PRM_stats_files, read.csv, drugs)

```

```{r clean-data, echo = F, warning=F, message=F}

# Binary tables 

# Remove all the "[]" present in the binary table caused by python's lists
binary_tables <- lapply(binary_tables, clean_binary_table)

# Clean lineage column 
binary_tables <- lapply(binary_tables, function(x) {x$main_lineage <- gsub("lineage", "", x$main_lineage); x})
binary_tables <- lapply(binary_tables, function(x) {x$sublin <- gsub("lineage", "", x$sublin); x})

# Trim white space
binary_tables <- lapply(binary_tables, trimws_df)

# Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
binary_tables_dup <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>%
  unnest(other_vars))
})

# Make col for just the other_vars genes, stripping out the change
binary_tables_dup <- lapply(binary_tables_dup, function(x){x$other_vars_gene <- gsub("-.*", "", x$other_vars); x})

# Rifampicin filter - remove lineage-specific PRMS
# NB - TOTAL HACK TO REMOVE THE LINEAGE-SPECIFIC RIF PRMs
rif_lin_PRM <- c("rpoB-p.Val695Leu", "rpoB-c.-199C>T", "rpoB-p.Ile925Val")
binary_tables$rifampicin$PRM <- ifelse(binary_tables$rifampicin$PRM %in% rif_lin_PRM,
       NA,
       binary_tables$rifampicin$PRM)

# Make binary cols for CM KRM and PRM
binary_tables <- lapply(binary_tables, function(x){
  data.frame(cbind(x, 
                   CM_bin = ifelse(is.na(x$CM), "absent", "present"), 
                   KRM_bin = ifelse(is.na(x$KRM), "absent", "present"), 
                   PRM_bin = ifelse(is.na(x$PRM), "absent", "present")))
})

# Make katG binary
binary_tables$isoniazid$KRM_katG_bin <- ifelse(is.na(binary_tables$isoniazid$KRM_katG), "absent", "present")

# KCM - combine gene-change
KCM$gene_change <- paste0(KCM$Gene, "-", KCM$Mutation)

# Subset KCM 
inh_KCM <- KCM[KCM[,"Drug"] == "isoniazid", ]
rif_KCM <- KCM[KCM[,"Drug"] == "rifampicin", ]


# Put PCM_merged together and join in the drug names
PCM_merged <- do.call("rbind", PCM_merged)
PCM_merged$gene_change <- paste0(PCM_merged$gene, "-", PCM_merged$change)

PCM_merged <- merge(PCM_merged, CM_RM_gene_assoc, 
                    by.x = "gene", by.y = "CM_gene", 
                    sort = F)

# Remove the PCMs in the KCM
PCM_merged <- PCM_merged[!(PCM_merged$gene_change %in% KCM$gene_change), ]

# Transpose summary
summaries <- do.call('cbind', lapply(summaries, function(x){data.frame(n = t(x))}))
names(summaries) <- drugs

# PRM stats
PRM_stats <- lapply(PRM_stats, sort_pos, 'mutation')
# Clean RIF of the lineage-specific PRMs
PRM_stats$rifampicin <- PRM_stats$rifampicin[!(PRM_stats$rifampicin$gene_mutation %in% rif_lin_PRM), ]
PRM_stats <- do.call("rbind", PRM_stats)
PRM_stats <- select(PRM_stats, drug, gene, mutation, n_samps, dst_prop, dr_prop, n_lins)
PRM_stats$dst_prop <- 1 - PRM_stats$dst_prop
names(PRM_stats) <- c("Drug", "Gene", "Change", "n samples", "DST sensitive", "DR-type sensitive", "n lineages")

```

```{r PRM-summary, echo = F, warning=F}

# PRM subset and summary
PRM_tabs <- lapply(binary_tables, function(x){subset(x, !(is.na(PRM)))})
PRM_tabs <- rm_null(lapply(PRM_tabs, function(x){
  tryCatch({
    sort_pos(pivot(x, formula(PRM ~ 'n')), 'PRM')
  }, error = function(e){})
            }))
PRM_tabs <- add_drug_col(PRM_tabs, drugs)
PRM_summary <- do.call("rbind", PRM_tabs)
PRM_summary <- select(PRM_summary, Drug, PRM, n)

```

```{r isoniazid-rare-mutations-cnts, echo = F, warning=F}

isoniazid_binary_table <- binary_tables$isoniazid
rare_katg_table <- subset(isoniazid_binary_table, is.na(KRM_katG) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_katG)) )
n_samps_rare_katg <- nrow(rare_katg_table)
n_rare_katg <- length(uniq_col(rare_katg_table$other_katG))
rare_katg_cnts <- tab2df(table(table(unlist(strsplit(rare_katg_table$other_katG, "; ")))))

# rare_katg_table[rare_katg_table$other_katG == "katG-p.Gln439His", ]

```

```{r rifampicin-rare-mutations-cnts, echo = F, warning=F}

rifampicin_binary_table <- binary_tables$rifampicin
rare_rpob_table <- subset(rifampicin_binary_table, is.na(KRM_rpoB) & !(is.na(CM)) & is.na(PRM) & !(is.na(other_rpoB)) )
n_samps_rare_rpob <- nrow(rare_rpob_table)
n_rare_rpob <- length(uniq_col(rare_rpob_table$other_rpoB))
rare_rpob_cnts <- tab2df(table(table(unlist(strsplit(rare_rpob_table$other_rpoB, "; ")))))

```

```{r rare-mutations-cnts, echo = F, warning=F}

rare_cnts <- plyr::rbind.fill(rare_katg_cnts, rare_rpob_cnts)
row.names(rare_cnts) <- drugs

```

```{r PRM-KRM-co-occurrence, echo = F, warning=F}

PRM_KRM <- rm_norow(lapply(binary_tables, function(x){subset(x, !(is.na(PRM)))}))
PRM_KRM <- lapply(PRM_KRM, function(x) {
  reshape2::dcast(x, PRM + KRM + other_vars ~ 'n', value.var = 'wgs_id', fun.aggregate = len_uniq) 
  })
PRM_KRM <- do.call("rbind", add_drug_col(PRM_KRM, drugs))
PRM_KRM <- dplyr::rename(PRM_KRM, unknown = other_vars)

```

```{r KRM-CM-or-no-CM-PRM-co-occurrence, echo = F, warning=F}

inh_bin <- binary_tables$isoniazid

# 4
# a
KRM_and_CM_and_PRM <- subset(inh_bin, KRM_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present")

n_samps_KRM_and_CM_and_PRM_rare_katg <- length(rm_na(KRM_and_CM_and_PRM$katG))

# c
KRM_and_no_CM_and_PRM <- subset(inh_bin, KRM_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present")

n_samps_KRM_and_no_CM_and_PRM_rare_katg <- length(rm_na(KRM_and_no_CM_and_PRM$katG))

```

```{r isoniazid-Ser315Thr, echo = F, warning=F}

ser315thr_tab <- subset(isoniazid_binary_table, grepl("Ser315Thr", isoniazid_binary_table$KRM))
# Move ser315thr string to front e.g. fabG1-c.-15C>T; katG-p.Ser315Thr -> katG-p.Ser315Thr; fabG1-c.-15C>T
ser315thr_tab$KRM <- ifelse(grepl("; katG-p.Ser315Thr", ser315thr_tab$KRM), swap_str(ser315thr_tab$KRM), ser315thr_tab$KRM)
# Drop the specific change for the co-occurring gene - e.g. katG-p.Ser315Thr; fabG1-c.-15C>T ->  katG-p.Ser315Thr; fabG1
ser315thr_tab$KRM_2 <- drop_change(ser315thr_tab$KRM)
ser315thr_pivot <- pivot(ser315thr_tab, formula(KRM_2 ~ 'n'))
ser315thr_pivot <- dplyr::rename(ser315thr_pivot, mutation = KRM_2)

```

```{r table-1, echo = F, warning=F}

# Make pivot tables of basic stats

# Add drug col, select and bind
meta_tables <- add_drug_col(binary_tables, drugs)
meta_table <- do.call('rbind', lapply(meta_tables, function(x){ select(x, Drug, wgs_id, main_lineage, sublin, country_code, drtype, dst)}))

# Trim whitespace
meta_table$dst <- trimws(meta_table$dst)

# Lin
lin_pivot <- pivot(meta_table, formula(main_lineage ~ 'n') )

# DR
drtype_pivot <- pivot(meta_table, formula(drtype ~ 'n') )
drtype_pivot <- trimws_df(drtype_pivot[order(match(drtype_pivot$drtype, dr_status_vect)), ])

# # Predicted DR to individual drugs
KRM_table <- do.call('rbind', lapply(meta_tables, function(x){ select(x, Drug, wgs_id, KRM)}))
KRM_table$KRM_binary <- ifelse(is.na(KRM_table$KRM), 0, 1)
KRM_pivot <- trimws_df(split_pivot(KRM_table, formula(Drug + KRM_binary ~ 'n'), "KRM_binary"))

# DST
# dst_pivot <- pivot(meta_table, formula(Drug + dst ~ 'n') )
dst_pivot <- trimws_df(split_pivot(meta_table, formula(Drug + dst ~ 'n'), "dst"))


table_1 <- data.frame()
lin_dr <- rbind(table_1,
                setNames(lin_pivot, names(table_1)),
                setNames(drtype_pivot, names(table_1)))

lin_dr <- cbind(rep("-", nrow(lin_dr)), lin_dr)

                 
KRM_dst <- rbind(table_1,
                 setNames(KRM_pivot, names(table_1)),
                 setNames(dst_pivot, names(table_1)))

table_1 <- rbind_force(lin_dr, KRM_dst)

names(table_1) <- c("Drug", "Chararcteristic", "n (%)")

table_1$Category <- c(rep("Lineage", nrow(lin_pivot)),
                      rep("DR status", nrow(drtype_pivot)),
                      rep("(predicted) resistance", nrow(KRM_pivot)),
                      rep("DST", nrow(dst_pivot)))

table_1 <- table_1[, c("Category", "Drug", "Chararcteristic", "n (%)")]

```

```{r basic-stats, echo = F, warning=F}

inh_bin <- binary_tables$isoniazid
rif_bin <- binary_tables$rifampicin
total_samps <- nrow(binary_tables[[1]])

inh_n_samps_DST_res <- dst_pivot[dst_pivot[, "Drug"] == "isoniazid" & dst_pivot[, "dst"] == "1", "n"]
rif_n_samps_DST_res <- dst_pivot[dst_pivot[, "Drug"] == "rifampicin" & dst_pivot[, "dst"] == "1", "n"]

inh_n_samps_CM <- nrow(subset(inh_bin, !(is.na(CM))))
rif_n_samps_CM <- nrow(subset(rif_bin, !(is.na(CM))))

inh_n_samps_CM_and_no_KRM <- nrow(subset(inh_bin, !(is.na(CM)) & is.na(KRM)))
rif_n_samps_CM_and_no_KRM <- nrow(subset(rif_bin, !(is.na(CM)) & is.na(KRM)))

```

```{r working, echo = F, warning=F}

# hier_tabs <- lapply(binary_tables, function(x){
# 
#   tab <- select(x, CM, KRM, PRM)
# 
#   tab_bin <- data.frame(apply(tab, 2, function(x){
#     ifelse(is.na(x), "absent", "present")
#   }))
# 
#   table(select(tab_bin, CM))
#   table(select(tab_bin, CM, KRM))
#   table(tab_bin)
# 
# })

```

| | |
| ----------- | ----------- |
|Gary Napier^1^ | gary.napier@lshtm.ac.uk |
|Susana Campino^1^ | susana.campino@lshtm.ac.uk |
|Jody E. Phelan^1^,* | jody.phelan@lshtm.ac.uk |
|Taane G. Clark^1^,^2^,* | taane.clark@lshtm.ac.uk |

^1^Faculty of Infectious and Tropical Diseases, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK
^2^Faculty of Epidemiology and Population Health, London School of Hygiene & Tropical Medicine, WC1E 7HT London, UK

\* joint authors


## Abstract

**Background:** Resistance to the first-line drugs isoniazid (INH) and rifampicin (RIF) in Mycobacterium tuberculosis (multidrug-resistant TB; MDR-TB) threatens disease control. Mutations that confer resistance to anti-TB drugs often come with a fitness cost. To overcome these fitness costs, *M. tuberculosis* can develop further, beneficial, compensatory mutations in genes that code for proteins performing similar biological roles or interact with drug-targets. Here we leverage the presence of known and putative compensatory mutations along with phenotypic and genotypic data to detect the presence of novel mutations occurring in the target genes of INH (*katG*) and RIF (*rpoB*), inferring their causative role in resistance to these drugs.

**Results:** Across ~32k isolates, there were `r inh_n_samps_DST_res` with INH and `r rif_n_samps_DST_res` with RIF phenotypic resistance. Known mutations in *katG* and *rpoB* explained most of the resistance. However, `r fmt(inh_n_samps_CM_and_no_KRM)` (`r fmt_pc(inh_n_samps_CM_and_no_KRM/total_samps)`) and `r fmt(rif_n_samps_CM_and_no_KRM)` (`r fmt_pc(rif_n_samps_CM_and_no_KRM/total_samps)`) samples had compensatory mutations, but did not have any known resistance mutations in INH or RIF target genes, respectively. By examining isolates with compensatory mutations but no known resistance mutation, we identified a number of inferred novel resistance mutations. We also characterise the putative *katG* mutations in terms of their co-occurrence with other INH variants - *katG*-Ser315Thr and other INH genes, e.g. *fabG1*. 

**Conclusions:** Detecting resistance mutations from whole genome sequencing can inform clinical management, but also revealed new variants linked to resistance. We have identified new resistance markers and, after validation, have added to a list of mutations for genotypic drug-resistance predictions.

**URL:**

**Keywords:** 

## Background

[Generic remark about tuberculosis being bad and being caused by the *M. tuberculosis* pathogen]

In gaining resistance against anti-tuberculosis drugs, drug target proteins are often compromised in their usual function, and *M. tuberculosis* therefore also incurs a fitness cost. Costs can be manifest in terms of any phenotypic difference, say, virulence or transmissibility. For example the *katG* gene codes for the KatG enzyme, a catalase-peroxidase which protects the bacterium from reactive oxygen species (ROS) damage and to detoxify hydrogen peroxide [@Trivedi2012], two products of the immune system which attack the pathogen. The enzyme is also the target of the pro-drug isoniazid (INH), converting it to multiple reactive species [@Zhang2009]. When mutations occur in the KatG protein, mediated by the *katG* gene, that disrupt INH binding, the bacteria is often left simultaneously with drug resistance and a faulty protein. This is not always the case, since some mutations can confer drug resistance without a fitness cost. For example the *katG*-Ser315Thr mutation confers resistance but minimally affects finess, hence is highly prevalent among Pre-MDR-TB and MDR-TB strains [@Hazbon2006][@Phelan2019]. 

```{r n-samples-tree, echo = F, warning=F}

# Tree

# INH
# 1
total_samps
# 2
nrow(subset(inh_bin, KRM_katG_bin == "present"))
nrow(subset(inh_bin, KRM_katG_bin == "absent"))
# 3
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present"))
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent"))
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present"))
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent"))
# 4
# a
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present"))
# b
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# c
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# d
nrow(subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# e
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present"))
# f
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent"))
# g
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present"))
# h
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent"))
# 5
# a
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           !(is.na(other_katG))))
# b
nrow(subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent" &
           is.na(other_katG)))




# # RIF
# # 1
# total_samps
# # 2
# nrow(subset(rif_bin, KRM_bin == "present"))
# nrow(rif_n_samps_no_KRM <- subset(rif_bin, KRM_bin == "absent"))
# # 3
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "present"))
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "absent"))
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "present"))
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "absent"))
# # 4
# # a
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "present" &
#          PRM_bin == "present"))
# # b
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "present" &
#          PRM_bin == "absent"))
# # c
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "absent" &
#          PRM_bin == "present"))
# # d
# nrow(subset(rif_bin, KRM_bin == "present" &
#          CM_bin == "absent" &
#          PRM_bin == "absent"))
# # e
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "present" &
#          PRM_bin == "present"))
# # f
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "present" &
#          PRM_bin == "absent"))
# # g
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "absent" &
#          PRM_bin == "present"))
# # h
# nrow(subset(rif_bin, KRM_bin == "absent" &
#          CM_bin == "absent" &
#          PRM_bin == "absent"))


# a
a <- subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "present")$wgs_id
# b
b <- subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "present" &
         PRM_bin == "absent")$wgs_id
# c
c <- subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "present")$wgs_id
# d
d <- subset(inh_bin, KRM_katG_bin == "present" &
         CM_bin == "absent" &
         PRM_bin == "absent")$wgs_id
# e
e <- subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "present")$wgs_id
# f
f <- subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "present" &
         PRM_bin == "absent")$wgs_id
# g
g <- subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "present")$wgs_id
# h
h <- subset(inh_bin, KRM_katG_bin == "absent" &
         CM_bin == "absent" &
         PRM_bin == "absent")$wgs_id

samps_list <- list(a, b, c, d, e, f, g, h)
for(i in seq(letters[1:8])){
  write.table(samps_list[[i]], file = paste0(letters[i], ".txt"),
              quote = F, col.names = F, row.names = F)
}

```

## Results

```{r results-isolate-collection, echo = F, warning=F}

# **Isolate collection**
inh_n_samps_TBP_res <- KRM_pivot[KRM_pivot[, "Drug"] == "isoniazid" & KRM_pivot[, "KRM_binary"] == "1", "n"]
rif_n_samps_TBP_res <- KRM_pivot[KRM_pivot[, "Drug"] == "rifampicin" & KRM_pivot[, "KRM_binary"] == "1", "n"]
n_samps_MDR <- drtype_pivot[drtype_pivot[, "drtype"] == "MDR-TB", "n"]


```

**Isolate collection**

*Summary* \
Number INH resistant (DST) - `r inh_n_samps_DST_res` \
Number RIF resistant (DST) - `r rif_n_samps_DST_res` \

Number INH resistant (TBprofiler) - `r inh_n_samps_TBP_res`
Number RIF resistant (TBprofiler) - `r rif_n_samps_TBP_res` \

Number MDR resistant (TBprofiler) - `r n_samps_MDR`

A total of `r total_samps` were suitable for analysis after filtering out mixed lineages. Of these, many had been tested as phenotypically resistant (INH = `r inh_n_samps_DST_res`; RIF = `r rif_n_samps_DST_res`). Supplementing the DST data, `r inh_n_samps_TBP_res` samples were predicted to be resistant to INH and `r rif_n_samps_TBP_res` samples were predicted to be resistant to RIF by the presence of known resistance mutations in TB-profiler [@phelan2019]. Also predicted by genotypic profiling in TB-profiler, `r n_samps_MDR` samples were classified as MDR-TB (see **Table \@ref(tab:table_1)**)

```{r results-resistance-mutations, echo = F, warning=F}

# **Resistance mutations, and how many explain the phenotype**
inh_n_samps_KRM <- nrow(subset(inh_bin, KRM_bin == "present"))
rif_n_samps_KRM <- nrow(subset(rif_bin, KRM_bin == "present"))

inh_n_samps_KRM_pc <- fmt_pc(nrow(subset(inh_bin, KRM_bin == "present"))/total_samps)
rif_n_samps_KRM_pc <- fmt_pc(nrow(subset(rif_bin, KRM_bin == "present"))/total_samps)

inh_n_samps_DST_res_KRM_present <- nrow(subset(inh_bin, dst == "1" & KRM_bin == "present"))
inh_n_samps_DST_res_KRM_present_pc <- fmt_pc(inh_n_samps_DST_res_KRM_present/tonum(inh_n_samps_DST_res))

rif_n_samps_DST_res_KRM_present <- nrow(subset(rif_bin, dst == "1" & KRM_bin == "present"))
rif_n_samps_DST_res_KRM_present_pc <- fmt_pc(rif_n_samps_DST_res_KRM_present/tonum(rif_n_samps_DST_res))

```

**Resistance mutations, and how many explain the phenotype**

*Summary* \
Number INH samples DST resistant and have a known resistance mutation - `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) \
Number RIF samples DST resistant and have a known resistance mutation - `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) \

The total number of samples with a known resistance mutation in an INH target gene (*katG*, *fabG1*, *inhA*) was `r inh_n_samps_KRM` (`r inh_n_samps_KRM_pc`), and `r rif_n_samps_KRM` (`r rif_n_samps_KRM_pc`) for the RIF target gene (*rpoB*).

Explaining the DST resistance status, `r fmt(inh_n_samps_DST_res_KRM_present)` (`r inh_n_samps_DST_res_KRM_present_pc`) samples had a known resistance mutation in an INH target gene, while `r fmt(rif_n_samps_DST_res_KRM_present)` (`r rif_n_samps_DST_res_KRM_present_pc`) samples had an *rpoB* mutation (RIF).


```{r results-compensatory-mutations, echo = F, warning=F}

# **Compensatory mutations**
inh_CM <- select(subset(inh_bin, CM_bin == "present"), CM)
inh_n_samps_KRM_CM <- nrow(subset(inh_bin, KRM_bin == "present" & 
                                    CM_bin == "present"))
inh_n_samps_KRM_no_CM <- nrow(subset(inh_bin, KRM_bin == "absent" & 
                                    CM_bin == "present"))
inh_n_CM_haplotypes <- length(unique(inh_CM$CM))
inh_CM_uniq <- uniq_vars(inh_CM$CM)
inh_CM_uniq_in_KCM <- inh_CM_uniq[inh_CM_uniq %in% inh_KCM$gene_change]
inh_PCM <- inh_CM_uniq[!(inh_CM_uniq %in% inh_KCM$gene_change)]

rif_CM <- select(subset(rif_bin, CM_bin == "present"), CM)
rif_n_samps_KRM_CM <- nrow(subset(rif_bin, KRM_bin == "present" & 
                                    CM_bin == "present"))
rif_n_samps_KRM_no_CM <- nrow(subset(rif_bin, KRM_bin == "absent" & 
                                    CM_bin == "present"))
rif_n_CM_haplotypes <- length(unique(rif_CM$CM))
rif_CM_uniq <- uniq_vars(rif_CM$CM)
rif_CM_uniq_in_KCM <- rif_CM_uniq[rif_CM_uniq %in% rif_KCM$gene_change]
rif_PCM <-  rif_CM_uniq[!(rif_CM_uniq %in% rif_KCM$gene_change)]

# Number of CM mutations in samples where there are KRM
# inh_CM_KRM_uniq <- uniq_vars(subset(inh_bin, !(is.na(CM)) & !(is.na(KRM)))$CM)
# Number of CM mutations in samples where there are no KRM
# inh_CM_no_KRM_uniq <- uniq_vars(subset(inh_bin, !(is.na(CM)) & is.na(KRM))$CM)




```

**Compensatory mutations**

*summary* \
Number INH samples with a CM - `r inh_n_samps_CM` \
Number INH CM haplotypes - `r inh_n_CM_haplotypes` \
Number INH unique CM - `r length(inh_CM_uniq)` \
Number INH CM in known CM list - `r length(inh_CM_uniq_in_KCM)` \
Number INH CM previously non-DR associated - `r length(inh_PCM)` \

Number RIF samples with a CM - `r rif_n_samps_CM` \
Number RIF CM haplotypes - `r rif_n_CM_haplotypes` \
Number RIF unique CM - `r length(rif_CM_uniq)` \
Number RIF CM in known CM list - `r length(rif_CM_uniq_in_KCM)` \
Number RIF CM previously non-DR associated - `r length(rif_PCM)` \

There were `r inh_n_samps_CM` and `r fmt(rif_n_samps_CM)` samples with a compensatory mutation, of which `r fmt(inh_n_samps_KRM_CM)` and `r fmt(rif_n_samps_KRM_CM)` samples had a known resistance mutation, and `r fmt(inh_n_samps_KRM_no_CM)` and `r fmt(inh_n_samps_KRM_no_CM)` did not have known resistance mutations in INH and RIF, respectively. 

The number of compensatory mutation haplotypes were `r inh_n_CM_haplotypes` (INH) and `r rif_n_CM_haplotypes` (RIF), while the number of unique compensatory mutations were `r length(inh_CM_uniq)` (INH) and `r length(rif_CM_uniq)` (RIF).

Of these, `r length(inh_CM_uniq_in_KCM)` (INH) and `r length(rif_CM_uniq_in_KCM)` (RIF) were from compensatory mutations previously known, whereas `r length(inh_PCM)` INH and `r length(rif_PCM)` RIF compensatory mutations were putative. 

```{r results-PRMs, echo = F, warning=F}

# **Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers**
inh_PRM <- select(subset(inh_bin, PRM_bin == "present"), PRM)
inh_n_PRM_haplotypes <- length(unique(inh_PRM$PRM))
inh_PRM_uniq <- unique(unlist(strsplit(inh_PRM$PRM, "; ")))

rif_PRM <- select(subset(rif_bin, PRM_bin == "present"), PRM)
rif_n_PRM_haplotypes <- length(unique(rif_PRM$PRM))
rif_PRM_uniq <- unique(unlist(strsplit(rif_PRM$PRM, "; ")))

```

**Filling in the gaps of phenotypic resistance, leveraging compensatory, and revealing novel resistance markers**

*Summary* \
Number INH samples with a PRM - `r nrow(inh_PRM)` \
Number INH PRM haplotypes - `r inh_n_PRM_haplotypes` \
Number INH unique PRM - `r length(inh_PRM_uniq)` \

Number RIF samples with a PRM - `r nrow(rif_PRM)` \
Number RIF PRM haplotypes - `r rif_n_PRM_haplotypes` \
Number RIF unique PRM - `r length(rif_PRM_uniq)` \

<!-- LAYER 4 -->
Number INH samples with KRM (*katG*) and CM and PRM - `r fmt(length(a))`
Number INH samples with KRM (*katG*), no CM and PRM - `r fmt(length(c))`
Number INH samples with no KRM (*katG*) and CM and PRM - `r fmt(length(e))`
Number INH samples with no KRM (*katG*), no CM and PRM - `r fmt(length(g))`

<!-- LAYER 5 --> 
Number INH samples with no KRM (*katG*), CM, no PRM, other *katG* - `r inh_n_samps_no_KRM_CM_PRM_other_katG`
Number INH unique rare *katG* mutations - `r `

After applying filters to the samples with compensatory mutations to find potential resistance mutations, and then re-searching the full TB-profiler database for samples with said mutations, `r nrow(inh_PRM)` INH isolates were found. No samples were found with potential resistance mutations in *rpoB*. In these INH samples, there were `r inh_n_PRM_haplotypes`, with `r length(inh_PRM_uniq)` unique mutations (see **Table \@ref(tab:PRM-summary-table) and Table \@ref(tab:PRM-stats-table)**).

Within the INH samples with putative resistance mutations, `r fmt(length(c))` samples had a known *katG* mutation but no compensatory mutations (see **Table \@ref(tab:PRM-KRM-other-table)**), `r fmt(length(e))` samples had no known *katG* mutation, but did have a compensatory mutation, while the majority, `r fmt(length(g))` had no known *katG* mutation and no compensatory mutations (see **Figure \@ref(fig:samples-breakdown)**).

Further to the potential resistance mutations, there were also found `r n_samps_rare_katg` `r n_samps_rare_rpob` samples with no known *katG* resistance mutation, having a compensatory mutation, but not having a potential resistance mutation by the criteria outlined in *Methods*, nevertheless with a non-drug-resistance-associated *katG* mutation ('rare *katG* mutation'). 


<!-- SENSITIVITY -->


## Discussion

## Conclusions

## Methods

*Sequence data and processing*

Samples were obtained from multiple sources, isolates belonging to individual patients.

Quality inspection of sequence reads was condcucted in fastQC (v0.11.9) (www.bioinformatics.babraham.ac.uk/projects/fastqc/). Reads were trimmed to remove low-quality sequences in Trimmomatic (v0.39)	- parameters: *PE, -phred33, LEADING:3, TRAILING:3, SLIDINGWINDOW:4:20, MINLEN:36* [@Bolger2014] and sequences were aligned to the H37Rv reference genome (AL123456) with BWA mem (v0.7.17) [@Li2013], [@Cole1998]. Joint SNP calling was carried out in gatk GenotypeGVCFs (v4.1.3.0) - --java-options "-Xmx40g" - [@Depristo2011].

SNP were filtered to exclude indels and heterozygous SNPs. Monomorphic SNPs and those in non-unique regions of the genome (e.g. ppe genes) were excluded. These procedures were conducted using bcftools view; bcftools filter (v1.9) (using htslib 1.9); parameters *-V indels -e 'GT="het"' -S . -a -c 1 -Oz* [@Li2009]. Multiple variant calling was done with gatk GenotypeGVCFs (v4.1.3.0) (gatk.broadinstitute.org). 

Multi-FASTA formated filed were created from the filtered SNP file and H37Rv reference fasta with bedtools makewindows (v2.28.0); parameters -w 50000 -s 50000 [@Quinlan2010]. 

Phylogenetic trees were made with IQ-TREE (v1.6.12), involving a general time reversible model with rate heterogeneity set to a discrete Gamma model and an ascertainment bias correction (parameters −m GTR+G+ASC), with 1000 bootstrap samples [@Nguyen2015].

Drug resistance types and lineages were predicted *in silico* with TB-Profiler (v2.4) [@Phelan2019] [@Coll2015]. TB-Profiler also used identified all known drug-resistance mutations and all other mutations in relevant genes for finding compensatory mutations and putative novel resistance mutations. 

*Finding putative resistance markers using compensatory mutations*

From the database of TB-profiler - all non-synonymous mutations in the relevant compensatory genes were found (*ahpC* in the case of INH and *rpoC* and *rpoA* for RIF).

All non-drug resistance associated ('unknown') mutations in compensatory genes were filtered such that they were present in three or more samples, <50% samples were predicted *Sensitive* by TB-profiler and <50% of samples were DST sensitive, for each relevant drug.

These successful mutations were then combined with the list of known compensatory mutations [@Phelan2019] to form a full list of compensatory mutations.

To find putative resistance mutations, all non-synonymous mutations in the relevant resistance genes were pulled from the TB-profiler database (*katG* for INH and *rpoB* for RIF), with the exclusion of isolates with mixed lineages. Some variants were excluded from consideration for PRM. *katG*-Arg463Leu is known not to be associated with INH resistance [@Doorn2001]. 

For each drug, mutations were found in samples where a compensatory mutation was present and no known resistance mutations were present, but a non-resistance-associated mutation was present in the relevant target genes. 

These mutations were then filtered to exclude known drug-resistance-associated variants, and subjected to the same criteria as the putative compensatory mutations - i.e. present in three or more samples, <50% samples were predicted *Sensitive* by TB-profiler and <50% of samples were DST sensitive. 

Using this list of potential resistance mutations, all TB-profiler samples were then searched for their presence, regardless of compensatory mutation status. It should be noted that therefore not all samples with potential resistance mutations necessarily have compensatory mutations, and vice versa. 

## List of abbreviations

* *Mtb* = *Mycobacterium tuberculosis*
* TB = tuberculosis
<!-- * RM = any resistance mutation; -->
<!-- * KRM = known RM; -->
<!-- * PRM = potential RM; -->
<!-- * CM = any compensatory mutation; -->
<!-- * KCM = known CM; -->
<!-- * PCM = potential CM -->
* INH = isoniazid
* RIF = rifampicin
* WGS = whole genome sequencing

## Declarations

## Ethics approval and consent to participate
Not applicable

## Consent for publication
Not applicable 

## Competing interests
The authors declare that they have no competing interests

## Funding
GN is funded by an BBSRC-LiDO PhD studentship. JEP is funded by a Newton Institutional Links Grant (British Council, no. 261868591). TGC is funded by the Medical Research Council UK (Grant no. MR/M01360X/1, MR/N010469/1, MR/R025576/1, and MR/R020973/1). SC is funded by Medical Research Council UK grants (ref. MR/M01360X/1, MR/R025576/1, and MR/R020973/1). The authors declare no conflicts of interest. The funders had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript should be declared.

## Authors' contributions
JEP and TGC conceived and directed the project. GN performed bioinformatic and statistical analyses under the supervision of SC, JEP and TGC. GN, SC, JEP and TGC interpreted results. GN wrote the first draft of the manuscript with inputs from JEP and TGC. All authors commented and edited on various versions of the draft manuscript and approved the final version. GN, SC, JEP, and TGC compiled the final manuscript.

## Acknowledgements

## Tables

```{r table-1-display, echo = F, warning=F}

cap <- sprintf("*Mycobacterium tuberculosis* samples (n=%s)", total_samps)
knitr::kable(table_1, caption = cap, format = "pipe", row.names = F)

```


<!-- ```{r summary-table, echo = F, warning=F} -->

<!-- # knitr::kable(summaries,  caption = "Summary of each PRM") %>% kable_styling(font_size = table_font_sz) -->
<!-- knitr::kable(summaries, caption = "Potential resistance mutations ", format = "pipe") -->

<!-- ``` -->


```{r PRM-summary-table, echo = F, warning=F}

cap <- sprintf("Potential resistance mutation (PRM) haplotypes for isoniazid (n=%s)", nrow(inh_PRM))
knitr::kable(PRM_summary, caption = , row.names = F, format = "pipe")

```

```{r PRM-stats-table, echo = F, warning=F}

cap <- "Number of samples, DST sensitive proportion, (predicted) DR type proportion and number of lineages for each PRM"
knitr::kable(PRM_stats, row.names = F, caption = cap, format = "pipe")

```

**Rare variants**

```{r rare-counts-table, echo = F, warning=F}

cap <- "Number of variants by frequency of occurrence for *katG* (isoniazid) and *rpoB* (rifampicin) among samples with no KRM, at least one CM, but no PRM. These variants are not counted as ‘PRM’ because of their low frequency (only one or two samples). Variants occurring more than twice were rejected because most samples were DR-type sensitive or DST sensitive, or in the case of INH, the samples already have a known resistance mutation in a gene other than *katG*."
knitr::kable(rare_cnts, caption = cap, format = "pipe")

```

**Samples with PRMs - co-occurrences of KRM and unknown mutations**

```{r PRM-KRM-other-table, echo = F, warning=F}

cap <- "Known resistance mutations and non-DR-associated (‘unknown’) mutations co-occurring in samples with putative drug resistance"
knitr::kable(PRM_KRM, row.names = F, caption = cap, format = "pipe")

```

**Isoniazid**

Comparison to katG-Ser315Thr and katG-Ser315Thr-KRM in other isoniazid resistance genes.

```{r isoniazid-Ser315Thr-table, echo = F, warning=F, }

cap <- "Comparison to katG-Ser315Thr and katG-Ser315Thr-KRM in other isoniazid resistance genes."
knitr::kable(ser315thr_pivot, row.names = F, caption = cap, format = "pipe")

```

## Figures

```{r samples-breakdown, echo=FALSE, out.width = "2000px", fig.cap="Breakdown of sample numbers by presence of mutation type for INH and RIF" }

knitr::include_graphics(inh_numbers_file)
knitr::include_graphics(rif_numbers_file)

```


<!-- ```{r samples-breakdown, echo=FALSE, out.width = "2000px", fig.cap="Breakdown of sample numbers by presence of mutation type for INH" } -->

<!-- knitr::include_graphics("../results/samples_breakdown_inh.png") -->

<!-- ``` -->


```{r isoniazid-tree, echo=FALSE, out.width = "2000px", fig.cap="Phylogenetic tree of isolates with resistance; Isoniazid resistance (n=XX), with potentially new katG mutations"}

knitr::include_graphics(tree_files[grep("isoniazid", tree_files)])

```


```{r inh-ahpc-katg-links, echo=FALSE, out.width = "2000px", fig.cap="Co-occurrences between *ahpC* compensatory mutation positions and *katG* positions. Gaps in *katG* positions indicate no compensatory mutation was found for that sample." }

binary_expand_PRM <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(PRM = strsplit(as.character(PRM), "; ")) %>%
  unnest(PRM))
})

inh <- binary_expand_PRM$isoniazid
inh_PRM <- subset(inh, !(is.na(PRM)))
inh_PRM <- select(inh_PRM, CM, PRM)

inh_PRM_pos <- data.frame(apply(inh_PRM, 2, find_pos))

colnames(inh_PRM_pos) <- c("ahpC", "katG")

df <- inh_PRM_pos %>%
  make_long(ahpC, katG)

ggplot(df, aes(x = x, next_x = next_x, node = node, next_node = next_node, label = node)) +
  coord_flip()+
  geom_sankey(flow.alpha = .6, fill = 'lightblue') +
  geom_sankey_text(size = 3, color = "black", angle = 45, width = 1) +
  theme_minimal()+
  labs(x = NULL)+
  theme(legend.position = "none",
        plot.title = element_text(hjust = .5),
        axis.text.x=element_blank()) +
  ggtitle("Isoniazid")


```

### References
































