---
title: "Compensatory mutations in *ahpC* reveal novel *katG* mutations implicated in isoniazid resistance in Mycobacterium tuberculosis"
output: 
  bookdown::word_document2: 
    fig_caption: yes
    table_caption: yes
    number_sections: no
mainfont: Calibri
fontsize: 10pt
bibliography: all_refs.bib
csl: biomed-central.csl
---

```{r setup, include=FALSE, echo = F, warning=F, message=F}

# SETUP ----

rm(list = ls())

knitr::opts_chunk$set(echo = F)
# options(scipen=1, digits=2)
options(scipen=999, digits = 3)
table_font_sz <- 8

```

```{r library, include=FALSE, echo = F, warning=F, message=F}

library(tidyr)
library(dplyr)
library(knitr)
library(reshape2)
library(janitor)

```

```{r functions, echo = F, warning=F, message=F}

source("https://raw.githubusercontent.com/GaryNapier/Packages_functions/master/Functions.R")

uniq_col <- function(x, split = "; "){
  unique(unlist(strsplit(x, split)))
}

col_pc <- function(df, col){
  round((df[, col] / sum(df[, col]))*100, 3)
}

len_uniq <- function(x){
  length(unique(x))
}

rbind_force <- function(df1, df2){
  rbind(df1, setNames(df2, names(df1)))
}

# Make quick pivot table summarising total unique values by column
# my_formula arg must be call to the formula() function: formula(<col> ~ 'n')
# Result:
#             PRM            n
#  katG-c.2223A>G  3   (4.05%)
# katG-p.Arg484His  5   (6.76%)
# ...
# katG-p.Tyr413Cys  8  (10.81%)
#           Total 74 (100.00%)
pivot <- function(x, my_formula, value_var = "wgs_id"){
  drop_cols(to_table(reshape2::dcast(x, my_formula, value.var = value_var, fun.aggregate = len_uniq),
                     pc_dir = 'col'), 
            'Total')
}

duped <- function(x, col){
  dups <- x[duplicated(x[col]), col]
  x[x[, col] %in% dups, ]
}

# Swap first and second parts of string, splitting on a charachter (or substring)
# e.g. 
# x <- "katG-p.Ile335Val; katG-p.Ser315Thr"
# swap_str(x, "; ")
# "katG-p.Ser315Thr; katG-p.Ile335Val"
swap_str <- function(x, split_chr = "; "){
  paste0(unlist(strsplit(x, split_chr))[c(2, 1)], collapse = split_chr)
}


# Print each element of vector vertically instead of across the screen
# e.g. 
# > x <- c("a", "b", "c")
# > x
# [1] "a" "b" "c"
# > print_vert(x)
# a
# b
# c
print_vert <- function(x){
  cat(paste0(x, '\n'))
}

# Remove change from second gene-change pair
# e.g.
# x <- "katG-p.Ser315Thr; fabG1-c.-15C>T"
# > drop_change(x, "-")
# [1] "katG-p.Ser315Thr; fabG1"
drop_change <- function(string, split_on = "-"){
  unlist(lapply(strsplit(string, split_on), function(x){paste0(x[c(1, 2)], collapse = split_on)}))
}

clean_binary_table <- function(x){
  x <- data.frame(apply(x, 2, function(x){gsub("\\[\\]", NA, as.character(x))}))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\[||\\]", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\), \\(", "; ", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\', \\'", "-", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\(||\\)", "", as.character(x)) }))
  x <- data.frame(apply(x, 2, function(x){ gsub("\\'", "", as.character(x)) }))
  x
}

sort_pos <- function(x, col){
  
  x$pos <- as.numeric(stringr::str_extract(x[, col], "[0-9]+"))
  x <- x[order(x$pos), ]
  drop_cols(x, 'pos')
  
}

lapply_nms <- function(x, fun, nms){
  res <- lapply(x, fun)
  names(res) <- nms
  res
}

```

```{r variables, echo = F, warning=F, message=F}


# VARIABLES ----

id_col <- "wgs_id"
rnd <- 3
# drugs <- c("rifampicin", "isoniazid", "ethambutol", "pyrazinamide", "streptomycin", 
#              "ofloxacin", "moxifloxacin", "levofloxacin", "amikacin", "kanamycin", "capreomycin", "ciprofloxacin", "prothionamide", 
#              "ethionamide", "para_aminosalicylic_acid", "cycloserine",
#              "clarithromycin", "clofazimine", "bedaquiline", "linezolid", "rifabutin", "delamanid")
drugs <- c("isoniazid", "rifampicin")
fq <- "fluoroquinolones, ofloxacin, levofloxacin, ciprofloxacin, moxifloxacin"
agc <- "capreomycin, streptomycin, amikacin, kanamycin, aminoglycosides"
align <- c("l", "r", "r")
blue_col <- "cornflowerblue"
plot_text_sz <- 0.5
main_cex <- 1.5
# dr_status_vect <- c("Sensitive", "Pre-MDR", "MDR", "Pre-XDR", "XDR", "Other")
dr_status_vect <- c("Sensitive", "Pre-MDR-TB", "HR-TB", "RR-TB", "MDR-TB", "Pre-XDR-TB", "XDR-TB", "Other")

```

```{r paths, echo = F, warning=F, message=F}

# PATHS ----

setwd("~/Documents/comp_mut/r_scripts/")

metadata_path <- "../../metadata/"
local_metadata_path <- "../metadata/"
pipeline_db_path <- "../../pipeline/db/"
db_path <- "../../tbdb/"
results_path <- "../results/"
newick_path <- ""

```

```{r files, echo = F, warning=F, message=F}

# FILES ----

# Metadata
metadata_file <- paste0(metadata_path, "tb_data_18_02_2021.csv")
# Database data
lineage_conversion_file <- paste0(pipeline_db_path, "lineage_conversions.txt")
dr_genes_file <- paste0(db_path, "tbdb.csv")
KCM_file <- paste0(pipeline_db_path, "compensatory_mutations.csv")
CM_RM_gene_assoc_file <- paste0(local_metadata_path, "CM_RM_gene_assoc.csv")

```

```{r drug-files, echo = F, warning=F, message=F}

binary_table_files <- paste0(results_path, drugs, "_binary_table.csv")
PCM_merged_files <- paste0(results_path,  drugs, "_PCM_merged.csv")
summary_files <- paste0(results_path, drugs, "_summary.csv")
PRM_stats_files <- paste0(results_path, drugs, "_PRM_stats.csv")
tree_files <- paste0(results_path, drugs, "_tree.png")

```

<!-- ```{r isoniazid-files, echo = F, warning=F, message=F} -->

<!-- # Results -->
<!-- isoniazid_binary_table_file <- paste0(results_path, "isoniazid_binary_table.csv") -->
<!-- isoniazid_PCM_merged_file <- paste0(results_path, "isoniazid_PCM_merged.csv") -->
<!-- isoniazid_summary_file <- paste0(results_path, "isoniazid_summary.csv") -->
<!-- isoniazid_PRM_stats_file <- paste0(results_path, "isoniazid_PRM_stats.csv") -->
<!-- isoniazid_tree_file <- paste0(results_path, "isoniazid_tree.png") -->

<!-- ``` -->

<!-- ```{r rifampicin-files, echo = F, warning=F, message=F} -->

<!-- rifampicin_binary_table_file <- paste0(results_path, "rifampicin_binary_table.csv") -->
<!-- rifampicin_PCM_merged_file <- paste0(results_path, "rifampicin_PCM_merged.csv") -->
<!-- rifampicin_summary_file <- paste0(results_path, "rifampicin_summary.csv") -->
<!-- rifampicin_PRM_stats_file <- paste0(results_path, "rifampicin_PRM_stats.csv") -->
<!-- rifampicin_tree_file <- paste0(results_path, "rifampicin_tree.png") -->

<!-- ``` -->


```{r read-in-data, echo = F, warning=F, message=F}

# READ IN DATA ----

# Metadata
metadata <- read.csv(metadata_file)
# Database data
lineage_conversions <- read.delim(lineage_conversion_file)
# dr_genes <- read.delim(dr_genes_file, header = F)
dr_genes <- read.csv(dr_genes_file)
KCM <- read.csv(KCM_file)
CM_RM_gene_assoc <- read.csv(CM_RM_gene_assoc_file)

```

```{r drugs-read-in-data, echo = F, warning=F, message=F}

binary_tables <- lapply_nms(binary_table_files, read.csv, drugs)
PCM_merged <- lapply_nms(PCM_merged_files, read.csv, drugs)
summaries <- lapply_nms(summary_files, read.csv, drugs)
PRM_stats <- lapply_nms(PRM_stats_files, read.csv, drugs)

```


<!-- ```{r isoniazid-read-in-data, echo = F, warning=F, message=F} -->

<!-- isoniazid_binary_table <- read.csv(isoniazid_binary_table_file) -->
<!-- isoniazid_PCM_merged <- read.csv(isoniazid_PCM_merged_file) -->
<!-- isoniazid_summary <- read.csv(isoniazid_summary_file) -->
<!-- isoniazid_PRM_stats <- read.csv(isoniazid_PRM_stats_file) -->

<!-- ``` -->

<!-- ```{r rifampicin-read-in-data, echo = F, warning=F, message=F} -->

<!-- rifampicin_binary_table <- read.csv(rifampicin_binary_table_file) -->
<!-- rifampicin_PCM_merged <- read.csv(rifampicin_PCM_merged_file) -->
<!-- rifampicin_summary <- read.csv(rifampicin_summary_file) -->
<!-- rifampicin_PRM_stats <- read.csv(rifampicin_PRM_stats_file) -->

<!-- ``` -->


```{r clean-data, echo = F, warning=F, message=F}

# Remove all the "[]" present in the binary table caused by python's lists
binary_tables <- lapply(binary_tables, clean_binary_table)

# Duplicate sample row if other_vars has more than one var (i.e. on the "; " character/string)
binary_tables_dup <- lapply(binary_tables, function(x){
  data.frame(x %>%
  mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>%
  unnest(other_vars))
})

# Make col for just the other_vars genes, stripping out the change
binary_tables_dup <- lapply(binary_tables_dup, function(x){x$other_vars_gene <- gsub("-.*", "", x$other_vars); x})

# KCM - combine gene-change
KCM$gene_change <- paste0(KCM$Gene, "-", KCM$Mutation)

# Subset KCM 
# isoniazid_KCM <- KCM[KCM[,"Gene"] == "ahpC", ]


# Put PCM_merged together and join in the drug names
PCM_merged <- do.call("rbind", PCM_merged)
PCM_merged$gene_change <- paste0(PCM_merged$gene, "-", PCM_merged$change)

PCM_merged <- merge(PCM_merged, CM_RM_gene_assoc, 
                    by.x = "gene", by.y = "CM_gene", 
                    sort = F)

# Remove the PCMs in the KCM
PCM_merged <- PCM_merged[!(PCM_merged$gene_change %in% KCM$gene_change), ]

# Transpose summary
summaries <- do.call('cbind', lapply(summaries, function(x){data.frame(n = t(x))}))
names(summaries) <- drugs

# PRM stats
PRM_stats <- lapply(PRM_stats, sort_pos, 'mutation')
PRM_stats <- do.call("rbind", PRM_stats)
PRM_stats <- select(PRM_stats, drug, gene, mutation, n_samps, dst_prop, dr_prop, n_lins)
PRM_stats$dst_prop <- 1 - PRM_stats$dst_prop
names(PRM_stats) <- c("Drug", "Gene", "Change", "n samples", "DST sensitive", "DR-type sensitive", "n lineages")

```

```{r PRM-summary, echo = F, warning=F}

# PRM subset and summary
# isoniazid_PRM_tab <- subset(isoniazid_binary_table, !(is.na(PRM)))
PRM_tabs <- lapply(binary_tables, function(x){subset(x, !(is.na(PRM)))})
# isoniazid_PRM_summary <- sort_pos(pivot(isoniazid_PRM_tab, formula(PRM ~ 'n')), 'PRM')
PRM_tabs <- lapply(PRM_tabs, function(x){sort_pos(pivot(x, formula(PRM ~ 'n')), 'PRM')})
PRM_tabs <- lapply_mod(PRM_tabs, function(i){PRM_tabs[[i]]$Drug <- rep(drugs[i], nrow(PRM_tabs[[i]])); PRM_tabs[[i]]})
PRM_summary <- do.call("rbind", PRM_tabs)
PRM_summary <- select(PRM_summary, Drug, PRM, n)

```

<!-- ```{r isoniazid-clean-data, echo = F, warning=F, message=F} -->

<!-- # Remove all the "[]" present in the binary table caused by python's lists -->
<!-- isoniazid_binary_table <- clean_binary_table(isoniazid_binary_table) -->

<!-- # Split out the other_vars to make extra row per mutation -->
<!-- isoniazid_binary_split <- data.frame(isoniazid_binary_table %>% -->
<!--   mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>% -->
<!--   unnest(other_vars)) -->

<!-- # Make col for just the other_vars genes, stripping out the change -->
<!-- isoniazid_binary_split$other_vars_gene <- gsub("-.*", "", isoniazid_binary_split$other_vars) -->

<!-- # Subset KCM for isoniazid -->
<!-- isoniazid_KCM <- KCM[KCM[,"Gene"] == "ahpC", ] -->

<!-- # Remove the PCMs in the KCM -->
<!-- isoniazid_PCM_merged <- isoniazid_PCM_merged[!(isoniazid_PCM_merged$change %in% isoniazid_KCM$Mutation), ] -->

<!-- # Transpose summary -->
<!-- isoniazid_summary <- data.frame(n = t(isoniazid_summary)) -->

<!-- # PRM stats -->
<!-- isoniazid_PRM_stats <- select(isoniazid_PRM_stats, gene, mutation, n_samps, dst_prop, dr_prop, n_lins) -->
<!-- isoniazid_PRM_stats <- sort_pos(isoniazid_PRM_stats, 'mutation') -->
<!-- isoniazid_PRM_stats$dst_prop <- 1 - isoniazid_PRM_stats$dst_prop -->
<!-- names(isoniazid_PRM_stats) <- c("Gene-change", "n samples", "DST sensitive", "DR type sensitive", "n lineages") -->

<!-- ``` -->

<!-- ```{r rifampicin-clean-data, echo = F, warning=F, message=F} -->

<!-- # Remove all the "[]" present in the binary table caused by python's lists -->
<!-- rifampicin_binary_table <- clean_binary_table(rifampicin_binary_table) -->

<!-- # Split out the other_vars to make extra row per mutation -->
<!-- rifampicin_binary_split <- data.frame(rifampicin_binary_table %>% -->
<!--   mutate(other_vars = strsplit(as.character(other_vars), "; ")) %>% -->
<!--   unnest(other_vars)) -->

<!-- # Make col for just the other_vars genes, stripping out the change -->
<!-- rifampicin_binary_split$other_vars_gene <- gsub("-.*", "", rifampicin_binary_split$other_vars) -->

<!-- # Subset KCM for rifampicin -->
<!-- rifampicin_KCM <- KCM[KCM[,"Gene"] == "rpoB", ] -->

<!-- # Remove the PCMs in the KCM -->
<!-- rifampicin_PCM_merged <- rifampicin_PCM_merged[!(rifampicin_PCM_merged$change %in% rifampicin_KCM$Mutation), ] -->

<!-- # Transpose summary -->
<!-- rifampicin_summary <- data.frame(n = t(rifampicin_summary)) -->

<!-- # PRM stats -->
<!-- rifampicin_PRM_stats <- select(rifampicin_PRM_stats, gene, mutation, n_samps, dst_prop, dr_prop, n_lins) -->
<!-- rifampicin_PRM_stats <- sort_pos(rifampicin_PRM_stats, 'mutation') -->
<!-- rifampicin_PRM_stats$dst_prop <- 1 - rifampicin_PRM_stats$dst_prop -->
<!-- names(rifampicin_PRM_stats) <- c("Gene-change", "n samples", "DST sensitive", "DR type sensitive", "n lineages") -->

<!-- ``` -->


## Abstract

## Introduction

## Results

RM = any resistance mutation;
KRM = known RM;
PRM = potential RM;
CM = any compensatory mutation;
KCM = known CM;
PCM = potential CM



```{r summary-table, echo = F, warning=F, caption = "Summary of each PRM"}

 # knitr::kable(isoniazid_summary)
knitr::kable(summaries)

```

```{r PRM-summary-table, echo = F, warning=F, caption = "n samples by PRM haplotype"}

# knitr::kable(isoniazid_PRM_summary)
knitr::kable(PRM_summary)

```

```{r PRM-stats-table, echo = F, warning=F, caption = "Number of samples, DST sensitive proportion, (predicted) DR type proportion and number of lineages for each PRM"}

# knitr::kable(isoniazid_PRM_stats)
knitr::kable(PRM_stats)

```

<!-- **Rare isoniazid (katG) variants** -->

**Rare variants**

```{r isoniazid-rare-mutations, echo = F, warning=F}

isoniazid_binary_table <- binary_tables$isoniazid
rare_katg_table <- subset(isoniazid_binary_table, !(is.na(CM)) & is.na(KRM) & is.na(PRM) & !(is.na(katG)) )
n_samps_rare_katg <- nrow(rare_katg_table)
n_rare_katg <- length(uniq_col(rare_katg_table$katG))
rare_katg_cnts <- table(table(unlist(strsplit(rare_katg_table$katG, "; "))))

```

```{r rifampicin-rare-mutations, echo = F, warning=F}

rifampicin_binary_table <- binary_tables$rifampicin
rare_rpob_table <- subset(rifampicin_binary_table, !(is.na(CM)) & is.na(KRM) & is.na(PRM) & !(is.na(rpoB)) )
n_samps_rare_rpob <- nrow(rare_rpob_table)
n_rare_rpob <- length(uniq_col(rare_rpob_table$rpoB))
rare_rpob_cnts <- table(table(unlist(strsplit(rare_rpob_table$rpoB, "; "))))

```

```{r rare-mutations, echo = F, warning=F}



```


```{r basic-stats, echo = F, warning=F}

# Basic stats

# n_samps_CM_and_no_KRM_and_no_PRM <- isoniazid_summary["n_samps_CM_and_no_KRM_and_no_PRM", ]
# summaries["n_samps_CM_and_no_KRM_and_no_PRM", "isoniazid"]

```

<!-- The `r n_samps_CM_and_no_KRM_and_no_PRM` samples with CM but no KRM and no PRM have many INH mutations just found in one or two of these samples (‘*rare INH mutations*’).  -->

<!-- e.g. Looking at TBProfiler - sample ERR1035291 has CM ahpC-c.-74G>A and missense katG-p.Thr394Pro, but this is the only sample with this *katG* mutation.  -->

<!-- `r n_samps_rare_katg`/`r n_samps_CM_and_no_KRM_and_no_PRM` of the samples with CM but no KRM and no PRM have any katG missense mutation, with `r n_rare_katg` distinct mutations. -->

<!-- Most of these occur only once or twice and were therefore filtered out. -->

<!-- The mutations occurring 3 or 4 times were only in one lineage so were also filtered -->


```{r rare-counts, echo = F, warning=F, caption = "Frequency of low-occurring variants"}

`r knitr::kable(rare_katg_cnts)`

```

**Samples with PRMs - co-occurrences of other mutations**

```{r isoniazid-PRM-KRM, echo = F, warning=F}

isoniazid_PRM_KRM <- pivot(isoniazid_PRM_tab, formula(KRM ~ 'n'))

# isoniazid_PRM_fabg <- pivot(isoniazid_PRM_tab, formula(fabG1 ~ 'n'))
isoniazid_PRM_other <- pivot(isoniazid_PRM_tab, formula(other_vars ~ 'n'))

isoniazid_PRM_KRM <- cbind('status' = rep('KRM', nrow(isoniazid_PRM_KRM)), isoniazid_PRM_KRM)
isoniazid_PRM_other <- cbind('status' = rep('unknown', nrow(isoniazid_PRM_other)), isoniazid_PRM_other)

isoniazid_PRM_KRM_other <- rbind_force(isoniazid_PRM_KRM, isoniazid_PRM_other)
isoniazid_PRM_KRM_other <- dplyr::rename(isoniazid_PRM_KRM_other, mutation = KRM)
names(isoniazid_PRM_KRM_other) <- c("Status", "Gene-change", "n samples (%)")

```

KRM and unknown mutations co-occurring in samples with PRM:
`r knitr::kable(isoniazid_PRM_KRM_other)`

Comparison to Ser315Thr and Ser315Thr-KRM in other isoniazid resistance genes.

```{r isoniazid-Ser315Thr, echo = F, warning=F}

# total n samps with a Ser315Thr mutation:  6877
# n samples with p.Ser315Thr and a fabG1 DR mutation:  909
# n samples with p.Ser315Thr and no fabG1 DR mutations:  5968
# proportion:  0.132
# n samps with a p.Ser315Thr mutation and a comp mutation:  177
# proportion:  0.026

ser315thr_tab <- subset(isoniazid_binary_table, grepl("Ser315Thr", isoniazid_binary_table$KRM))

# Move ser315thr string to front e.g. fabG1-c.-15C>T; katG-p.Ser315Thr -> katG-p.Ser315Thr; fabG1-c.-15C>T
ser315thr_tab$KRM <- ifelse(grepl("; katG-p.Ser315Thr", ser315thr_tab$KRM), swap_str(ser315thr_tab$KRM), ser315thr_tab$KRM)
# Drop the specific change for the co-occurring gene - e.g. katG-p.Ser315Thr; fabG1-c.-15C>T ->  katG-p.Ser315Thr; fabG1
ser315thr_tab$KRM_2 <- drop_change(ser315thr_tab$KRM)
ser315thr_pivot <- pivot(ser315thr_tab, formula(KRM_2 ~ 'n'))
ser315thr_pivot <- dplyr::rename(ser315thr_pivot, mutation = KRM_2)

```

`r knitr::kable(ser315thr_pivot)`



##### Rifampicin

**Summary:**

```{r rifampicin-summary-table, echo = F, warning=F}

# Summary

n_samps_CM_and_no_KRM_and_no_PRM <- rifampicin_summary["n_samps_CM_and_no_KRM_and_no_PRM", ]
 
# PRM subset and summary
rifampicin_PRM_tab <- subset(rifampicin_binary_table, !(is.na(PRM)))
rifampicin_PRM_summary <- pivot(rifampicin_PRM_tab, formula(PRM ~ 'n'))

```

`r knitr::kable(rifampicin_summary)`

**Rare rifampicin variants**

```{r rifampicin-rare-mutations, echo = F, warning=F}

rifampicin_rare_table <- subset(rifampicin_binary_table, !(is.na(CM)) & is.na(KRM) & is.na(PRM) & (!(is.na(rpoC)) | !(is.na(rpoB))) )
n_samps_rare_rifampicin <- nrow(rifampicin_rare_table)
n_rare_rifampicin <- length(uniq_col(rifampicin_rare_table$other_vars))
rare_rifampicin_table <- table(unlist(strsplit(rifampicin_rare_table$other_vars, "; ")))

```

`r knitr::kable(rare_rifampicin_table)`



## Methods

**Sequence data and processing**


### Figures

```{r isoniazid-tree, echo=FALSE, out.width = "2000px", fig.cap="Tree of isoniazid samples with potential new katG mutations" }

# Run tree script

# system(paste("Rscript file_to_source.R", arg1, arg2))
# 
# for (drug in drugs){
#   tree_file <- paste0(newick_dir, drug, ".filt.val.gt.g.snps.fa.treefile")
#   
# system(sprintf("Rscript comp_mut_tree.R 
#                --tree_file %s 
#                --metadata_file %s
#                --project_code %s
#                --column drug 
#                --outfile %s", tree_file, , , ))

knitr::include_graphics(isoniazid_tree_file)

```

### Tables



### References
































